// @zegocloud/zego-uikit-rn/lib/commonjs/signal_plugin/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _services = _interopRequireDefault(require("./services"));
var _defines = require("./defines");
var _room_properties_service = _interopRequireDefault(require("./services/room_properties_service"));
var _in_room_message_service = _interopRequireDefault(require("./services/in_room_message_service"));
var _user_in_room_attributes_service = _interopRequireDefault(require("./services/user_in_room_attributes_service"));
var _ZegoUIKitSignalingPlugin;
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
class ZegoUIKitSignalingPlugin {
  constructor() {
    _defineProperty(this, "_signaling", _defines.ZegoUIKitPluginType.signaling);
    if (!ZegoUIKitSignalingPlugin.shared) {
      ZegoUIKitSignalingPlugin.shared = this;
    }
    return ZegoUIKitSignalingPlugin.shared;
  }
  static getInstance() {
    if (!ZegoUIKitSignalingPlugin.shared) {
      ZegoUIKitSignalingPlugin.shared = new ZegoUIKitSignalingPlugin();
    }
    return ZegoUIKitSignalingPlugin.shared;
  }
  static getVersion() {
    return _services.default.getInstance().getVersion().then(zimVersion => {
      return `signaling_plugin:1.0.0;zim:${zimVersion}`;
    });
  }
  getZIMInstance() {
    return _services.default.getInstance().getZIMInstance();
  }
  getPluginType() {
    return this._signaling;
  }
  setBackgroundMessageHandler() {
    _services.default.getInstance().setBackgroundMessageHandler();
  }
  setThroughMessageReceivedHandler() {
    _services.default.getInstance().setThroughMessageReceivedHandler();
  }
  setAndroidOfflineDataHandler(handler) {
    _services.default.getInstance().setAndroidOfflineDataHandler(handler);
  }
  setIOSOfflineDataHandler(handler) {
    _services.default.getInstance().setIOSOfflineDataHandler(handler);
  }
  onCallKitAnswerCall(handler) {
    _services.default.getInstance().onCallKitAnswerCall(handler);
  }
  onCallKitEndCall(handler) {
    _services.default.getInstance().onCallKitEndCall(handler);
  }
  reportCallKitCallEnded(uuid, reason) {
    _services.default.getInstance().reportCallKitCallEnded(uuid, reason);
  }
  reportIncomingCall(cxCallUpdate, uuid) {
    return _services.default.getInstance().reportIncomingCall(cxCallUpdate, uuid);
  }
  queryCallList(count, nextFlag) {
    return _services.default.getInstance().queryCallList(count, nextFlag);
  }
  sendInRoomTextMessage(roomID, message) {
    return _in_room_message_service.default.getInstance().sendInRoomTextMessage(roomID, message);
  }
  sendInRoomCommandMessage(roomID, message) {
    return _in_room_message_service.default.getInstance().sendInRoomCommandMessage(roomID, message);
  }
  renewToken(token) {
    return _services.default.getInstance().renewToken(token);
  }
  invoke(method, params) {
    switch (method) {
      case 'init':
        return _services.default.getInstance().init(params.appID, params.appSign);
      case 'uninit':
        return _services.default.getInstance().uninit();
      case 'login':
        return _services.default.getInstance().login(params.userID, params.userName, params.token);
      case 'enableNotifyWhenAppRunningInBackgroundOrQuit':
        return _services.default.getInstance().enableNotifyWhenAppRunningInBackgroundOrQuit(params.certificateIndex, params.isIOSDevelopmentEnvironment, params.appName);
      case 'setAdvancedConfig':
        return _services.default.getInstance().setAdvancedConfig(params.key, params.value);
      case 'logout':
        return _services.default.getInstance().logout();
      case 'sendInvitation':
        return _services.default.getInstance().sendInvitation(params.inviterName, params.invitees, params.timeout, params.type, params.data, params.notificationConfig);
      case 'cancelInvitation':
        return _services.default.getInstance().cancelInvitation(params.invitees, params.data);
      case 'reportCallKitCallEnded':
        return _services.default.getInstance().reportCallKitCallEnded(params.uuid, params.reason);
      case 'reportIncomingCall':
        return _services.default.getInstance().reportIncomingCall(params.cxCallUpdate, params.uuid);
      case 'refuseInvitation':
        return _services.default.getInstance().refuseInvitation(params.inviterID, params.data);
      case 'acceptInvitation':
        return _services.default.getInstance().acceptInvitation(params.inviterID, params.data);
      // UserInRoomAttributes
      case 'joinRoom':
        return _user_in_room_attributes_service.default.getInstance().joinRoom(params.roomID);
      case 'leaveRoom':
        return _user_in_room_attributes_service.default.getInstance().leaveRoom();
      case 'leaveAllRoom':
        return _user_in_room_attributes_service.default.getInstance().leaveAllRoom();
      case 'setUsersInRoomAttributes':
        return _user_in_room_attributes_service.default.getInstance().setUsersInRoomAttributes(params.key, params.value, params.userIDs);
      case 'queryUsersInRoomAttributes':
        return _user_in_room_attributes_service.default.getInstance().queryUsersInRoomAttributes(params.nextFlag, params.count);
      // RoomProperties
      case 'updateRoomProperty':
        return _room_properties_service.default.getInstance().updateRoomProperty(params.key, params.value, params.isDeleteAfterOwnerLeft, params.isForce, params.isUpdateOwner);
      case 'deleteRoomProperties':
        return _room_properties_service.default.getInstance().deleteRoomProperties(params.keys, params.isForce);
      case 'beginRoomPropertiesBatchOperation':
        return _room_properties_service.default.getInstance().beginRoomPropertiesBatchOperation(params.isDeleteAfterOwnerLeft, params.isForce, params.isUpdateOwner);
      case 'endRoomPropertiesBatchOperation':
        return _room_properties_service.default.getInstance().endRoomPropertiesBatchOperation();
      case 'queryRoomProperties':
        return _room_properties_service.default.getInstance().queryRoomProperties();
      default:
        return Promise.reject();
    }
  }
  registerPluginEventHandler(event, callbackID, callback) {
    switch (event) {
      case 'connectionStateChanged':
        _services.default.getInstance().onConnectionStateChanged(callbackID, callback);
        break;
      case 'invitationReceived':
        _services.default.getInstance().onCallInvitationReceived(callbackID, callback);
        break;
      case 'invitationTimeout':
        _services.default.getInstance().onCallInvitationTimeout(callbackID, callback);
        break;
      case 'invitationResponseTimeout':
        _services.default.getInstance().onCallInviteesAnsweredTimeout(callbackID, callback);
        break;
      case 'invitationAccepted':
        _services.default.getInstance().onCallInvitationAccepted(callbackID, callback);
        break;
      case 'invitationRefused':
        _services.default.getInstance().onCallInvitationRejected(callbackID, callback);
        break;
      case 'invitationCanceled':
        _services.default.getInstance().onCallInvitationCancelled(callbackID, callback);
        break;
      // UserInRoomAttributes
      case 'usersInRoomAttributesUpdated':
        _user_in_room_attributes_service.default.getInstance().onUsersInRoomAttributesUpdated(callbackID, callback);
        break;
      // RoomProperties
      case 'roomPropertyUpdated':
        _room_properties_service.default.getInstance().onRoomPropertyUpdated(callbackID, callback);
        break;
      case 'inRoomTextMessageReceived':
        _in_room_message_service.default.getInstance().onInRoomTextMessageReceived(callbackID, callback);
        break;
      case 'onInRoomCommandMessageReceived':
        _in_room_message_service.default.getInstance().onInRoomCommandMessageReceived(callbackID, callback);
        break;
      case 'onRequireNewToken':
        _services.default.getInstance().onRequireNewToken(callbackID, callback);
        break;
      case 'onLoginSuccess':
        _services.default.getInstance().onLoginSuccess(callbackID, callback);
        break;
      default:
        break;
    }
  }
}
exports.default = ZegoUIKitSignalingPlugin;
_ZegoUIKitSignalingPlugin = ZegoUIKitSignalingPlugin;
_defineProperty(ZegoUIKitSignalingPlugin, "shared", void 0);
