// @zegocloud/zego-uikit-prebuilt-call-rn/lib/commonjs/call_invitation/services/bell.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _reactNative = require("react-native");
var _reactNativeSound = _interopRequireDefault(require("react-native-sound"));
var _logger = require("../../utils/logger");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const {
  ZegoUIKitPrebuiltCallRNModule
} = _reactNative.NativeModules;

// console.warn('Sound', Sound);
// console.warn('Vibration', Vibration);

const BellManage = {
  _incomingCallFileName: '',
  _outgoingCallFileName: '',
  _whooshIncoming: null,
  _whooshOutgoing: null,
  vibrate: () => {
    _reactNative.Vibration.vibrate([0, 500, 1000, 500], true);
  },
  cancleVirate: () => {
    _reactNative.Vibration.cancel();
  },
  initRingtoneConfig: ringtoneConfig => {
    BellManage._incomingCallFileName = ringtoneConfig.incomingCallFileName;
    BellManage._outgoingCallFileName = ringtoneConfig.outgoingCallFileName;
  },
  initIncomingSound: () => {
    (0, _logger.zloginfo)('[BellManage][initIncomingSound]');
    // Sound.setCategory('Playback');
    return new Promise((resolve, reject) => {
      BellManage._whooshIncoming = new _reactNativeSound.default(BellManage._incomingCallFileName, _reactNativeSound.default.MAIN_BUNDLE, error => {
        if (error) {
          (0, _logger.zlogerror)('Failed to load the sound of zego_incoming', error);
          reject();
        } else {
          (0, _logger.zloginfo)('Load the sound of zego_incoming successfully');
          BellManage._whooshIncoming.setNumberOfLoops(-1);
          resolve(BellManage._whooshIncoming);
        }
      });
    });
  },
  initOutgoingSound: () => {
    return new Promise((resolve, reject) => {
      BellManage._whooshOutgoing = new _reactNativeSound.default(BellManage._outgoingCallFileName, _reactNativeSound.default.MAIN_BUNDLE, error => {
        if (error) {
          (0, _logger.zlogerror)('Failed to load the sound of zego_outgoing', error);
          reject();
        } else {
          (0, _logger.zloginfo)('Load the sound of zego_outgoing successfully');
          BellManage._whooshOutgoing.setNumberOfLoops(-1);
          resolve(BellManage._whooshOutgoing);
        }
      });
    });
  },
  playIncomingSound: () => {
    ZegoUIKitPrebuiltCallRNModule.changeToSpeaker();
    BellManage._whooshIncoming.play(success => {
      if (success) {
        (0, _logger.zloginfo)('successfully finished playing zego_incoming');
      } else {
        (0, _logger.zloginfo)('playback zego_incoming failed due to audio decoding errors');
      }
    });
    (0, _logger.zloginfo)('successfully playing zego_incoming');
  },
  stopIncomingSound: () => {
    if (!BellManage._whooshIncoming) {
      (0, _logger.zloginfo)('[BellManage][stopOutgoingSound] _whooshIncoming is null');
      return;
    }
    BellManage._whooshIncoming.stop(() => {
      (0, _logger.zloginfo)('successfully stop playing zego_incoming');
    });
  },
  playOutgoingSound: () => {
    ZegoUIKitPrebuiltCallRNModule.changeToReceiver();
    BellManage._whooshOutgoing.play(success => {
      if (success) {
        (0, _logger.zloginfo)('successfully finished playing zego_outgoing');
      } else {
        (0, _logger.zloginfo)('playback zego_outgoing failed due to audio decoding errors');
      }
    });
    (0, _logger.zloginfo)('successfully playing zego_outgoing');
  },
  stopOutgoingSound: () => {
    BellManage._whooshOutgoing.stop(() => {
      (0, _logger.zloginfo)('successfully stop playing zego_outgoing');
    });
  },
  releaseIncomingSound: () => {
    (0, _logger.zloginfo)('[BellManage][releaseIncomingSound]');
    BellManage._whooshIncoming.release();
    BellManage._whooshIncoming = null;
    (0, _logger.zloginfo)('successfully release the audio player resource');
  },
  releaseOutgoingSound: () => {
    BellManage._whooshOutgoing.release();
    BellManage._whooshOutgoing = null;
    (0, _logger.zloginfo)('successfully release the audio player resource');
  }
};
var _default = exports.default = BellManage;
