// @zegocloud/zego-uikit-prebuilt-call-rn/lib/commonjs/call_invitation/services/call_invite_helper.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _invite_state_manager = _interopRequireDefault(require("./invite_state_manager"));
var _bell = _interopRequireDefault(require("./bell"));
var _logger = require("../../utils/logger");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
class CallInviteHelper {
  constructor() {
    _defineProperty(this, "_currentCallUUID", void 0);
    _defineProperty(this, "_instance", void 0);
    _defineProperty(this, "_offlineData", void 0);
    _defineProperty(this, "onAcceptCallbackMap", {});
    _defineProperty(this, "onRefuseCallbackMap", {});
  }
  static getInstance() {
    return this._instance || (this._instance = new CallInviteHelper());
  }
  setOfflineData(data) {
    (0, _logger.zloginfo)('[CallInviteHelper][setOfflineData]', data);
    this._offlineData = data;
  }
  getOfflineData() {
    return this._offlineData;
  }
  acceptCall(callID, data) {
    (0, _logger.zloginfo)(`[CallInviteHelper][acceptCall] callID: ${callID}, data: ${JSON.stringify(data)}`);
    _invite_state_manager.default.updateInviteDataAfterAccepted(callID);
    this.notifyAccept(data);
  }
  refuseCall(callID) {
    _invite_state_manager.default.updateInviteDataAfterRejected(callID);
    this.notifyRefuse();
  }
  notifyAccept(data) {
    Object.keys(this.onAcceptCallbackMap).forEach(callbackID => {
      if (this.onAcceptCallbackMap[callbackID]) {
        this.onAcceptCallbackMap[callbackID](data);
      }
    });
  }
  onCallAccepted(callbackID, callback) {
    if (typeof callback !== 'function') {
      delete this.onAcceptCallbackMap[callbackID];
    } else {
      this.onAcceptCallbackMap[callbackID] = callback;
    }
  }
  notifyRefuse() {
    Object.keys(this.onRefuseCallbackMap).forEach(callbackID => {
      if (this.onRefuseCallbackMap[callbackID]) {
        this.onRefuseCallbackMap[callbackID]();
      }
    });
  }
  onCallRefused(callbackID, callback) {
    if (typeof callback !== 'function') {
      delete this.onRefuseCallbackMap[callbackID];
    } else {
      this.onRefuseCallbackMap[callbackID] = callback;
    }
  }
}
exports.default = CallInviteHelper;
