// @zegocloud/zego-uikit-prebuilt-call-rn/lib/commonjs/call_invitation/services/offline_call_event_listener.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _reactNative = require("react-native");
var _zegoUikitRn = _interopRequireWildcard(require("@zegocloud/zego-uikit-rn"));
var _logger = require("../../utils/logger");
var _package_version = require("../../utils/package_version");
var _inner_text_helper = _interopRequireDefault(require("../services/inner_text_helper"));
var _notification_helper = _interopRequireDefault(require("./notification_helper"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
const {
  ZegoUIKitPrebuiltCallRNModule
} = _reactNative.NativeModules;
class OfflineCallEventListener {
  constructor() {
    _defineProperty(this, "TAG", 'OfflineCallEventListener');
    _defineProperty(this, "_instance", void 0);
    _defineProperty(this, "config", {});
    _defineProperty(this, "_isSystemCalling", false);
  }
  static getInstance() {
    return this._instance || (this._instance = new OfflineCallEventListener());
  }
  useSystemCallingUI(plugins = []) {
    this._isSystemCalling = true;
    _zegoUikitRn.default.installPlugins(plugins);
    _zegoUikitRn.default.logComponentsVersion(new Map([['PrebuiltCall', (0, _package_version.getPackageVersion)()]]));
    const signalingPlugin = _zegoUikitRn.default.getSignalingPlugin().getZegoUIKitSignalingPlugin();

    // https://doc-zh.zego.im/article/17416
    signalingPlugin.getInstance().setBackgroundMessageHandler();
    signalingPlugin.getInstance().setAndroidOfflineDataHandler(async data => {
      (0, _logger.zloginfo)('OfflineDataHandler: ', data);
      _notification_helper.default.getInstance().init(plugins);
      const cancelInvitation = data && data.operation_type === "cancel_invitation";
      if (!cancelInvitation) {
        _notification_helper.default.getInstance().showOfflineNotification(data.zim_call_id, Date.now().toString(), data);
      } else {
        _notification_helper.default.getInstance().dismissNotification(data.zim_call_id, 'BeCancelled', this.TAG);
      }
    });
    signalingPlugin.getInstance().setIOSOfflineDataHandler((data, callUUID) => {
      (0, _logger.zloginfo)("setIOSOfflineDataHandler", callUUID, data);
      _zegoUikitRn.default.getSignalingPlugin().setAdvancedConfig('zim_voip_call_id', data.zim_call_id);
      _notification_helper.default.getInstance().init();
      _notification_helper.default.getInstance().showOfflineNotification(data.zim_call_id, callUUID, data);
    });
    _zegoUikitRn.ZegoUIKitLogger.logFlush();
  }
  init(config) {
    (0, _logger.zloginfo)('######OfflineCallEventListener init', config);
    this.config = config;

    // Setup for background invitation
    if (!this._isSystemCalling) return;
    this.setupOnlineCallKit();
  }

  // For Android
  setupOnlineCallKit() {
    const {
      ringtoneConfig
    } = this.config;
    if (!ringtoneConfig) {
      (0, _logger.zloginfo)('ringtoneConfig is required for calling notification');
    }
    if (_reactNative.Platform.OS === 'android' && ringtoneConfig && ringtoneConfig.incomingCallFileName) {
      (0, _logger.zloginfo)('call ZegoUIKitPrebuiltCallRNModule.setupCallKit');
      ZegoUIKitPrebuiltCallRNModule.setupCallKit({
        incomingCallFileName: ringtoneConfig.incomingCallFileName,
        incomingDeclineButtonText: _inner_text_helper.default.instance().getInnerText().incomingCallPageDeclineButton,
        incomingAcceptButtonText: _inner_text_helper.default.instance().getInnerText().incomingCallPageAcceptButton
      });
    }
  }
  uninit() {
    (0, _logger.zloginfo)('######OfflineCallEventListener uninit');
  }
}
exports.default = OfflineCallEventListener;
