// @zegocloud/zego-uikit-prebuilt-call-rn/lib/commonjs/call_invitation/services/plugins.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _zegoUikitRn = _interopRequireWildcard(require("@zegocloud/zego-uikit-rn"));
var _logger = require("../../utils/logger");
var _reactNativeEncryptedStorage = _interopRequireDefault(require("react-native-encrypted-storage"));
var _EventEmitter = require("../../utils/EventEmitter");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
const _appInfo = {};
const _localUser = {};
let _pluginConnectionState;
let ZIMKitPlugin = null;
const _install = plugins => {
  _zegoUikitRn.default.installPlugins(plugins);
  plugins.forEach(plugin => {
    if (plugin.ZIMKit) {
      (0, _logger.zloginfo)('[Plugins] install ZIMKit success.');
      ZIMKitPlugin = plugin;
    } else if (plugin.default && typeof plugin.default.getModuleName === 'function') {
      const temp = plugin.default.getModuleName();
      if (temp === 'ZIMKit') {
        (0, _logger.zloginfo)('[Plugins] install ZIMKit success.');
        ZIMKitPlugin = plugin;
      }
    }
  });
};
const _storeLoginInfo = async () => {
  try {
    await _reactNativeEncryptedStorage.default.setItem("_ZegoUIKitPrebuiltCallLoginInfo", JSON.stringify({
      userID: _localUser.userID,
      userName: _localUser.userName,
      appID: _appInfo.appID,
      appSign: _appInfo.appSign
    }));
  } catch (error) {
    (0, _logger.zlogerror)('[Plugins] _storeLoginInfo error', error);
  }
};
const ZegoPrebuiltPlugins = {
  init: (appID, appSign, userID, userName, plugins) => {
    _zegoUikitRn.ZegoUIKitLogger.logSetUserID(userID);
    (0, _logger.zloginfo)('[ZegoPrebuiltPlugins] init', {
      appID,
      userID,
      userName
    });
    const callbackID = 'ZegoPrebuiltPlugins' + String(Math.floor(Math.random() * 10000));
    _install(plugins);
    _zegoUikitRn.default.getSignalingPlugin().init(appID, appSign);
    _zegoUikitRn.default.getSignalingPlugin().onConnectionStateChanged(callbackID, ({
      state
    }) => {
      _pluginConnectionState = state;
    });
    _appInfo.appID = appID;
    _appInfo.appSign = appSign;
    _localUser.userID = userID;
    _localUser.userName = userName;
    _storeLoginInfo();
    return _zegoUikitRn.default.getSignalingPlugin().login(userID, userName).then(() => {
      (0, _logger.zloginfo)(`[Plugins] login success. userID: ${userID}, userName: ${userName}`);
      _EventEmitter.eventEmitter.emit(_EventEmitter.EventName.LOGINED, {
        userID,
        userName
      });
    });
  },
  reconnectIfDisconnected: () => {
    (0, _logger.zloginfo)('[Plugins] reconnectIfDisconnected', _pluginConnectionState, _zegoUikitRn.ZegoInvitationConnectionState.disconnected);
    if (_pluginConnectionState === _zegoUikitRn.ZegoInvitationConnectionState.disconnected) {
      _zegoUikitRn.default.getSignalingPlugin().logout().then(() => {
        (0, _logger.zloginfo)('[Plugins] auto logout success.');
        _zegoUikitRn.default.getSignalingPlugin().login(_localUser.userID, _localUser.userName).then(() => {
          (0, _logger.zloginfo)('[Plugins] auto reconnect success.');
        });
      });
    }
  },
  uninit: () => {
    (0, _logger.zloginfo)('[ZegoPrebuiltPlugins] uninit');
    delete _appInfo.appID;
    delete _appInfo.appSign;
    delete _localUser.userID;
    delete _localUser.userName;
    _zegoUikitRn.default.getSignalingPlugin().logout();
    _zegoUikitRn.default.getSignalingPlugin().uninit();
  },
  getLocalUser: () => {
    return _localUser;
  },
  getAppInfo: () => {
    return _appInfo;
  },
  getZIMKitPlugin: () => {
    return ZIMKitPlugin;
  },
  loadLoginInfoFromLocalEncryptedStorage: async () => {
    try {
      const loginInfoStr = await _reactNativeEncryptedStorage.default.getItem("_ZegoUIKitPrebuiltCallLoginInfo");
      if (loginInfoStr) {
        return JSON.parse(loginInfoStr);
      }
    } catch (error) {
      (0, _logger.zlogerror)('[Plugins] getAppInfoFromLocalEncryptedStorage error', error);
      return Promise.resolve(undefined);
    }
  }
};
var _default = exports.default = ZegoPrebuiltPlugins;
