// @zegocloud/zego-uikit-rn/lib/module/signal_plugin/core/room_properties_core.js
var _ZegoPluginRoomPropertiesCore;
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
import { zlogerror, zloginfo } from '../utils/logger';
import ZegoPluginResult from './defines';
import ZegoPluginUserInRoomAttributesCore from './user_in_room_attributes_core';
import ZegoUIKitCorePlugin from "../../components/internal/ZegoUIKitCorePlugin";
export default class ZegoPluginRoomPropertiesCore {
  constructor() {
    _defineProperty(this, "_onRoomPropertyUpdatedCallbackMap", {});
    if (!ZegoPluginRoomPropertiesCore.shared) {
      ZegoPluginRoomPropertiesCore.shared = this;
    }
    return ZegoPluginRoomPropertiesCore.shared;
  }
  static getInstance() {
    if (!ZegoPluginRoomPropertiesCore.shared) {
      ZegoPluginRoomPropertiesCore.shared = new ZegoPluginRoomPropertiesCore();
    }
    return ZegoPluginRoomPropertiesCore.shared;
  }
  // ------- internal events register ------
  _registerEngineCallback() {
    ZegoUIKitCorePlugin.getZIMPlugin().default.getInstance().on('roomAttributesUpdated', (zim, {
      roomID,
      infos
    }) => {
      zloginfo(`[ZegoPluginRoomPropertiesCore]NotifyRoomPropertiesUpdated`, infos);
      infos.forEach(info => {
        this._notifyRoomPropertiesUpdated(info);
      });
    });
    ZegoUIKitCorePlugin.getZIMPlugin().default.getInstance().on('roomAttributesBatchUpdated', (zim, {
      roomID,
      infos
    }) => {
      zloginfo(`[ZegoPluginRoomPropertiesCore]NotifyRoomPropertiesUpdated`, infos);
      infos.forEach(info => {
        this._notifyRoomPropertiesUpdated(info);
      });
    });
    zloginfo('[ZegoPluginRoomPropertiesCore]Register callback for ZIM...');
  }
  _unregisterEngineCallback() {
    zloginfo('[ZegoPluginRoomPropertiesCore]Unregister callback from ZIM...');
    ZegoUIKitCorePlugin.getZIMPlugin().default.getInstance().off('roomAttributesUpdated');
    ZegoUIKitCorePlugin.getZIMPlugin().default.getInstance().off('roomAttributesBatchUpdated');
  }
  // ------- internal events exec ------
  _notifyRoomPropertiesUpdated(notifyData) {
    Object.keys(this._onRoomPropertyUpdatedCallbackMap).forEach(callbackID => {
      if (this._onRoomPropertyUpdatedCallbackMap[callbackID]) {
        this._onRoomPropertyUpdatedCallbackMap[callbackID](notifyData);
      }
    });
  }

  // ------- external method ------
  updateRoomProperty(attributes, config) {
    if (!ZegoUIKitCorePlugin.getZIMPlugin().default.getInstance()) {
      zlogerror('[ZegoPluginRoomPropertiesCore]Please initialize it first.');
      return Promise.reject();
    }
    const roomID = ZegoPluginUserInRoomAttributesCore.getInstance().getRoomBaseInfo().roomID;
    return new Promise((resolve, reject) => {
      ZegoUIKitCorePlugin.getZIMPlugin().default.getInstance().setRoomAttributes(attributes, roomID, config).then(({
        roomID: resRoomID,
        errorKeys
      }) => {
        zloginfo(`[ZegoPluginRoomPropertiesCore]Update the room properties successfully.`);
        resolve({
          ...new ZegoPluginResult('', ''),
          errorKeys
        });
      }).catch(error => {
        zlogerror(`[ZegoPluginRoomPropertiesCore]Failed to update room properties, code: ${error.code}, message: ${error.message}`);
        reject(error);
      });
    });
  }
  deleteRoomProperties(keys, config) {
    if (!ZegoUIKitCorePlugin.getZIMPlugin().default.getInstance()) {
      zlogerror('[ZegoPluginRoomPropertiesCore]Please initialize it first.');
      return Promise.reject();
    }
    const roomID = ZegoPluginUserInRoomAttributesCore.getInstance().getRoomBaseInfo().roomID;
    return new Promise((resolve, reject) => {
      ZegoUIKitCorePlugin.getZIMPlugin().default.getInstance().deleteRoomAttributes(keys, roomID, config).then(({
        roomID: resRoomID,
        errorKeys
      }) => {
        zloginfo(`[ZegoPluginRoomPropertiesCore]Delete the room properties successfully.`);
        resolve({
          ...new ZegoPluginResult('', ''),
          errorKeys
        });
      }).catch(error => {
        zlogerror(`[ZegoPluginRoomPropertiesCore]Failed to delete room properties, code: ${error.code}, message: ${error.message}`);
        reject(error);
      });
    });
  }
  beginRoomPropertiesBatchOperation(config) {
    if (!ZegoUIKitCorePlugin.getZIMPlugin().default.getInstance()) {
      zlogerror('[ZegoPluginRoomPropertiesCore]Please initialize it first.');
      return Promise.reject();
    }
    const roomID = ZegoPluginUserInRoomAttributesCore.getInstance().getRoomBaseInfo().roomID;
    ZegoUIKitCorePlugin.getZIMPlugin().default.getInstance().beginRoomAttributesBatchOperation(roomID, config);
    zloginfo(`[ZegoPluginRoomPropertiesCore]Begin batch operate room properties successfully.`);
  }
  endRoomPropertiesBatchOperation() {
    if (!ZegoUIKitCorePlugin.getZIMPlugin().default.getInstance()) {
      zlogerror('[ZegoPluginRoomPropertiesCore]Please initialize it first.');
      return Promise.reject();
    }
    const roomID = ZegoPluginUserInRoomAttributesCore.getInstance().getRoomBaseInfo().roomID;
    return new Promise((resolve, reject) => {
      ZegoUIKitCorePlugin.getZIMPlugin().default.getInstance().endRoomAttributesBatchOperation(roomID).then(({
        roomID: resRoomID
      }) => {
        zloginfo(`[ZegoPluginRoomPropertiesCore]End batch operate room properties successfully.`);
        resolve(new ZegoPluginResult('', ''));
      }).catch(error => {
        zlogerror(`[ZegoPluginRoomPropertiesCore]Failed to end batch operate room properties, code: ${error.code}, message: ${error.message}`);
        reject(error);
      });
    });
  }
  queryRoomProperties() {
    if (!ZegoUIKitCorePlugin.getZIMPlugin().default.getInstance()) {
      zlogerror('[ZegoPluginRoomPropertiesCore]Please initialize it first.');
      return Promise.reject();
    }
    const roomID = ZegoPluginUserInRoomAttributesCore.getInstance().getRoomBaseInfo().roomID;
    return new Promise((resolve, reject) => {
      ZegoUIKitCorePlugin.getZIMPlugin().default.getInstance().queryRoomAllAttributes(roomID).then(({
        roomID: resRoomID,
        roomAttributes
      }) => {
        zloginfo(`[ZegoPluginRoomPropertiesCore]Query room all attributes successfully.`);
        resolve({
          roomAttributes,
          ...new ZegoPluginResult('', '')
        });
      }).catch(error => {
        zlogerror(`[ZegoPluginRoomPropertiesCore]Failed to query room all properties, code: ${error.code}, message: ${error.message}`);
        reject(error);
      });
    });
  }
  onRoomPropertyUpdated(callbackID, callback) {
    if (!ZegoUIKitCorePlugin.getZIMPlugin().default.getInstance()) {
      zlogerror('[ZegoPluginRoomPropertiesCore]Please initialize it first.');
    }
    if (typeof callback !== 'function') {
      if (callbackID in this._onRoomPropertyUpdatedCallbackMap) {
        zloginfo('[Core][onRoomPropertyUpdated] Remove callback for: [', callbackID, '] because callback is not a valid function!');
        delete this._onRoomPropertyUpdatedCallbackMap[callbackID];
      }
    } else {
      this._onRoomPropertyUpdatedCallbackMap[callbackID] = callback;
    }
  }
}
_ZegoPluginRoomPropertiesCore = ZegoPluginRoomPropertiesCore;
_defineProperty(ZegoPluginRoomPropertiesCore, "shared", void 0);
