// @zegocloud/zego-uikit-prebuilt-call-rn/lib/module/call_invitation/services/callevent_notify_app.js
var _CallEventNotifyApp;
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
import { zloginfo } from "../../utils/logger";
export default class CallEventNotifyApp {
  constructor() {
    // @ts-ignore
    _defineProperty(this, "eventCallbacks", {});
  }
  static getInstance() {
    return this._instance || (this._instance = new CallEventNotifyApp());
  }
  init(callbacks) {
    this.eventCallbacks['onIncomingCallDeclineButtonPressed'] = callbacks.onIncomingCallDeclineButtonPressed;
    this.eventCallbacks['onIncomingCallAcceptButtonPressed'] = callbacks.onIncomingCallAcceptButtonPressed;
    this.eventCallbacks['onIncomingCallReceived'] = callbacks.onIncomingCallReceived;
    this.eventCallbacks['onIncomingCallCanceled'] = callbacks.onIncomingCallCanceled;
    this.eventCallbacks['onIncomingCallTimeout'] = callbacks.onIncomingCallTimeout;
    this.eventCallbacks['onIncomingCallRejectedCauseBusy'] = callbacks.onIncomingCallRejectedCauseBusy;
    this.eventCallbacks['onOutgoingCallCancelButtonPressed'] = callbacks.onOutgoingCallCancelButtonPressed;
    this.eventCallbacks['onOutgoingCallAccepted'] = callbacks.onOutgoingCallAccepted;
    this.eventCallbacks['onOutgoingCallRejectedCauseBusy'] = callbacks.onOutgoingCallRejectedCauseBusy;
    this.eventCallbacks['onOutgoingCallDeclined'] = callbacks.onOutgoingCallDeclined;
    this.eventCallbacks['onOutgoingCallTimeout'] = callbacks.onOutgoingCallTimeout;
    zloginfo(`[CallEventNotifyApp][init] succeed`);
  }
  isFunction(callback) {
    return typeof callback === 'function';
  }
  notifyEvent(event, ...params) {
    let callback = this.eventCallbacks[event];
    if (!this.isFunction(callback)) {
      zloginfo(`[CallEventNotifyApp][${event}] is not configured or the timing is too early`);
      return;
    }
    zloginfo(`[CallEventNotifyApp][${event}] notify will`);
    callback(...params);
    zloginfo(`[CallEventNotifyApp][${event}] notify succeed`);
  }
}
_CallEventNotifyApp = CallEventNotifyApp;
_defineProperty(CallEventNotifyApp, "_instance", void 0);
