// @zegocloud/zego-uikit-prebuilt-call-rn/lib/module/call_invitation/services/hangup_helper.js
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
import { Alert } from 'react-native';
export default class HangupHelper {
  constructor() {
    _defineProperty(this, "_instance", void 0);
    _defineProperty(this, "_debounce", false);
    _defineProperty(this, "_onAutoJumpCallbackMap", {});
  }
  static getInstance() {
    return this._instance || (this._instance = new HangupHelper());
  }
  setDebounce(debounce) {
    this._debounce = !!debounce;
  }
  getDebounce() {
    return this._debounce;
  }
  showLeaveAlert(hangUpConfirmInfo) {
    return new Promise((resolve, reject) => {
      if (hangUpConfirmInfo) {
        const {
          title = "Leave the call",
          message = "Are you sure to leave the call?",
          cancelButtonName = "Cancel",
          confirmButtonName = "Confirm"
        } = hangUpConfirmInfo;
        Alert.alert(title, message, [{
          text: cancelButtonName,
          onPress: () => {
            reject();
          },
          style: "cancel"
        }, {
          text: confirmButtonName,
          onPress: () => {
            resolve();
          }
        }], {
          cancelable: false
        });
      } else {
        resolve();
      }
    });
  }
  notifyAutoJump() {
    Object.keys(this._onAutoJumpCallbackMap).forEach(callbackID => {
      if (this._onAutoJumpCallbackMap[callbackID]) {
        this._onAutoJumpCallbackMap[callbackID]();
      }
    });
  }
  onAutoJump(callbackID, callback) {
    if (typeof callback !== 'function') {
      delete this._onAutoJumpCallbackMap[callbackID];
    } else {
      this._onAutoJumpCallbackMap[callbackID] = callback;
    }
  }
}
