// @zegocloud/zego-uikit-prebuilt-call-rn/lib/commonjs/call_invitation/pages/ZegoCallInvitationWaiting.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ZegoUIKitPrebuiltCallWaitingScreen;
var _react = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
var _native = require("@react-navigation/native");
var _zegoUikitRn = _interopRequireWildcard(require("@zegocloud/zego-uikit-rn"));
var _services = _interopRequireDefault(require("../../services"));
var _logger = require("../../utils/logger");
var _bell = _interopRequireDefault(require("../services/bell"));
var _callevent_notify_app = _interopRequireDefault(require("../services/callevent_notify_app"));
var _invite_state_manager = _interopRequireDefault(require("../services/invite_state_manager"));
var _plugins = _interopRequireDefault(require("../services/plugins"));
var _ZegoCallInvationForeground = _interopRequireDefault(require("./ZegoCallInvationForeground"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
function ZegoUIKitPrebuiltCallWaitingScreen(props) {
  const TAG = 'ZegoUIKitPrebuiltCallWaitingScreen';
  const navigation = (0, _native.useNavigation)();
  const autoBackTimerRef = (0, _react.useRef)(null);
  const isNavigatingBackRef = (0, _react.useRef)(false);
  const {
    appID,
    appSign
  } = _services.default.getInstance().getInitAppInfo();
  const {
    userID,
    userName
  } = _services.default.getInstance().getInitUser();
  const initConfig = _services.default.getInstance().getInitConfig();
  const {
    avatarBuilder,
    waitingPageConfig
  } = initConfig;
  const {
    route
  } = props;
  const {
    roomID,
    isVideoCall,
    invitees,
    inviterID,
    invitationID,
    callName,
    timeout
  } = route.params;
  const getInviteeIDList = () => {
    return invitees.map(invitee => {
      return invitee.userID;
    });
  };
  const grantPermissions = async callback => {
    // Android: Dynamically obtaining device permissions
    if (_reactNative.Platform.OS === 'android') {
      // Check if permission granted
      let grantedAudio = _reactNative.PermissionsAndroid.check(_reactNative.PermissionsAndroid.PERMISSIONS.RECORD_AUDIO);
      let grantedCamera = _reactNative.PermissionsAndroid.check(_reactNative.PermissionsAndroid.PERMISSIONS.CAMERA);
      const ungrantedPermissions = [];
      try {
        const isAudioGranted = await grantedAudio;
        const isVideoGranted = await grantedCamera;
        if (!isAudioGranted) {
          ungrantedPermissions.push(_reactNative.PermissionsAndroid.PERMISSIONS.RECORD_AUDIO);
        }
        if (!isVideoGranted && isVideoCall) {
          ungrantedPermissions.push(_reactNative.PermissionsAndroid.PERMISSIONS.CAMERA);
        }
      } catch (error) {
        ungrantedPermissions.push(_reactNative.PermissionsAndroid.PERMISSIONS.RECORD_AUDIO, _reactNative.PermissionsAndroid.PERMISSIONS.CAMERA);
      }
      // If not, request it
      (0, _logger.zloginfo)(`[ZegoUIKitPrebuiltCallWaitingScreen] requestMultiple, ungrantedPermissions: ${ungrantedPermissions}`);
      return _reactNative.PermissionsAndroid.requestMultiple(ungrantedPermissions).then(result => {
        (0, _logger.zloginfo)(`[ZegoUIKitPrebuiltCallWaitingScreen] requestMultiple, result: ${result}`);
        if (callback) {
          callback();
        }
      });
    } else if (callback) {
      callback();
    }
  };
  const hangUpHandle = () => {
    (0, _logger.zloginfo)('[ZegoUIKitPrebuiltCallWaitingScreen] hangUpHandle, Leave room on waiting page');
    _callevent_notify_app.default.getInstance().notifyEvent('onOutgoingCallCancelButtonPressed', navigation, roomID, invitees, isVideoCall ? 1 : 0);
    if (_invite_state_manager.default.isAutoCancelInvite(invitationID)) {
      _zegoUikitRn.default.getSignalingPlugin().cancelInvitation(getInviteeIDList(), JSON.stringify({
        "call_id": roomID,
        "operation_type": "cancel_invitation",
        'inviter': {
          userID: _plugins.default.getLocalUser().userID,
          userName: _plugins.default.getLocalUser().userName
        }
      }));
      _invite_state_manager.default.updateInviteDataAfterCancel(invitationID);
    }
    navigationBack();
  };
  const configLocalTimeoutProcess = timer => {
    if (timer != null) {
      (0, _logger.zloginfo)('[ZegoUIKitPrebuiltCallWaitingScreen] config localTimeoutProcess');
      autoBackTimerRef.current = timer;
    } else if (autoBackTimerRef.current) {
      (0, _logger.zloginfo)('[ZegoUIKitPrebuiltCallWaitingScreen] remove localTimeoutProcess');
      clearTimeout(autoBackTimerRef.current);
      autoBackTimerRef.current = null;
    }
  };
  const navigationBack = () => {
    if (isNavigatingBackRef.current) {
      (0, _logger.zloginfo)('[ZegoUIKitPrebuiltCallWaitingScreen] navigationBack ignored, already navigating back');
      return;
    }
    isNavigatingBackRef.current = true;
    (0, _logger.zloginfo)('[ZegoUIKitPrebuiltCallWaitingScreen] navigationBack');
    configLocalTimeoutProcess(null);
    _zegoUikitRn.default.uninit();
    _bell.default.stopOutgoingSound();
    _invite_state_manager.default.initInviteData();
    navigation.goBack();
  };
  (0, _react.useEffect)(() => {
    _bell.default.playOutgoingSound();
    _zegoUikitRn.default.init(appID, appSign, {
      userID,
      userName
    }).then(() => {
      grantPermissions(() => {
        _zegoUikitRn.default.turnCameraOn('', isVideoCall);
        _zegoUikitRn.default.turnMicrophoneOn('', true);
        _zegoUikitRn.default.setAudioOutputToSpeaker(true);
      });
    });
    const autoBackTimeout = (timeout ?? 60) + 5;
    (0, _logger.zloginfo)(`[ZegoUIKitPrebuiltCallWaitingScreen] will auto navigationBack after ${autoBackTimeout} seconds`);
    configLocalTimeoutProcess(setTimeout(() => {
      autoBackTimerRef.current = null;
      (0, _logger.zloginfo)('[ZegoUIKitPrebuiltCallWaitingScreen] auto navigationBack by local timer');
      navigationBack();
    }, autoBackTimeout * 1000));
    const unsubscribe1 = navigation.addListener('blur', () => {
      (0, _logger.zloginfo)('[Navigation] ZegoUIKitPrebuiltCallWaitingScreen, blur');
    });
    const unsubscribe2 = navigation.addListener('focus', () => {
      (0, _logger.zloginfo)('[Navigation] ZegoUIKitPrebuiltCallWaitingScreen, focus');
    });
    const unsubscribe3 = navigation.addListener('beforeRemove', () => {
      (0, _logger.zloginfo)('[Navigation] ZegoUIKitPrebuiltCallWaitingScreen, beforeRemove');
    });
    return () => {
      (0, _logger.zloginfo)('[ZegoUIKitPrebuiltCallWaitingScreen] useEffect, unmount');
      configLocalTimeoutProcess(null);
      _bell.default.stopOutgoingSound();
      unsubscribe1();
      unsubscribe2();
      unsubscribe3();
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
  const handleBackButton = () => {
    return true;
  };
  (0, _react.useEffect)(() => {
    (0, _logger.zloginfo)(`ZegoUIKitPrebuiltCallWaitingScreen useEffect, invitationID: ${invitationID}`);
    const callbackID = 'ZegoUIKitPrebuiltCallWaitingScreen' + String(Math.floor(Math.random() * 10000));
    if (invitees.length === 1) {
      _zegoUikitRn.default.getSignalingPlugin().onInvitationResponseTimeout(callbackID, ({
        callID,
        invitees,
        data
      }) => {
        (0, _logger.zloginfo)(`onInvitationResponseTimeout implement by ${TAG}, callID: ${callID}, data: ${data}`);
        navigationBack();
        let dataParsed = data ? JSON.parse(data) : {};
        const dataInDataParsed = dataParsed.data ? JSON.parse(dataParsed.data) : {};
        _callevent_notify_app.default.getInstance().notifyEvent('onOutgoingCallTimeout', dataInDataParsed.call_id, dataParsed.invitees);
        (0, _logger.zloginfo)(`onInvitationResponseTimeout implement by ${TAG}, process done`);
      });
      _zegoUikitRn.default.getSignalingPlugin().onInvitationRefused(callbackID, ({
        callID,
        invitee,
        data
      }) => {
        (0, _logger.zloginfo)(`onInvitationRefused implement by ${TAG}, callID: ${callID}, data: ${data}`);
        _zegoUikitRn.default.uninit();
        const callIDs = Array.from(_invite_state_manager.default._invitationMap.keys());
        if (callIDs.includes(callID)) {
          navigationBack();
        }
        let dataParsed = data ? JSON.parse(data) : {};
        if (dataParsed.reason === 'busy') {
          _callevent_notify_app.default.getInstance().notifyEvent('onOutgoingCallRejectedCauseBusy', dataParsed.call_id, dataParsed.invitee);
        } else {
          _callevent_notify_app.default.getInstance().notifyEvent('onOutgoingCallDeclined', dataParsed.call_id, dataParsed.invitee);
        }
      });

      // After the caller initiates a call and then loses network connectivity, 
      // the server will cancel the call and notify both the caller and the callee that the call has been canceled.
      _zegoUikitRn.default.getSignalingPlugin().onInvitationCanceled(callbackID, result => {
        (0, _logger.zloginfo)(`[ZegoUIKitPrebuiltCallWaitingScreen] onInvitationCanceled, result: ${JSON.stringify(result)}`);
        if (result.callID !== invitationID) {
          (0, _logger.zlogwarning)('[ZegoUIKitPrebuiltCallWaitingScreen] onInvitationCanceled, callID not match');
          return;
        }
        navigationBack();
      });
    } else {
      // CallInviteStateManage.onSomeoneAcceptedInvite(callbackID, () => {
      //   zloginfo('Someone accepted the invitation');
      //   BellManage.stopOutgoingSound();
      // });
      _invite_state_manager.default.onInviteCompletedWithNobody(callbackID, () => {
        (0, _logger.zloginfo)('[ZegoUIKitPrebuiltCallWaitingScreen] onInviteCompletedWithNobody');
        navigationBack();
      });
    }
    _zegoUikitRn.default.getSignalingPlugin().onInvitationAccepted(callbackID, ({
      invitee,
      data
    }) => {
      (0, _logger.zloginfo)(`onInvitationAccepted implement by ${TAG}, invitee: ${JSON.stringify(invitee)}, data: ${data}`);
      configLocalTimeoutProcess(null);
      _bell.default.stopOutgoingSound();

      // ZegoUIKit.leaveRoom().then(() => {
      (0, _logger.zloginfo)('Jump to call room page.');
      const state = navigation.getState();
      (0, _logger.zloginfo)(`Current navigation routes: [${state.routes.map(r => r.name).join(', ')}]`);
      const isAlreadyInCall = state.routes.some(r => r.name === 'ZegoUIKitPrebuiltCallInCallScreen');
      if (isAlreadyInCall) {
        (0, _logger.zloginfo)('Already in ZegoUIKitPrebuiltCallInCallScreen page, do not need to jump again.');
      } else {
        (0, _logger.zloginfo)('Navigate to ZegoUIKitPrebuiltCallInCallScreen');
        navigation.navigate('ZegoUIKitPrebuiltCallInCallScreen', {
          origin: 'ZegoUIKitPrebuiltCallWaitingScreen',
          roomID,
          isVideoCall,
          invitees: getInviteeIDList(),
          inviter: inviterID,
          invitationID,
          useFrontFacingCamera: _zegoUikitRn.default.isUsingFrontFacingCamera()
        });
      }
      // });

      let dataParsed = data ? JSON.parse(data) : {};
      _callevent_notify_app.default.getInstance().notifyEvent('onOutgoingCallAccepted', dataParsed.call_id, dataParsed.invitee);
      (0, _logger.zloginfo)(`onInvitationAccepted implement by ${TAG}, process done`);
    });
    const backHandler = _reactNative.BackHandler.addEventListener('hardwareBackPress', handleBackButton);
    navigation.setOptions({
      gestureEnabled: false
    });
    return () => {
      (0, _logger.zloginfo)('[ZegoUIKitPrebuiltCallWaitingScreen] useEffect, unmount');
      _zegoUikitRn.default.getSignalingPlugin().onInvitationResponseTimeout(callbackID);
      _zegoUikitRn.default.getSignalingPlugin().onInvitationRefused(callbackID);
      _zegoUikitRn.default.getSignalingPlugin().onInvitationCanceled(callbackID);
      _zegoUikitRn.default.getSignalingPlugin().onInvitationAccepted(callbackID);
      _invite_state_manager.default.onInviteCompletedWithNobody(callbackID);
      backHandler.remove();
      navigation.setOptions({
        gestureEnabled: true
      });
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.container
  }, isVideoCall ? /*#__PURE__*/_react.default.createElement(_zegoUikitRn.ZegoAudioVideoView, {
    userID: userID,
    roomID: roomID,
    useVideoViewAspectFill: true
    // eslint-disable-next-line react/no-unstable-nested-components
    ,
    foregroundBuilder: () => /*#__PURE__*/_react.default.createElement(_ZegoCallInvationForeground.default, {
      isVideoCall: isVideoCall,
      invitee: invitees[0],
      onHangUp: hangUpHandle,
      avatarBuilder: avatarBuilder,
      waitingPageConfig: waitingPageConfig,
      callName: callName,
      isGroupCall: invitees.length > 1
    })
  }) : /*#__PURE__*/_react.default.createElement(_ZegoCallInvationForeground.default, {
    isVideoCall: isVideoCall,
    invitee: invitees[0],
    onHangUp: hangUpHandle,
    avatarBuilder: avatarBuilder,
    waitingPageConfig: waitingPageConfig,
    callName: callName,
    isGroupCall: invitees.length > 1
  }), isVideoCall ? /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.topMenuContainer
  }, /*#__PURE__*/_react.default.createElement(_zegoUikitRn.ZegoSwitchCameraButton, {
    iconBuilder: waitingPageConfig.switchCameraBuilder
  })) : /*#__PURE__*/_react.default.createElement(_reactNative.View, null));
}
const styles = _reactNative.StyleSheet.create({
  container: {
    width: '100%',
    height: '100%',
    flex: 1
  },
  topMenuContainer: {
    zIndex: 2,
    width: '100%',
    paddingRight: 15,
    top: 30,
    alignItems: 'flex-end'
  }
});
