// @zegocloud/zego-uikit-prebuilt-call-rn/lib/commonjs/call_invitation/pages/ZegoCallInvationForeground.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ZegoCallInvationForeground;
var _react = _interopRequireDefault(require("react"));
var _reactNative = require("react-native");
var _ZegoBottomBar = _interopRequireDefault(require("../../call/ZegoBottomBar"));
var _ZegoMenuBarButtonName = _interopRequireDefault(require("../../call/ZegoMenuBarButtonName"));
var _defines = require("../services/defines");
var _inner_text_helper = _interopRequireDefault(require("../services/inner_text_helper"));
var _reactDelegateComponent = _interopRequireDefault(require("react-delegate-component"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function ZegoCallInvationForeground(props) {
  const {
    isVideoCall,
    invitee,
    onHangUp,
    avatarBuilder,
    waitingPageConfig,
    callName,
    isGroupCall
  } = props;
  const userName = invitee.userName;
  const getShortName = name => {
    if (!name) {
      return '';
    }
    const nl = name.split(' ');
    var shotName = '';
    nl.forEach(part => {
      if (part !== '') {
        shotName += part.substring(0, 1);
      }
    });
    return shotName;
  };
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.container, waitingPageConfig.backgroundColor ? {
      backgroundColor: waitingPageConfig.backgroundColor
    } : null, isVideoCall ? styles.opacity : null]
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: {
      position: 'absolute',
      width: '100%',
      height: '100%'
    }
  }, waitingPageConfig.backgroundBuilder ? waitingPageConfig.backgroundBuilder(invitee, isVideoCall) : null), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.content
  }, isGroupCall ? /*#__PURE__*/_react.default.createElement(_reactNative.View, null) : waitingPageConfig.avatarBuilder ? waitingPageConfig.avatarBuilder(invitee) : /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.avatar
  }, !avatarBuilder ? /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: styles.nameLabel
  }, getShortName(userName)) : /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: {
      width: '100%',
      height: '100%',
      alignItems: 'center',
      justifyContent: 'center'
    }
  }, /*#__PURE__*/_react.default.createElement(_reactDelegateComponent.default, {
    to: avatarBuilder,
    props: {
      userInfo: invitee
    }
  }))), waitingPageConfig.nameBuilder ? waitingPageConfig.nameBuilder(userName) : /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: styles.userName
  }, callName ? callName : _inner_text_helper.default.instance().getOutgoingCallPageTitle(userName, isVideoCall ? _defines.ZegoInvitationType.videoCall : _defines.ZegoInvitationType.voiceCall)), waitingPageConfig.stateBuilder ? waitingPageConfig.stateBuilder() : /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: styles.calling
  }, _inner_text_helper.default.instance().getOutgoingCallPageMessage(isVideoCall ? _defines.ZegoInvitationType.videoCall : _defines.ZegoInvitationType.voiceCall))), /*#__PURE__*/_react.default.createElement(_ZegoBottomBar.default, {
    menuBarButtons: [_ZegoMenuBarButtonName.default.hangUpButton],
    onHangUp: onHangUp,
    buttonBuilders: {
      hangupBuilder: waitingPageConfig.hangupBuilder
    }
  }));
}
const styles = _reactNative.StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#333333',
    zIndex: 2
  },
  opacity: {
    backgroundColor: 'transparent'
  },
  content: {
    width: '100%',
    zIndex: 2,
    alignItems: 'center',
    marginTop: 114
  },
  avatar: {
    width: 100,
    height: 100,
    borderRadius: 1000,
    backgroundColor: '#ffffff',
    alignItems: 'center',
    justifyContent: 'center',
    marginBottom: 5,
    overflow: 'hidden'
  },
  nameLabel: {
    color: '#222222',
    fontSize: 22
  },
  userName: {
    fontSize: 21,
    lineHeight: 29.5,
    marginBottom: 23.5,
    color: '#ffffff'
  },
  calling: {
    fontSize: 16,
    lineHeight: 22.5,
    color: '#ffffff',
    opacity: 0.7
  }
});
