// @zegocloud/zego-uikit-prebuilt-call-rn/lib/commonjs/call_invitation/pages/ZegoCallInvitationRoom.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ZegoUIKitPrebuiltCallInCallScreen;
var _react = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
var _native = require("@react-navigation/native");
var _zegoUikitRn = _interopRequireDefault(require("@zegocloud/zego-uikit-rn"));
var _call = _interopRequireDefault(require("../../call"));
var _minimizing_helper = _interopRequireDefault(require("../../call/services/minimizing_helper"));
var _prebuilt_helper = _interopRequireDefault(require("../../call/services/prebuilt_helper"));
var _services = _interopRequireDefault(require("../../services"));
var _defines = require("../../services/defines");
var _timing_helper = _interopRequireDefault(require("../../services/timing_helper"));
var _logger = require("../../utils/logger");
var _bell = _interopRequireDefault(require("../services/bell"));
var _defines2 = require("../services/defines");
var _hangup_helper = _interopRequireDefault(require("../services/hangup_helper"));
var _invite_state_manager = _interopRequireDefault(require("../services/invite_state_manager"));
var _notification_helper = _interopRequireDefault(require("../services/notification_helper"));
var _plugins = _interopRequireDefault(require("../services/plugins"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
function ZegoUIKitPrebuiltCallInCallScreen(props) {
  const navigation = (0, _native.useNavigation)();
  const {
    appID,
    appSign
  } = _services.default.getInstance().getInitAppInfo();
  const {
    userID,
    userName
  } = _services.default.getInstance().getInitUser();
  const initConfig = _services.default.getInstance().getInitConfig();
  const {
    requireConfig,
    avatarBuilder,
    requireInviterConfig
  } = initConfig;
  const {
    route
  } = props;
  const isMinimizeSwitch = _minimizing_helper.default.getInstance().getIsMinimizeSwitch();
  (0, _logger.zloginfo)(`#####ZegoUIKitPrebuiltCallInCallScreen####### isMinimizeSwitch: ${isMinimizeSwitch}, route.params: ${JSON.stringify(route.params)}`);
  let routeParams = _prebuilt_helper.default.getInstance().getRouteParams();
  if (!isMinimizeSwitch) {
    routeParams.origin = route.params.origin;
    routeParams.roomID = route.params.roomID;
    routeParams.isVideoCall = route.params.isVideoCall;
    routeParams.invitees = route.params.invitees;
    routeParams.inviter = route.params.inviter;
    routeParams.invitationID = route.params.invitationID;
    routeParams.useFrontFacingCamera = route.params.useFrontFacingCamera;
    _timing_helper.default.getInstance().resetDuration();
  } else {
    routeParams.useFrontFacingCamera = _zegoUikitRn.default.isUsingFrontFacingCamera();
  }
  (0, _logger.zloginfo)(`#####ZegoUIKitPrebuiltCallInCallScreen####### routeParams: ${JSON.stringify(routeParams)}`);
  const {
    origin,
    roomID,
    isVideoCall,
    invitees,
    inviter,
    invitationID,
    useFrontFacingCamera
  } = routeParams;
  const callInvitationData = {
    type: isVideoCall ? _defines2.ZegoInvitationType.videoCall : _defines2.ZegoInvitationType.voiceCall,
    invitees,
    callID: roomID
  };
  const requireDefaultConfig = data => {
    const callConfig = data.invitees.length > 1 ? _defines2.ZegoInvitationType.videoCall === data.type ? _defines.GROUP_VIDEO_CALL_CONFIG : _defines.GROUP_VOICE_CALL_CONFIG : _defines2.ZegoInvitationType.videoCall === data.type ? _defines.ONE_ON_ONE_VIDEO_CALL_CONFIG : _defines.ONE_ON_ONE_VOICE_CALL_CONFIG;
    return {
      ...callConfig
    };
  };
  const config = typeof requireConfig === 'function' ? {
    ...requireDefaultConfig(callInvitationData),
    ...requireConfig(callInvitationData),
    requireInviterConfig
  } : requireDefaultConfig(callInvitationData);

  // Store requireConfig.onCallEnd in PrebuiltHelper.routeParams, so FloatingMinimizedView can call it before notifyDestroyPrebuilt
  if (!isMinimizeSwitch && !routeParams.onCallEnd) {
    var _config$timingConfig;
    routeParams.onCallEnd = config.onCallEnd;
    routeParams.onDurationUpdate = (_config$timingConfig = config.timingConfig) === null || _config$timingConfig === void 0 ? void 0 : _config$timingConfig.onDurationUpdate;
  }
  const _onCallEnd = (roomID, reason, duration) => {
    // reason 0-localHangUp, 1-remoteHangUp, 2-kickOut
    (0, _logger.zloginfo)(`[ZegoUIKitPrebuiltCallInCallScreen][_onCallEnd], roomID: ${roomID}, reason: ${reason}, duration: ${duration}`);
    if (reason === _defines.ZegoCallEndReason.localHangUp) {
      hangUpHandle();
    } else {
      callEndHandle();
    }

    // callback.
    if (typeof config.onCallEnd == 'function') {
      (0, _logger.zloginfo)('[ZegoUIKitPrebuiltCallInCallScreen][_onCallEnd] notify requireConfig.onCallEnd will');
      config.onCallEnd(roomID, reason, duration);
      (0, _logger.zloginfo)('[ZegoUIKitPrebuiltCallInCallScreen][_onCallEnd] notify requireConfig.onCallEnd succeed');
    } else {
      const isMinimize = _minimizing_helper.default.getInstance().getIsMinimize();
      if (isMinimize) {
        _prebuilt_helper.default.getInstance().notifyDestroyPrebuilt();
      } else {
        navigation.goBack();
        origin === 'ZegoUIKitPrebuiltCallWaitingScreen' && navigation.goBack();
      }
    }
    clearAutoHangUpTimer();
    (0, _logger.zloginfo)('[ZegoUIKitPrebuiltCallInCallScreen][_onCallEnd] process done');
  };
  const callEndHandle = () => {
    if (_reactNative.AppState.currentState === 'background') {
      let currentInCallID = _notification_helper.default.getInstance().getCurrentInCallID();
      if (currentInCallID) {
        _notification_helper.default.getInstance().hangUp(currentInCallID, 'other');
      }
    }
  };
  const hangUpHandle = () => {
    callEndHandle();
    // Determine if the current is Inviter
    if (inviter === userID && _invite_state_manager.default.isAutoCancelInvite(invitationID)) {
      _zegoUikitRn.default.getSignalingPlugin().cancelInvitation(invitees, JSON.stringify({
        "call_id": roomID,
        "operation_type": "cancel_invitation",
        'inviter': {
          userID: _plugins.default.getLocalUser().userID,
          userName: _plugins.default.getLocalUser().userName
        }
      }));
      _invite_state_manager.default.updateInviteDataAfterCancel(invitationID);
    }
  };
  let hangUpTimer = null;
  const startAutoHangUpTimer = detectSeconds => {
    if (hangUpTimer) {
      clearTimeout(hangUpTimer);
    }
    (0, _logger.zloginfo)(`[ZegoUIKitPrebuiltCallInCallScreen] startAutoHangUpTimer, will auto hangup after ${detectSeconds} seconds if inviter does not join this call.`);
    const newTimer = setTimeout(() => {
      (0, _logger.zloginfo)('[ZegoUIKitPrebuiltCallInCallScreen] autoHangUp');
      const duration = _timing_helper.default.getInstance().getDuration();
      _onCallEnd(roomID, _defines.ZegoCallEndReason.localHangUp, duration);
    }, detectSeconds * 1000);
    hangUpTimer = newTimer;
  };
  const clearAutoHangUpTimer = () => {
    if (hangUpTimer) {
      (0, _logger.zloginfo)('[ZegoUIKitPrebuiltCallInCallScreen] clearAutoHangUpTimer');
      clearTimeout(hangUpTimer);
      hangUpTimer = null;
    }
  };
  (0, _react.useEffect)(() => {
    (0, _logger.zloginfo)(`ZegoUIKitPrebuiltCallInCallScreen useEffect`);
    const callbackID = 'ZegoUIKitPrebuiltCallInCallScreen ' + String(Math.floor(Math.random() * 10000));

    // if group call from `ZegoUIKitPrebuiltCallWaitingScreen`, don't need add this anymore.
    if (invitees.length > 1 && inviter === userID && origin != 'ZegoUIKitPrebuiltCallWaitingScreen') {
      _bell.default.playOutgoingSound();
      _invite_state_manager.default.onSomeoneAcceptedInvite(callbackID, () => {
        (0, _logger.zloginfo)('Someone accepted the invitation');
        _bell.default.stopOutgoingSound();
      });
      _invite_state_manager.default.onInviteCompletedWithNobody(callbackID, () => {
        (0, _logger.zloginfo)('Invite completed with nobody');
        // navigation.navigate('ZegoInnerChildrenPage');
        // just go back, don't need call `onCallEnd`.
        navigation.goBack();
        origin === 'ZegoUIKitPrebuiltCallWaitingScreen' && invitees.length === 1 && navigation.goBack();
      });
    }
    _hangup_helper.default.getInstance().onAutoJump(callbackID, () => {
      (0, _logger.zloginfo)('########onAutoJump#########', origin, invitees.length);
      hangUpHandle();
      navigation.goBack();
      origin === 'ZegoUIKitPrebuiltCallWaitingScreen' && invitees.length === 1 && navigation.goBack();
    });
    const unsubscribe1 = navigation.addListener('blur', () => {
      (0, _logger.zloginfo)('[Navigation] ZegoUIKitPrebuiltCallInCallScreen, blur');
    });
    const unsubscribe2 = navigation.addListener('focus', () => {
      (0, _logger.zloginfo)('[Navigation] ZegoUIKitPrebuiltCallInCallScreen, focus');
    });
    const unsubscribe3 = navigation.addListener('beforeRemove', () => {
      (0, _logger.zloginfo)('[Navigation] ZegoUIKitPrebuiltCallInCallScreen, beforeRemove');
    });
    return () => {
      (0, _logger.zloginfo)('ZegoUIKitPrebuiltCallInCallScreen useEffect return');
      _hangup_helper.default.getInstance().onAutoJump(callbackID);
      _invite_state_manager.default.onSomeoneAcceptedInvite(callbackID);
      _invite_state_manager.default.onInviteCompletedWithNobody(callbackID);
      _bell.default.stopOutgoingSound();
      _invite_state_manager.default.initInviteData();
      unsubscribe1();
      unsubscribe2();
      unsubscribe3();
    };
  }, []);
  return /*#__PURE__*/_react.default.createElement(_call.default, {
    appID: appID,
    appSign: appSign,
    userID: userID,
    userName: userName,
    callID: roomID,
    config: {
      ...config,
      useFrontFacingCamera,
      onJoinRoom: () => {
        (0, _logger.zloginfo)('[ZegoUIKitPrebuiltCallInCallScreen] onJoinRoom');
        // If I am not the person who initiated the call, then set a timer to hang up according to the configuration settings
        if (userID !== inviter) {
          if (typeof config.requireInviterConfig == 'object' && config.requireInviterConfig.enabled && config.requireInviterConfig.detectSeconds > 0) {
            startAutoHangUpTimer(config.requireInviterConfig.detectSeconds);
          }
        }
      },
      onCallEnd: (callID, reason, duration) => {
        _onCallEnd(callID, reason, duration);
      },
      avatarBuilder: avatarBuilder,
      onHangUpConfirmation: typeof config.onHangUpConfirmation === 'function' ? () => {
        return config.onHangUpConfirmation();
      } : undefined,
      onUserJoin: userInfoList => {
        const foundInviterJoin = userInfoList.some((userInfo, _) => {
          if (inviter === userInfo.userID) {
            return true;
          }
          return false;
        });
        if (foundInviterJoin) {
          (0, _logger.zloginfo)('[ZegoUIKitPrebuiltCallInCallScreen] found inviter joined room, will clearAutoHangUpTimer');
          clearAutoHangUpTimer();
        }
      }
    }
  });
}
