// @zegocloud/zego-uikit-prebuilt-call-rn/lib/commonjs/call/services/minimizing_helper.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _zegoUikitRn = _interopRequireDefault(require("@zegocloud/zego-uikit-rn"));
var _logger = require("../../utils/logger");
var _MinimizingHelper;
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
class MinimizingHelper {
  constructor() {
    _defineProperty(this, "_isMinimize", false);
    _defineProperty(this, "_isMinimizeSwitch", false);
    _defineProperty(this, "_activeUserID", '');
    _defineProperty(this, "_onActiveUserIDUpdateCallbackMap", {});
    _defineProperty(this, "_onWindowMinimizeCallbackMap", {});
    _defineProperty(this, "_onWindowMaximizeCallbackMap", {});
    _defineProperty(this, "_onEntryNormalCallbackMap", {});
    _defineProperty(this, "_updateTimer", null);
    _defineProperty(this, "_appInfo", {});
    _defineProperty(this, "_localUser", {});
    _defineProperty(this, "_roomID", '');
    _defineProperty(this, "_config", {});
    _defineProperty(this, "_plugins", []);
    _defineProperty(this, "_onPrebuiltInitCallbackMap", {});
  }
  static getInstance() {
    return this._instance || (this._instance = new MinimizingHelper());
  }
  getIsMinimize() {
    return this._isMinimize;
  }
  setIsMinimizeSwitch(isMinimizeSwitch, from) {
    this._isMinimizeSwitch = !!isMinimizeSwitch;
    (0, _logger.zloginfo)(`[MinimizingHelper] setIsMinimizeSwitch: ${this._isMinimizeSwitch}, from: ${from}`);
  }
  getIsMinimizeSwitch() {
    return this._isMinimizeSwitch;
  }
  setInitParams(appID, appSign, userID, userName, roomID, config = {}) {
    this._appInfo = {
      appID,
      appSign
    };
    this._localUser = {
      userID,
      userName
    };
    this._roomID = roomID;
    Object.assign(this._config, config);
  }
  getInitAppInfo() {
    return this._appInfo;
  }
  getInitUser() {
    return this._localUser;
  }
  getInitRoomID() {
    return this._roomID;
  }
  getInitConfig() {
    return this._config;
  }
  getInitPlugins() {
    return this._plugins;
  }
  minimizeWindow() {
    (0, _logger.zloginfo)('[MinimizingHelper][minimizeWindow]');
    const callbackID = 'MinimizingHelper' + String(Math.floor(Math.random() * 10000));
    this.unRegisterAudioVideoListCallback(callbackID);
    this.registerAudioVideoListCallback(callbackID);
    this.notifyMinimize();
    this.startUpdateTimer();
  }
  startUpdateTimer() {
    this.updateActiveUserIDByTimer();
    this.initUpdateTimer();
  }
  initUpdateTimer() {
    clearInterval(this._updateTimer);
    this._updateTimer = null;
  }
  updateActiveUserIDByTimer() {
    const allUsers = _zegoUikitRn.default.getAllUsers().filter(user => user.userID && user.userID !== this._localUser.userID && (user.isCameraOn || user.isMicrophoneOn));
    const activeUserID = allUsers.length ? allUsers[0].userID : this._localUser.userID;
    this._activeUserID = activeUserID || '';

    // console.log('[MinimizingHelper]updateActiveUserIDByTimer', this._activeUserID);
    this.notifyActiveUserIDUpdate(this._activeUserID);
  }
  registerAudioVideoListCallback(callbackID) {
    _zegoUikitRn.default.onUserLeave(callbackID, userInfoList => {
      const temp = userInfoList.filter(user => user.userID === this._activeUserID);
      if (temp.length) {
        this.updateActiveUserIDByTimer();
      }
    });
  }
  unRegisterAudioVideoListCallback(callbackID) {
    _zegoUikitRn.default.onUserLeave(callbackID);
  }
  notifyPrebuiltInit() {
    Object.keys(this._onPrebuiltInitCallbackMap).forEach(callbackID => {
      if (this._onPrebuiltInitCallbackMap[callbackID]) {
        this._onPrebuiltInitCallbackMap[callbackID]();
      }
    });
  }
  notifyActiveUserIDUpdate(activeUserID) {
    Object.keys(this._onActiveUserIDUpdateCallbackMap).forEach(callbackID => {
      if (this._onActiveUserIDUpdateCallbackMap[callbackID]) {
        this._onActiveUserIDUpdateCallbackMap[callbackID](activeUserID);
      }
    });
  }
  notifyMinimize() {
    (0, _logger.zloginfo)('[MinimizingHelper][notifyMinimize]');
    this._isMinimize = true;
    Object.keys(this._onWindowMinimizeCallbackMap).forEach(callbackID => {
      if (this._onWindowMinimizeCallbackMap[callbackID]) {
        this._onWindowMinimizeCallbackMap[callbackID]();
      }
    });
  }
  notifyMaximize() {
    (0, _logger.zloginfo)('[MinimizingHelper][notifyMaximize]');
    this._isMinimize = false;
    Object.keys(this._onWindowMaximizeCallbackMap).forEach(callbackID => {
      if (this._onWindowMaximizeCallbackMap[callbackID]) {
        this._onWindowMaximizeCallbackMap[callbackID]();
      }
    });
  }
  notifyEntryNormal() {
    (0, _logger.zloginfo)('[MinimizingHelper][notifyEntryNormal]');
    this._isMinimize = false;
    Object.keys(this._onEntryNormalCallbackMap).forEach(callbackID => {
      if (this._onEntryNormalCallbackMap[callbackID]) {
        this._onEntryNormalCallbackMap[callbackID]();
      }
    });
  }
  onPrebuiltInit(callbackID, callback) {
    if (typeof callback !== 'function') {
      delete this._onPrebuiltInitCallbackMap[callbackID];
    } else {
      this._onPrebuiltInitCallbackMap[callbackID] = callback;
    }
  }
  onActiveUserIDUpdate(callbackID, callback) {
    if (typeof callback !== 'function') {
      delete this._onActiveUserIDUpdateCallbackMap[callbackID];
    } else {
      this._onActiveUserIDUpdateCallbackMap[callbackID] = callback;
    }
  }
  onWindowMinimized(callbackID, callback) {
    if (typeof callback !== 'function') {
      delete this._onWindowMinimizeCallbackMap[callbackID];
    } else {
      this._onWindowMinimizeCallbackMap[callbackID] = callback;
    }
  }
  onWindowMaximized(callbackID, callback) {
    if (typeof callback !== 'function') {
      delete this._onWindowMaximizeCallbackMap[callbackID];
    } else {
      this._onWindowMaximizeCallbackMap[callbackID] = callback;
    }
  }
  onEntryNormal(callbackID, callback) {
    if (typeof callback !== 'function') {
      delete this._onEntryNormalCallbackMap[callbackID];
    } else {
      this._onEntryNormalCallbackMap[callbackID] = callback;
    }
  }
}
exports.default = MinimizingHelper;
_MinimizingHelper = MinimizingHelper;
_defineProperty(MinimizingHelper, "_instance", void 0);
