// @zegocloud/zego-uikit-prebuilt-call-rn/lib/commonjs/call/services/prebuilt_helper.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _logger = require("../../utils/logger");
var _PrebuiltHelper;
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
class PrebuiltHelper {
  constructor() {
    _defineProperty(this, "_stateData", {});
    _defineProperty(this, "_routeParams", {});
    _defineProperty(this, "_notifyData", {});
    _defineProperty(this, "_onPrebuiltDestroyCallbackMap", {});
  }
  static getInstance() {
    return this._instance || (this._instance = new PrebuiltHelper());
  }
  // Use reference types directly, so the set method is not provided here
  getStateData() {
    return this._stateData;
  }
  getRouteParams() {
    return this._routeParams;
  }
  getNotifyData() {
    return this._notifyData;
  }
  clearState() {
    this._stateData = {
      memberConnectStateMap: {}
    };
  }
  clearRouteParams() {
    (0, _logger.zloginfo)(`[PrebuiltHelper][clearRouteParams]`);
    this._routeParams = {};
  }
  clearNotify() {
    this._notifyData = {};
  }
  notifyDestroyPrebuilt() {
    (0, _logger.zloginfo)('[PrebuiltHelper][notifyDestroyPrebuilt]', this._onPrebuiltDestroyCallbackMap);
    Object.keys(this._onPrebuiltDestroyCallbackMap).forEach(callbackID => {
      if (this._onPrebuiltDestroyCallbackMap[callbackID]) {
        this._onPrebuiltDestroyCallbackMap[callbackID]();
      }
    });
  }
  // Temporarily resolved an issue where dialog shutdown could not be triggered
  onPrebuiltDestroy(callbackID, callback) {
    if (typeof callback !== 'function') {
      delete this._onPrebuiltDestroyCallbackMap[callbackID];
    } else {
      this._onPrebuiltDestroyCallbackMap[callbackID] = callback;
    }
  }
}
exports.default = PrebuiltHelper;
_PrebuiltHelper = PrebuiltHelper;
_defineProperty(PrebuiltHelper, "_instance", void 0);
