// @zegocloud/zego-uikit-rn/lib/module/signal_plugin/services/in_room_message_service.js
var _ZegoPluginInRoomMessageService;
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
import ZegoPluginRoomMessageCore from '../core/in_room_message_core';
export default class ZegoPluginInRoomMessageService {
  constructor() {
    if (!ZegoPluginInRoomMessageService.shared) {
      ZegoPluginInRoomMessageService.shared = this;
    }
    return ZegoPluginInRoomMessageService.shared;
  }
  static getInstance() {
    if (!ZegoPluginInRoomMessageService.shared) {
      ZegoPluginInRoomMessageService.shared = new ZegoPluginInRoomMessageService();
    }
    return ZegoPluginInRoomMessageService.shared;
  }
  sendInRoomTextMessage(roomID, message) {
    return ZegoPluginRoomMessageCore.getInstance().sendInRoomTextMessage(roomID, message);
  }
  sendInRoomCommandMessage(roomID, message) {
    return ZegoPluginRoomMessageCore.getInstance().sendInRoomCommandMessage(roomID, message);
  }
  onInRoomTextMessageReceived(callbackID, callback) {
    ZegoPluginRoomMessageCore.getInstance().onInRoomTextMessageReceived(callbackID, callback);
  }
  onInRoomCommandMessageReceived(callbackID, callback) {
    ZegoPluginRoomMessageCore.getInstance().onInRoomCommandMessageReceived(callbackID, callback);
  }
}
_ZegoPluginInRoomMessageService = ZegoPluginInRoomMessageService;
_defineProperty(ZegoPluginInRoomMessageService, "shared", void 0);
