// @zegocloud/zego-uikit-rn/lib/module/signal_plugin/services/user_in_room_attributes_service.js
var _ZegoPluginUserInRoomAttributesService;
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
import ZegoPluginUserInRoomAttributesCore from '../core/user_in_room_attributes_core';
export default class ZegoPluginUserInRoomAttributesService {
  constructor() {
    if (!ZegoPluginUserInRoomAttributesService.shared) {
      ZegoPluginUserInRoomAttributesService.shared = this;
    }
    return ZegoPluginUserInRoomAttributesService.shared;
  }
  static getInstance() {
    if (!ZegoPluginUserInRoomAttributesService.shared) {
      ZegoPluginUserInRoomAttributesService.shared = new ZegoPluginUserInRoomAttributesService();
    }
    return ZegoPluginUserInRoomAttributesService.shared;
  }
  joinRoom(roomID) {
    return ZegoPluginUserInRoomAttributesCore.getInstance().joinRoom(roomID);
  }
  leaveRoom() {
    return ZegoPluginUserInRoomAttributesCore.getInstance().leaveRoom();
  }
  leaveAllRoom() {
    return ZegoPluginUserInRoomAttributesCore.getInstance().leaveAllRoom();
  }
  setUsersInRoomAttributes(key, value, userIDs) {
    const attributes = {
      [key]: value
    };
    return ZegoPluginUserInRoomAttributesCore.getInstance().setUsersInRoomAttributes(attributes, userIDs);
  }
  queryUsersInRoomAttributes(nextFlag, count) {
    const config = {
      nextFlag,
      count
    };
    return ZegoPluginUserInRoomAttributesCore.getInstance().queryUsersInRoomAttributes(config);
  }
  onUsersInRoomAttributesUpdated(callbackID, callback) {
    ZegoPluginUserInRoomAttributesCore.getInstance().onUsersInRoomAttributesUpdated(callbackID, callback);
  }
}
_ZegoPluginUserInRoomAttributesService = ZegoPluginUserInRoomAttributesService;
_defineProperty(ZegoPluginUserInRoomAttributesService, "shared", void 0);
