// @zegocloud/zego-uikit-prebuilt-call-rn/lib/module/utils/EventEmitter.js
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
export let EventName = /*#__PURE__*/function (EventName) {
  EventName[EventName["LOGINED"] = 0] = "LOGINED";
  EventName[EventName["SHOW_TOAST"] = 1] = "SHOW_TOAST";
  return EventName;
}({});
class EventEmitter {
  constructor() {
    _defineProperty(this, "callbacks", {});
  }
  on(event, fn) {
    if (!this.callbacks[event]) {
      this.callbacks[event] = [];
    }
    this.callbacks[event] = [...this.callbacks[event], fn];
    return this;
  }
  emit(event, ...args) {
    const callbacks = this.callbacks[event];
    if (callbacks) {
      callbacks.forEach(callback => callback.apply(this, args));
    }
    return this;
  }
  off(event, fn) {
    const callbacks = this.callbacks[event];
    if (callbacks) {
      if (fn) {
        this.callbacks[event] = callbacks.filter(callback => callback !== fn);
      } else {
        delete this.callbacks[event];
      }
    }
    return this;
  }
  destroy() {
    this.callbacks = {};
  }
}
export const eventEmitter = new EventEmitter();
