// @zegocloud/zego-uikit-rn/lib/commonjs/components/internal/defines.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ZegoUIKitVideoConfig = exports.ZegoUIKitPluginType = exports.ZegoToastType = exports.ZegoRoomPropertyUpdateType = exports.ZegoChangedCountOrProperty = exports.ZegoAudioVideoResourceMode = void 0;
var _zegoExpressEngineReactnative = require("zego-express-engine-reactnative");
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
const ZegoAudioVideoResourceMode = exports.ZegoAudioVideoResourceMode = {
  Default: 0,
  CDNOnly: 1,
  L3Only: 2,
  RTCOnly: 3,
  CDNPlus: 4
};
const ZegoChangedCountOrProperty = exports.ZegoChangedCountOrProperty = {
  userAdd: 1,
  userDelete: 2,
  microphoneStateUpdate: 3,
  cameraStateUpdate: 4,
  attributesUpdate: 5
};
const ZegoUIKitPluginType = exports.ZegoUIKitPluginType = {
  signaling: 1,
  // zim, fcm
  beauty: 2,
  // effects or avatar or deepAR
  whiteboard: 3 // superboard
};
const ZegoRoomPropertyUpdateType = exports.ZegoRoomPropertyUpdateType = {
  set: 0,
  update: 1,
  remote: 2
};
class ZegoUIKitVideoConfig {
  constructor(fps, bitrate, width, height) {
    /// Frame rate, control the frame rate of the camera and the frame rate of the encoder.
    _defineProperty(this, "fps", void 0);
    /// Bit rate in kbps.
    _defineProperty(this, "bitrate", void 0);
    /// resolution width, control the image width of camera image acquisition or encoder when publishing stream.
    _defineProperty(this, "width", void 0);
    /// resolution height, control the image height of camera image acquisition or encoder when publishing stream.
    _defineProperty(this, "height", void 0);
    this.fps = fps;
    this.bitrate = bitrate;
    this.width = width;
    this.height = height;
  }
  static preset180P() {
    return new ZegoUIKitVideoConfig(15, 300, 180, 320);
  }
  static preset270P() {
    return new ZegoUIKitVideoConfig(15, 400, 270, 480);
  }
  static preset360P() {
    return new ZegoUIKitVideoConfig(15, 600, 360, 640);
    ;
  }
  static preset540P() {
    return new ZegoUIKitVideoConfig(15, 1200, 540, 960);
  }
  static preset720P() {
    return new ZegoUIKitVideoConfig(15, 1500, 720, 1280);
  }
  static preset1080P() {
    return new ZegoUIKitVideoConfig(15, 3000, 1080, 1920);
  }
  static preset2K() {
    return new ZegoUIKitVideoConfig(15, 6000, 1440, 2560);
  }
  static preset4K() {
    return new ZegoUIKitVideoConfig(15, 12000, 2160, 3840);
  }
  toSDK(orientation) {
    var config = new _zegoExpressEngineReactnative.ZegoVideoConfig();
    config.bitrate = this.bitrate;
    config.fps = this.fps;
    const isPortraitUp = orientation === _zegoExpressEngineReactnative.ZegoOrientation.PortraitUp || orientation === _zegoExpressEngineReactnative.ZegoOrientation.PortraitDown;
    const width = isPortraitUp ? this.width : this.height;
    const height = isPortraitUp ? this.height : this.width;
    config.encodeWidth = width;
    config.captureWidth = width;
    config.encodeHeight = height;
    config.captureHeight = height;
    return config;
  }
}
exports.ZegoUIKitVideoConfig = ZegoUIKitVideoConfig;
const ZegoToastType = exports.ZegoToastType = {
  default: 0,
  info: 1,
  success: 2,
  warning: 3,
  error: 4
};
