// zego-zim-react-native/lib/ZIMErrorCode.js
export var ZIMErrorCode;
(function (ZIMErrorCode) {
    /** Success without exception. */
    ZIMErrorCode[ZIMErrorCode["Success"] = 0] = "Success";
    /** Failed, the guarantee is wrong. */
    ZIMErrorCode[ZIMErrorCode["Failed"] = 1] = "Failed";
    /** The incoming parameter is invalid. <br>Use case: Used to protect the bottom strategy when the interface call fails due to incorrect parameters. */
    ZIMErrorCode[ZIMErrorCode["CommonModuleParamsInvalid"] = 6000001] = "CommonModuleParamsInvalid";
    /** The SDK is not initialized. <br>Use case: Used for error return when the ZIM SDK is not initialized successfully. */
    ZIMErrorCode[ZIMErrorCode["CommonModuleNotInit"] = 6000002] = "CommonModuleNotInit";
    /** Invalid AppID. <br>Use case: Used for error return of failure due to invalid AppID. */
    ZIMErrorCode[ZIMErrorCode["CommonModuleInvalidAppID"] = 6000003] = "CommonModuleInvalidAppID";
    /** Trigger SDK internal frequency limit. */
    ZIMErrorCode[ZIMErrorCode["CommonModuleTriggerSDKFrequencyLimit"] = 6000004] = "CommonModuleTriggerSDKFrequencyLimit";
    /** Trigger background service frequency limit. */
    ZIMErrorCode[ZIMErrorCode["CommonModuleTriggerServerFrequencyLimit"] = 6000005] = "CommonModuleTriggerServerFrequencyLimit";
    /** The Switch server reported an error. */
    ZIMErrorCode[ZIMErrorCode["CommonModuleSwitchServerError"] = 6000006] = "CommonModuleSwitchServerError";
    /** ZIM service internal error. */
    ZIMErrorCode[ZIMErrorCode["CommonModuleIMServerError"] = 6000007] = "CommonModuleIMServerError";
    /** ZIM internal database error. */
    ZIMErrorCode[ZIMErrorCode["CommonModuleIMDatabaseError"] = 6000008] = "CommonModuleIMDatabaseError";
    /** Disconnected while sending request. */
    ZIMErrorCode[ZIMErrorCode["CommonModuleIMServerDisconnect"] = 6000009] = "CommonModuleIMServerDisconnect";
    /** Failed to upload log. */
    ZIMErrorCode[ZIMErrorCode["CommonModuleUploadLogError"] = 6000010] = "CommonModuleUploadLogError";
    /** Illegal param. */
    ZIMErrorCode[ZIMErrorCode["CommonModuleUserIsNotExist"] = 6000011] = "CommonModuleUserIsNotExist";
    /** User background query frequency limit. */
    ZIMErrorCode[ZIMErrorCode["CommonModuleUserInfoQueriedLimit"] = 6000012] = "CommonModuleUserInfoQueriedLimit";
    /** User background query frequency limit. The current package does not support this operation. */
    ZIMErrorCode[ZIMErrorCode["CommonModuleUnsupportedRequest"] = 6000013] = "CommonModuleUnsupportedRequest";
    /**  */
    ZIMErrorCode[ZIMErrorCode["CommonModuleUnactivatedIMServer"] = 6000014] = "CommonModuleUnactivatedIMServer";
    /** The upper limit of daily active users of the AppID was exceeded. Procedure. */
    ZIMErrorCode[ZIMErrorCode["CommonModuleExceedDAULimit"] = 6000015] = "CommonModuleExceedDAULimit";
    /** The upper limit of the monthly active users of the AppID is exceeded. Procedure. */
    ZIMErrorCode[ZIMErrorCode["CommonModuleExceedMAULimit"] = 6000016] = "CommonModuleExceedMAULimit";
    /** Login failed due to internal reasons. */
    ZIMErrorCode[ZIMErrorCode["NetworkModuleCommonError"] = 6000101] = "NetworkModuleCommonError";
    /** Login failed due to internal reasons. */
    ZIMErrorCode[ZIMErrorCode["NetworkModuleServerError"] = 6000102] = "NetworkModuleServerError";
    /** Token is invalid. */
    ZIMErrorCode[ZIMErrorCode["NetworkModuleTokenInvalid"] = 6000103] = "NetworkModuleTokenInvalid";
    /** Network error. */
    ZIMErrorCode[ZIMErrorCode["NetworkModuleNetworkError"] = 6000104] = "NetworkModuleNetworkError";
    /** The request to the background timed out. */
    ZIMErrorCode[ZIMErrorCode["NetworkModuleRequestTimeout"] = 6000105] = "NetworkModuleRequestTimeout";
    /** Token expired. */
    ZIMErrorCode[ZIMErrorCode["NetworkModuleTokenExpired"] = 6000106] = "NetworkModuleTokenExpired";
    /** Token version number is wrong. */
    ZIMErrorCode[ZIMErrorCode["NetworkModuleTokenVersionError"] = 6000107] = "NetworkModuleTokenVersionError";
    /** Token duration is too short. */
    ZIMErrorCode[ZIMErrorCode["NetworkModuleTokenTimeIsTooShort"] = 6000108] = "NetworkModuleTokenTimeIsTooShort";
    /** rsp_proto_parse_error. */
    ZIMErrorCode[ZIMErrorCode["NetworkModuleRspProtoParseError"] = 6000109] = "NetworkModuleRspProtoParseError";
    /** Logging in to multiple accounts on the current device. */
    ZIMErrorCode[ZIMErrorCode["NetworkModuleUserHasAlreadyLogged"] = 6000111] = "NetworkModuleUserHasAlreadyLogged";
    /** User is not logged. */
    ZIMErrorCode[ZIMErrorCode["NetworkModuleUserIsNotLogged"] = 6000121] = "NetworkModuleUserIsNotLogged";
    /** The user has logged out. */
    ZIMErrorCode[ZIMErrorCode["NetworkModuleUserHasAlreadyLoggedOut"] = 6000122] = "NetworkModuleUserHasAlreadyLoggedOut";
    /** The user is offline. Procedure. */
    ZIMErrorCode[ZIMErrorCode["NetworkModuleUserOffline"] = 6000123] = "NetworkModuleUserOffline";
    /** The userID passed in for the offline login does not match the userID from the last successful login. */
    ZIMErrorCode[ZIMErrorCode["NetworkModuleUserIDError"] = 6000124] = "NetworkModuleUserIDError";
    /** Internal error sending message. */
    ZIMErrorCode[ZIMErrorCode["MessageModuleCommonError"] = 6000201] = "MessageModuleCommonError";
    /** Error sending message background service. */
    ZIMErrorCode[ZIMErrorCode["MessageModuleServerError"] = 6000202] = "MessageModuleServerError";
    /** Message sending failed. */
    ZIMErrorCode[ZIMErrorCode["MessageModuleSendMessageFailed"] = 6000203] = "MessageModuleSendMessageFailed";
    /** The message target does not exist. */
    ZIMErrorCode[ZIMErrorCode["MessageModuleTargetDoesNotExist"] = 6000204] = "MessageModuleTargetDoesNotExist";
    /** The message edited failed. */
    ZIMErrorCode[ZIMErrorCode["MessageModuleEditMessageFailed"] = 6000205] = "MessageModuleEditMessageFailed";
    /** Message sending failed because the user actively canceled it. */
    ZIMErrorCode[ZIMErrorCode["MessageModuleSendMessageFailedForCancelled"] = 6000206] = "MessageModuleSendMessageFailedForCancelled";
    /** Failed to cancel the message being sent. */
    ZIMErrorCode[ZIMErrorCode["MessageModuleCancelSendingMessageFailed"] = 6000207] = "MessageModuleCancelSendingMessageFailed";
    /** the message is not pinned. */
    ZIMErrorCode[ZIMErrorCode["MessageModuleMessageIsNotPinned"] = 6000208] = "MessageModuleMessageIsNotPinned";
    /** File generic error. */
    ZIMErrorCode[ZIMErrorCode["MessageModuleFileError"] = 6000210] = "MessageModuleFileError";
    /** There are no errors in the file. */
    ZIMErrorCode[ZIMErrorCode["MessageModuleFileNotExist"] = 6000211] = "MessageModuleFileNotExist";
    /** File operation background error. */
    ZIMErrorCode[ZIMErrorCode["MessageModuleFileServerError"] = 6000212] = "MessageModuleFileServerError";
    /** The file type is not supported. Procedure. */
    ZIMErrorCode[ZIMErrorCode["MessageModuleFileTypeUnsupported"] = 6000213] = "MessageModuleFileTypeUnsupported";
    /** The file size is invalid. */
    ZIMErrorCode[ZIMErrorCode["MessageModuleFileSizeInvalid"] = 6000214] = "MessageModuleFileSizeInvalid";
    /** The duration of the audio and video files is invalid. */
    ZIMErrorCode[ZIMErrorCode["MessageModuleFileDurationInvalid"] = 6000215] = "MessageModuleFileDurationInvalid";
    /** File permission error. */
    ZIMErrorCode[ZIMErrorCode["MessageModuleFilePermissionDenied"] = 6000216] = "MessageModuleFilePermissionDenied";
    /** File download failed. Procedure */
    ZIMErrorCode[ZIMErrorCode["MessageModuleFileDownloadFailed"] = 6000217] = "MessageModuleFileDownloadFailed";
    /** File size download exceeds limit. */
    ZIMErrorCode[ZIMErrorCode["MessageModuleFileDownloadLimit"] = 6000218] = "MessageModuleFileDownloadLimit";
    /** Url resource not found. */
    ZIMErrorCode[ZIMErrorCode["MessageModuleFileDownloadUrlNotFound"] = 6000219] = "MessageModuleFileDownloadUrlNotFound";
    /** A server error occurred downloading the url. */
    ZIMErrorCode[ZIMErrorCode["MessageModuleFileDownloadHttpRequestServerError"] = 6000220] = "MessageModuleFileDownloadHttpRequestServerError";
    /** The message did not pass the review. */
    ZIMErrorCode[ZIMErrorCode["MessageModuleAuditRejected"] = 6000221] = "MessageModuleAuditRejected";
    /** Message auditing failed. */
    ZIMErrorCode[ZIMErrorCode["MessageModuleAuditFailed"] = 6000222] = "MessageModuleAuditFailed";
    /** The message custom pre-callback business audit did not pass. */
    ZIMErrorCode[ZIMErrorCode["MessageModuleAuditCustomSentRejected"] = 6000230] = "MessageModuleAuditCustomSentRejected";
    /** The cache file failed to open. */
    ZIMErrorCode[ZIMErrorCode["MessageModuleCacheFileOpenFailed"] = 6000240] = "MessageModuleCacheFileOpenFailed";
    /** The cache file export or import is in progress. */
    ZIMErrorCode[ZIMErrorCode["MessageModuleCacheFileExportOrImportIsInProgressCurrently"] = 6000241] = "MessageModuleCacheFileExportOrImportIsInProgressCurrently";
    /** The cache file can only import its own cache. */
    ZIMErrorCode[ZIMErrorCode["MessageModuleCacheFileOnlyCanImportCacheOfOneself"] = 6000242] = "MessageModuleCacheFileOnlyCanImportCacheOfOneself";
    /** The cache file export failed. */
    ZIMErrorCode[ZIMErrorCode["MessageModuleCacheFileExportFailed"] = 6000243] = "MessageModuleCacheFileExportFailed";
    /** The cache file import failed. */
    ZIMErrorCode[ZIMErrorCode["MessageModuleCacheFileImportFailed"] = 6000244] = "MessageModuleCacheFileImportFailed";
    /** The cache file JSON parsing failed. */
    ZIMErrorCode[ZIMErrorCode["MessageModuleCacheFileJsonParseFailed"] = 6000245] = "MessageModuleCacheFileJsonParseFailed";
    /** File request short connection error. */
    ZIMErrorCode[ZIMErrorCode["MessageModuleFileConnectError"] = 6000251] = "MessageModuleFileConnectError";
    /** File request encountered an HTTP error. */
    ZIMErrorCode[ZIMErrorCode["MessageModuleFileHttpError"] = 6000252] = "MessageModuleFileHttpError";
    /** File request encountered a certificate error. */
    ZIMErrorCode[ZIMErrorCode["MessageModuleFileVerificationError"] = 6000253] = "MessageModuleFileVerificationError";
    /** Client file request error. */
    ZIMErrorCode[ZIMErrorCode["MessageModuleFileClientError"] = 6000254] = "MessageModuleFileClientError";
    /** Call invitation error. */
    ZIMErrorCode[ZIMErrorCode["CallModuleCommonError"] = 6000270] = "CallModuleCommonError";
    /** Call cancellation error. */
    ZIMErrorCode[ZIMErrorCode["CallModuleCancelError"] = 6000271] = "CallModuleCancelError";
    /** Call invitation background error. */
    ZIMErrorCode[ZIMErrorCode["CallModuleServerError"] = 6000272] = "CallModuleServerError";
    /** User uninviter operation error. */
    ZIMErrorCode[ZIMErrorCode["CallModuleIsNotInvitor"] = 6000273] = "CallModuleIsNotInvitor";
    /** User non-invitee operation error. */
    ZIMErrorCode[ZIMErrorCode["CallModuleIsNotInvitee"] = 6000274] = "CallModuleIsNotInvitee";
    /** The call already has an error. */
    ZIMErrorCode[ZIMErrorCode["CallModuleAlreadyExists"] = 6000275] = "CallModuleAlreadyExists";
    /** There is no error in the call. */
    ZIMErrorCode[ZIMErrorCode["CallModuleDoesNotExist"] = 6000276] = "CallModuleDoesNotExist";
    /** Set the receipt message to read error. */
    ZIMErrorCode[ZIMErrorCode["MessageModuleReceiptReadError"] = 6000277] = "MessageModuleReceiptReadError";
    /** I want to withdraw the message but the time limit has expired. */
    ZIMErrorCode[ZIMErrorCode["MessageModuleMessageExceedsRevokeTime"] = 6000278] = "MessageModuleMessageExceedsRevokeTime";
    /** Tried to retract the message but it's been retracted. */
    ZIMErrorCode[ZIMErrorCode["MessageModuleMessageHasBeenRevoked"] = 6000279] = "MessageModuleMessageHasBeenRevoked";
    /** When a user sets or deletes a key in a message, the key has been set or deleted by the user. <br> Processing: Developers can pop-up prompt, or no special processing. */
    ZIMErrorCode[ZIMErrorCode["MessageModuleMessageReactionTypeExisted"] = 6000280] = "MessageModuleMessageReactionTypeExisted";
    /** When you initiate a call invitation, all invitees are not registered. <br> Solution: 1. Please check the registration status of the invitee. If the user to be invited has not yet registered, please register first. 2. If the user is registered, check whether the UserID in the invitee list is correct. */
    ZIMErrorCode[ZIMErrorCode["CallModuleInviteUserDoesNotExist"] = 6000281] = "CallModuleInviteUserDoesNotExist";
    /** Call the sendMessageReceiptsRead interface to set more than 10 messages to read at a time. Please reduce the number of incoming messages to less than 10. */
    ZIMErrorCode[ZIMErrorCode["MessageModuleMessageReceiptLimit"] = 6000282] = "MessageModuleMessageReceiptLimit";
    /** Note: The device that invokes the callJoin interface has called in advanced mode and is the primary device. No action is required. */
    ZIMErrorCode[ZIMErrorCode["CallModuleUserAlreadyInTheCall"] = 6000283] = "CallModuleUserAlreadyInTheCall";
    /** The sender has been added to the blacklist. */
    ZIMErrorCode[ZIMErrorCode["MessageModuleSenderInBlacklist"] = 6000284] = "MessageModuleSenderInBlacklist";
    /** No corresponding operation authority. */
    ZIMErrorCode[ZIMErrorCode["MessageModuleEditNoCorrespondingOperationAuthority"] = 6000285] = "MessageModuleEditNoCorrespondingOperationAuthority";
    /** The message has exceeded the edit time limit. */
    ZIMErrorCode[ZIMErrorCode["MessageModuleMessageExceedsEditTime"] = 6000286] = "MessageModuleMessageExceedsEditTime";
    /** The number of pinned messages exceeds limit. */
    ZIMErrorCode[ZIMErrorCode["MessageModuleTheNumberOfPinnedMessagesExceedsLimit"] = 6000287] = "MessageModuleTheNumberOfPinnedMessagesExceedsLimit";
    /** Room related operation error. */
    ZIMErrorCode[ZIMErrorCode["RoomModuleCommonError"] = 6000301] = "RoomModuleCommonError";
    /** Room operation background failed. */
    ZIMErrorCode[ZIMErrorCode["RoomModuleServerError"] = 6000302] = "RoomModuleServerError";
    /** Room creation failed. */
    ZIMErrorCode[ZIMErrorCode["RoomModuleCreateRoomError"] = 6000303] = "RoomModuleCreateRoomError";
    /** Room entry failed. */
    ZIMErrorCode[ZIMErrorCode["RoomModuleJoinRoomError"] = 6000304] = "RoomModuleJoinRoomError";
    /** Failed to leave the room. */
    ZIMErrorCode[ZIMErrorCode["RoomModuleLeaveRoomError"] = 6000306] = "RoomModuleLeaveRoomError";
    /** Failed to query a room member. */
    ZIMErrorCode[ZIMErrorCode["RoomModuleTheRoomMemberQueryFailed"] = 6000310] = "RoomModuleTheRoomMemberQueryFailed";
    /**  */
    ZIMErrorCode[ZIMErrorCode["RoomModuleTheRoomMembersQueryFailedCompletely"] = 6000311] = "RoomModuleTheRoomMembersQueryFailedCompletely";
    /** User has error in room. */
    ZIMErrorCode[ZIMErrorCode["RoomModuleUserIsAlreadyInTheRoom"] = 6000320] = "RoomModuleUserIsAlreadyInTheRoom";
    /** User not in room error. */
    ZIMErrorCode[ZIMErrorCode["RoomModuleUserIsNotInTheRoom"] = 6000321] = "RoomModuleUserIsNotInTheRoom";
    /**  */
    ZIMErrorCode[ZIMErrorCode["RoomModuleTheRoomDoesNotExist"] = 6000322] = "RoomModuleTheRoomDoesNotExist";
    /** There are no errors in the room. */
    ZIMErrorCode[ZIMErrorCode["RoomModuleTheRoomAlreadyExists"] = 6000323] = "RoomModuleTheRoomAlreadyExists";
    /** The number of existing rooms has reached its limit. */
    ZIMErrorCode[ZIMErrorCode["RoomModuleTheNumberOfExistingRoomsHasReachedLimit"] = 6000324] = "RoomModuleTheNumberOfExistingRoomsHasReachedLimit";
    /** The number of added rooms reached the upper limit. */
    ZIMErrorCode[ZIMErrorCode["RoomModuleTheNumberOfJoinedRoomsHasReachedLimit"] = 6000325] = "RoomModuleTheNumberOfJoinedRoomsHasReachedLimit";
    /** Try an interface such as createRoom, joinRoom, or enterRoom repeatedly, or perform an action on a room that is being connected. Please wait for return of room connection result returned by onRoomStateChanged before performing operations. */
    ZIMErrorCode[ZIMErrorCode["RoomModuleTheRoomIsConnecting"] = 6000326] = "RoomModuleTheRoomIsConnecting";
    /**  */
    ZIMErrorCode[ZIMErrorCode["RoomModuleRoomMemberIsBlocked"] = 6000327] = "RoomModuleRoomMemberIsBlocked";
    /** Room property operation failed. */
    ZIMErrorCode[ZIMErrorCode["RoomModuleRoomAttributesCommonError"] = 6000330] = "RoomModuleRoomAttributesCommonError";
    /** The room properties operation failed completely. */
    ZIMErrorCode[ZIMErrorCode["RoomModuleRoomAttributesOperationFailedCompletely"] = 6000331] = "RoomModuleRoomAttributesOperationFailedCompletely";
    /** The room properties operation part failed. */
    ZIMErrorCode[ZIMErrorCode["RoomModuleRoomAttributesOperationFailedPartly"] = 6000332] = "RoomModuleRoomAttributesOperationFailedPartly";
    /** Description Failed to query the room properties. */
    ZIMErrorCode[ZIMErrorCode["RoomModuleRoomAttributesQueryFailed"] = 6000333] = "RoomModuleRoomAttributesQueryFailed";
    /** The number of room properties reached the upper limit. */
    ZIMErrorCode[ZIMErrorCode["RoomModuleTheNumberOfRoomAttributesExceedsLimit"] = 6000334] = "RoomModuleTheNumberOfRoomAttributesExceedsLimit";
    /** The length of the room attribute Key reaches the upper limit. Procedure */
    ZIMErrorCode[ZIMErrorCode["RoomModuleTheLengthOfRoomAttributeKeyExceedsLimit"] = 6000335] = "RoomModuleTheLengthOfRoomAttributeKeyExceedsLimit";
    /** The length of the room attribute Value reached the upper limit. Procedure */
    ZIMErrorCode[ZIMErrorCode["RoomModuleTheLengthOfRoomAttributeValueExceedsLimit"] = 6000336] = "RoomModuleTheLengthOfRoomAttributeValueExceedsLimit";
    /** Total length of room attribute reached the upper limit. */
    ZIMErrorCode[ZIMErrorCode["RoomModuleTheTotalLengthOfRoomAttributesValueExceedsLimit"] = 6000337] = "RoomModuleTheTotalLengthOfRoomAttributesValueExceedsLimit";
    /** Room room member attribute failed. */
    ZIMErrorCode[ZIMErrorCode["RoomModuleRoomMemberAttributesCommonError"] = 6000350] = "RoomModuleRoomMemberAttributesCommonError";
    /** The total length of the room member attribute reaches the upper limit. */
    ZIMErrorCode[ZIMErrorCode["RoomModuleTheTotalLengthOfRoomMemberAttributesExceedsLimit"] = 6000351] = "RoomModuleTheTotalLengthOfRoomMemberAttributesExceedsLimit";
    /** The length of the room user attribute Key reached the upper limit. */
    ZIMErrorCode[ZIMErrorCode["RoomModuleTheLengthOfRoomMemberAttributesKeyExceedsLimit"] = 6000352] = "RoomModuleTheLengthOfRoomMemberAttributesKeyExceedsLimit";
    /** The length of the room member attribute Value reached the upper limit. Procedure. */
    ZIMErrorCode[ZIMErrorCode["RoomModuleTheLengthOfRoomMemberAttributesValueExceedsLimit"] = 6000353] = "RoomModuleTheLengthOfRoomMemberAttributesValueExceedsLimit";
    /** The number of room member attributes reached the upper limit. */
    ZIMErrorCode[ZIMErrorCode["RoomModuleTheMemberNumberOfRoomMemberAttributesExceedsLimit"] = 6000357] = "RoomModuleTheMemberNumberOfRoomMemberAttributesExceedsLimit";
    /** The push ID is invalid. Procedure. */
    ZIMErrorCode[ZIMErrorCode["ZpnsModulePushIDInvalid"] = 6000401] = "ZpnsModulePushIDInvalid";
    /** An error occurred during group-related operations. */
    ZIMErrorCode[ZIMErrorCode["GroupModuleCommonError"] = 6000501] = "GroupModuleCommonError";
    /** The group related background reported an error. */
    ZIMErrorCode[ZIMErrorCode["GroupModuleServerError"] = 6000502] = "GroupModuleServerError";
    /** Failed to create a group. Procedure. */
    ZIMErrorCode[ZIMErrorCode["GroupModuleCreateGroupError"] = 6000503] = "GroupModuleCreateGroupError";
    /** Failed to dissolve the group. */
    ZIMErrorCode[ZIMErrorCode["GroupModuleDismissGroupError"] = 6000504] = "GroupModuleDismissGroupError";
    /** Failed to join the group. */
    ZIMErrorCode[ZIMErrorCode["GroupModuleJoinGroupError"] = 6000505] = "GroupModuleJoinGroupError";
    /** Failed to leave the group. */
    ZIMErrorCode[ZIMErrorCode["GroupModuleLeaveGroupError"] = 6000506] = "GroupModuleLeaveGroupError";
    /** Failed to kick out a group member. */
    ZIMErrorCode[ZIMErrorCode["GroupModuleKickoutGroupMemberError"] = 6000507] = "GroupModuleKickoutGroupMemberError";
    /** Failed to invite users to join the group. */
    ZIMErrorCode[ZIMErrorCode["GroupModuleInviteUserIntoGroupError"] = 6000508] = "GroupModuleInviteUserIntoGroupError";
    /** Failed to transfer the group master. */
    ZIMErrorCode[ZIMErrorCode["GroupModuleTransferOwnerError"] = 6000509] = "GroupModuleTransferOwnerError";
    /** Failed to update group information.  */
    ZIMErrorCode[ZIMErrorCode["GroupModuleUpdateGroupInfoError"] = 6000510] = "GroupModuleUpdateGroupInfoError";
    /** Failed to query group information.  */
    ZIMErrorCode[ZIMErrorCode["GroupModuleQueryGroupInfoError"] = 6000511] = "GroupModuleQueryGroupInfoError";
    /** Failed to manipulate the group attribute. */
    ZIMErrorCode[ZIMErrorCode["GroupModuleGroupAttributesOperationFailed"] = 6000512] = "GroupModuleGroupAttributesOperationFailed";
    /** Failed to query group attributes. Procedure */
    ZIMErrorCode[ZIMErrorCode["GroupModuleGroupAttributesQueryFailed"] = 6000513] = "GroupModuleGroupAttributesQueryFailed";
    /** Description Failed to update group member information. */
    ZIMErrorCode[ZIMErrorCode["GroupModuleUpdateGroupMemberInfoError"] = 6000514] = "GroupModuleUpdateGroupMemberInfoError";
    /** Description Failed to query group member information. */
    ZIMErrorCode[ZIMErrorCode["GroupModuleQueryGroupMemberInfoError"] = 6000515] = "GroupModuleQueryGroupMemberInfoError";
    /** Failed to query the group list. Procedure */
    ZIMErrorCode[ZIMErrorCode["GroupModuleQueryGroupListError"] = 6000516] = "GroupModuleQueryGroupListError";
    /** Description Failed to query the group member list. */
    ZIMErrorCode[ZIMErrorCode["GroupModuleQueryGroupMemberListError"] = 6000517] = "GroupModuleQueryGroupMemberListError";
    /** The user is not in the group. */
    ZIMErrorCode[ZIMErrorCode["GroupModuleUserIsNotInTheGroup"] = 6000521] = "GroupModuleUserIsNotInTheGroup";
    /** The user is already in the group. */
    ZIMErrorCode[ZIMErrorCode["GroupModuleMemberIsAlreadyInTheGroup"] = 6000522] = "GroupModuleMemberIsAlreadyInTheGroup";
    /** The group does not exist. */
    ZIMErrorCode[ZIMErrorCode["GroupModuleGroupDoesNotExist"] = 6000523] = "GroupModuleGroupDoesNotExist";
    /** The group already exists. */
    ZIMErrorCode[ZIMErrorCode["GroupModuleGroupAlreadyExists"] = 6000524] = "GroupModuleGroupAlreadyExists";
    /** The group membership reaches the upper limit. */
    ZIMErrorCode[ZIMErrorCode["GroupModuleGroupMemberHasReachedLimit"] = 6000525] = "GroupModuleGroupMemberHasReachedLimit";
    /** The group property does not exist. */
    ZIMErrorCode[ZIMErrorCode["GroupModuleGroupAttributeDoesNotExist"] = 6000526] = "GroupModuleGroupAttributeDoesNotExist";
    /** Note: Create a group with a destroyed group ID. Use a different ID. */
    ZIMErrorCode[ZIMErrorCode["GroupModuleCreateGroupWithDismissed"] = 6000527] = "GroupModuleCreateGroupWithDismissed";
    /** The number of group attributes reaches the upper bound error. */
    ZIMErrorCode[ZIMErrorCode["GroupModuleTheNumberOfGroupAttributesExceedsLimit"] = 6000531] = "GroupModuleTheNumberOfGroupAttributesExceedsLimit";
    /** The group attribute Key length has reached the upper limit. */
    ZIMErrorCode[ZIMErrorCode["GroupModuleTheLengthOfGroupAttributeKeyExceedsLimit"] = 6000532] = "GroupModuleTheLengthOfGroupAttributeKeyExceedsLimit";
    /** The group attribute Value length has reached the upper limit. */
    ZIMErrorCode[ZIMErrorCode["GroupModuleTheLengthOfGroupAttributeValueExceedsLimit"] = 6000533] = "GroupModuleTheLengthOfGroupAttributeValueExceedsLimit";
    /** The group attribute Value length has reached the upper limit. */
    ZIMErrorCode[ZIMErrorCode["GroupModuleTheTotalLengthOfGroupAttributesValueExceedsLimit"] = 6000534] = "GroupModuleTheTotalLengthOfGroupAttributesValueExceedsLimit";
    /** Group operation permission. */
    ZIMErrorCode[ZIMErrorCode["GroupModuleGroupNoCorrespondingOperationAuthority"] = 6000541] = "GroupModuleGroupNoCorrespondingOperationAuthority";
    /** Forbid to join the group. */
    ZIMErrorCode[ZIMErrorCode["GroupModuleForbidJoinGroupError"] = 6000542] = "GroupModuleForbidJoinGroupError";
    /** Need to apply to join the group. */
    ZIMErrorCode[ZIMErrorCode["GroupModuleNeedApplyJoinGroupError"] = 6000543] = "GroupModuleNeedApplyJoinGroupError";
    /** Need to apply to invite to join the group. */
    ZIMErrorCode[ZIMErrorCode["GroupModuleNeedApplyInviteGroupError"] = 6000544] = "GroupModuleNeedApplyInviteGroupError";
    /** Session operation error. */
    ZIMErrorCode[ZIMErrorCode["ConversationModuleCommonError"] = 6000601] = "ConversationModuleCommonError";
    /** Session background error. */
    ZIMErrorCode[ZIMErrorCode["ConversationModuleServerError"] = 6000602] = "ConversationModuleServerError";
    /** There are no errors in the session. */
    ZIMErrorCode[ZIMErrorCode["ConversationModuleConversationDoesNotExist"] = 6000603] = "ConversationModuleConversationDoesNotExist";
    /** The session top list has reached its limit. */
    ZIMErrorCode[ZIMErrorCode["ConversationModuleConversationPinnedListReachedLimit"] = 6000604] = "ConversationModuleConversationPinnedListReachedLimit";
    /** Error opening database. */
    ZIMErrorCode[ZIMErrorCode["DatabaseModuleOpenDatabaseError"] = 6000701] = "DatabaseModuleOpenDatabaseError";
    /** Error modifying database. */
    ZIMErrorCode[ZIMErrorCode["DatabaseModuleModifyDatabaseError"] = 6000702] = "DatabaseModuleModifyDatabaseError";
    /** Error deleting database. */
    ZIMErrorCode[ZIMErrorCode["DatabaseModuleDeleteDatabaseError"] = 6000703] = "DatabaseModuleDeleteDatabaseError";
    /** Error selecting database. */
    ZIMErrorCode[ZIMErrorCode["DatabaseModuleSeleteDatabaseError"] = 6000704] = "DatabaseModuleSeleteDatabaseError";
    /** The database is not open. */
    ZIMErrorCode[ZIMErrorCode["DatabaseModuleDbIsNotOpened"] = 6000710] = "DatabaseModuleDbIsNotOpened";
    /** The number of friends has reached the upper limit. */
    ZIMErrorCode[ZIMErrorCode["FriendModuleFriendNumsLimit"] = 6000801] = "FriendModuleFriendNumsLimit";
    /** The friend application status is wrong. */
    ZIMErrorCode[ZIMErrorCode["FriendModuleFriendApplicationStatusError"] = 6000802] = "FriendModuleFriendApplicationStatusError";
    /** The other party is already your friend. */
    ZIMErrorCode[ZIMErrorCode["FriendModuleIsAlreadyYourFriend"] = 6000803] = "FriendModuleIsAlreadyYourFriend";
    /** The other party is already in your blacklist. */
    ZIMErrorCode[ZIMErrorCode["FriendModuleAlreadyAddToBlacklist"] = 6000804] = "FriendModuleAlreadyAddToBlacklist";
    /** Cannot add yourself to the blacklist. */
    ZIMErrorCode[ZIMErrorCode["FriendModuleCannotAddSelfToBlacklist"] = 6000805] = "FriendModuleCannotAddSelfToBlacklist";
    /** The other party has been removed from your blacklist. */
    ZIMErrorCode[ZIMErrorCode["FriendModuleAlreadyDeleteFromBlacklist"] = 6000806] = "FriendModuleAlreadyDeleteFromBlacklist";
    /** The other party is not in your blacklist. */
    ZIMErrorCode[ZIMErrorCode["FriendModuleUserNotInBlacklist"] = 6000807] = "FriendModuleUserNotInBlacklist";
    /** The blacklist list quantity has reached the upper limit. */
    ZIMErrorCode[ZIMErrorCode["FriendModuleBlacklistListQuantityLimit"] = 6000808] = "FriendModuleBlacklistListQuantityLimit";
    /** The friend list has reached the upper limit. */
    ZIMErrorCode[ZIMErrorCode["FriendModuleFriendOperationLimitExceeded"] = 6000809] = "FriendModuleFriendOperationLimitExceeded";
    /** Cannot add yourself to the friend list. */
    ZIMErrorCode[ZIMErrorCode["FriendModuleCannotAddSelfToFriendList"] = 6000810] = "FriendModuleCannotAddSelfToFriendList";
    /** The other party is not registered. */
    ZIMErrorCode[ZIMErrorCode["FriendModuleFriendAreUnregistered"] = 6000811] = "FriendModuleFriendAreUnregistered";
    /** The other party is not your friend. */
    ZIMErrorCode[ZIMErrorCode["FriendModuleNotYourFriend"] = 6000812] = "FriendModuleNotYourFriend";
    /** Cannot delete yourself. */
    ZIMErrorCode[ZIMErrorCode["FriendModuleCannotDeleteSelf"] = 6000813] = "FriendModuleCannotDeleteSelf";
    /** The friend application has expired. */
    ZIMErrorCode[ZIMErrorCode["FriendModuleFriendApplicationExpired"] = 6000814] = "FriendModuleFriendApplicationExpired";
    /** Failed to add to blacklist. */
    ZIMErrorCode[ZIMErrorCode["FriendModuleAddBlacklistFail"] = 6000815] = "FriendModuleAddBlacklistFail";
    /** Failed to delete from blacklist. */
    ZIMErrorCode[ZIMErrorCode["FriendModuleDelBlacklistFail"] = 6000816] = "FriendModuleDelBlacklistFail";
})(ZIMErrorCode || (ZIMErrorCode = {}));
