// @zegocloud/zego-uikit-rn/lib/commonjs/signal_plugin/services/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _reactNative = require("react-native");
var _core = _interopRequireDefault(require("../core"));
var _defines = _interopRequireDefault(require("../core/defines"));
var _logger = require("../utils/logger");
var _ZegoUIKitCorePlugin = _interopRequireDefault(require("../../components/internal/ZegoUIKitCorePlugin"));
var _report = _interopRequireDefault(require("../../utils/report"));
var _ZegoPluginInvitationService;
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
class ZegoPluginInvitationService {
  constructor() {
    _defineProperty(this, "_androidOfflineDataHandler", void 0);
    _defineProperty(this, "_iOSOfflineDataHandler", void 0);
    _defineProperty(this, "_callKitAnswerCallHandler", void 0);
    _defineProperty(this, "_callKitEndCallHandler", void 0);
    _defineProperty(this, "_advancedConfig", {});
    if (!ZegoPluginInvitationService.shared) {
      ZegoPluginInvitationService.shared = this;
    }
    return ZegoPluginInvitationService.shared;
  }
  static getInstance() {
    if (!ZegoPluginInvitationService.shared) {
      ZegoPluginInvitationService.shared = new ZegoPluginInvitationService();
    }
    return ZegoPluginInvitationService.shared;
  }
  setBackgroundMessageHandler() {
    (0, _logger.zloginfo)('[ZegoPluginInvitationService] ZPNs setBackgroundMessageHandler');
    _ZegoUIKitCorePlugin.default.getZPNsPlugin().default.setBackgroundMessageHandler(message => {
      (0, _logger.zloginfo)('ZPNs backgroundMessageHandler message: ', message);
      const dataObj = JSON.parse(message.extras.payload);
      dataObj.zim_call_id = message.extras.call_id;

      // Pre-emptively determine whether to cancel for accurate reporting
      const cancelInvitation = dataObj && dataObj.operation_type === "cancel_invitation";
      if (!cancelInvitation) {
        _report.default.reportEvent('invitationReceived', {
          'call_id': dataObj.zim_call_id,
          'inviter': dataObj.inviter.id,
          'app_state': 'restarted',
          'extended_data': message.extras.payload
        });
      }
      let offlineDataHandler = ZegoPluginInvitationService.getInstance().getAndroidOfflineDataHandler();
      offlineDataHandler(dataObj);
    });
  }
  setThroughMessageReceivedHandler() {
    (0, _logger.zloginfo)('[ZegoPluginInvitationService] ZPNs setThroughMessageReceivedHandler');
    _ZegoUIKitCorePlugin.default.getZPNsPlugin().default.getInstance().on("throughMessageReceived", message => {
      (0, _logger.zloginfo)('ZPNs throughMessageReceived message: ', message);
      const dataObj = JSON.parse(message.extras.payload);
      dataObj.zim_call_id = message.extras.call_id;
      let offlineDataHandler = ZegoPluginInvitationService.getInstance().getAndroidOfflineDataHandler();
      offlineDataHandler(dataObj);
    });
  }
  setAndroidOfflineDataHandler(handler) {
    (0, _logger.zloginfo)('[ZegoPluginInvitationService] ZPNs setAndroidOfflineDataHandler');
    this._androidOfflineDataHandler = handler;
  }
  getAndroidOfflineDataHandler() {
    return this._androidOfflineDataHandler;
  }
  setIOSOfflineDataHandler(handler) {
    (0, _logger.zloginfo)(`[ZegoPluginInvitationService][setIOSOfflineDataHandler]`);
    this._iOSOfflineDataHandler = handler;
  }
  onCallKitAnswerCall(handler) {
    this._callKitAnswerCallHandler = handler;
  }
  onCallKitEndCall(handler) {
    this._callKitEndCallHandler = handler;
  }
  getIOSOfflineDataHandler() {
    return this._iOSOfflineDataHandler;
  }
  getAnswerCallHandle() {
    return this._callKitAnswerCallHandler;
  }
  getEndCallHandle() {
    return this._callKitEndCallHandler;
  }
  reportCallKitCallEnded(uuid, reason) {
    if (_ZegoUIKitCorePlugin.default.getCallKitPlugin()) {
      return _ZegoUIKitCorePlugin.default.getCallKitPlugin().default.getInstance().reportCallEnded(reason, uuid);
    } else {
      (0, _logger.zlogwarning)('[ZegoPluginInvitationService][reportCallKitCallEnded] Cannot getCallKitPlugin');
    }
  }
  reportIncomingCall(cxCallUpdate, uuid) {
    if (_ZegoUIKitCorePlugin.default.getCallKitPlugin()) {
      return _ZegoUIKitCorePlugin.default.getCallKitPlugin().default.getInstance().reportIncomingCall(cxCallUpdate, uuid);
    } else {
      (0, _logger.zlogwarning)('[ZegoPluginInvitationService][reportIncomingCall] Cannot getCallKitPlugin');
      return Promise.reject('Cannot get CallKit plugin');
    }
  }
  getZIMInstance() {
    return _core.default.getInstance().getZIMInstance();
  }
  getVersion() {
    return _core.default.getInstance().getVersion();
  }
  init(appID, appSign) {
    _core.default.getInstance().create({
      appID,
      appSign
    });

    // Some advanced key need to be called after create
    Object.keys(this._advancedConfig).forEach(key => {
      let value = this._advancedConfig[key];
      _core.default.getInstance().setAdvancedConfig(key, value);
    });
  }
  uninit() {
    _core.default.getInstance().destroy();
  }
  login(userID, userName, token) {
    return _core.default.getInstance().login({
      userID: userID,
      userName: userName,
      userAvatarUrl: ""
    }, token);
  }
  logout() {
    return _core.default.getInstance().logout();
  }
  enableNotifyWhenAppRunningInBackgroundOrQuit(certificateIndex, isIOSDevelopmentEnvironment, appName) {
    if (!certificateIndex) {
      certificateIndex = 1;
    }
    (0, _logger.zloginfo)(`[Service] enableNotifyWhenAppRunningInBackgroundOrQuit, certificateIndex: ${certificateIndex}, isIOSSandboxEnvironment: ${isIOSDevelopmentEnvironment}, appName: ${appName}`);
    if (_ZegoUIKitCorePlugin.default.getZPNsPlugin()) {
      if (_reactNative.Platform.OS === 'android') {
        (0, _logger.zloginfo)('[ZegoPluginInvitationService] registerPush, Android');
        _ZegoUIKitCorePlugin.default.getZPNsPlugin().default.setPushConfig({
          "enableFCMPush": true,
          "enableHWPush": false,
          "enableMiPush": false,
          "enableOppoPush": false,
          "enableVivoPush": false,
          "appType": certificateIndex
        });
        _ZegoUIKitCorePlugin.default.getZPNsPlugin().default.getInstance().registerPush({});
      } else if (_reactNative.Platform.OS === 'ios' && _ZegoUIKitCorePlugin.default.getCallKitPlugin()) {
        (0, _logger.zloginfo)('[ZegoPluginInvitationService] registerPush, iOS');
        const CXProviderConfiguration = {
          localizedName: appName ?? 'My app',
          iconTemplateImageName: "AppIcon"
        };
        _ZegoUIKitCorePlugin.default.getCallKitPlugin().default.setInitConfiguration(CXProviderConfiguration);
        _ZegoUIKitCorePlugin.default.getZPNsPlugin().default.getInstance().applyNotificationPermission();

        // 0 - Production; 1 - Development; 2 - Automatic
        const iOSEnvironment = isIOSDevelopmentEnvironment == null ? 2 : isIOSDevelopmentEnvironment ? 1 : 0;
        (0, _logger.zloginfo)('[ZegoPluginInvitationService] registerPush, isIOSSandboxEnvironment', iOSEnvironment);
        _ZegoUIKitCorePlugin.default.getZPNsPlugin().default.setPushConfig({
          'appType': certificateIndex
        });
        _ZegoUIKitCorePlugin.default.getZPNsPlugin().default.getInstance().registerPush({
          enableIOSVoIP: true,
          iOSEnvironment: iOSEnvironment
        });
      }
      _ZegoUIKitCorePlugin.default.getZPNsPlugin().default.getInstance().on("registered", message => {
        (0, _logger.zloginfo)("[ZegoPluginInvitationService] ZPNs registered,", message);
        if (message.msg.includes("SERVICE_NOT_AVAILABLE")) {
          (0, _logger.zlogwarning)('[ZegoPluginInvitationService] ZPNs registered, Please check the network connectivity with FCM.');
        }
      });
      _ZegoUIKitCorePlugin.default.getZPNsPlugin().default.getInstance().on("notificationArrived", message => {
        // only support for apple, xiaomi
        (0, _logger.zloginfo)("[ZegoPluginInvitationService] ZPNs notificationArrived,", message);
      });

      // ZegoUIKitCorePlugin.getZPNsPlugin().default.getInstance().on("notificationClicked", (message) => {
      //   zloginfo("@@@@@@@@@@@@@@@@notificationClicked>>>>>>>>>>>>>>>############", getCallID(message))
      //   setZpnState("notificationClicked: " + getCallID(message))
      // })
      // ZegoUIKitCorePlugin.getZPNsPlugin().default.getInstance().on("throughMessageReceived", (message) => {
      //   zloginfo("@@@@@@@@@@@@@@@@throughMessageReceived>>>>>>>>>>>>>>>############", getCallID(message))
      //   setZpnState("throughMessageReceived: " + getCallID(message))
      // })

      if (_ZegoUIKitCorePlugin.default.getCallKitPlugin()) {
        (0, _logger.zloginfo)('[ZegoPluginInvitationService] register iOS didReceiveIncomingPush');
        _ZegoUIKitCorePlugin.default.getCallKitPlugin().default.getInstance().on("didReceiveIncomingPush", (extras, uuid) => {
          (0, _logger.zloginfo)('#########didReceiveIncomingPush', extras, uuid);
          let {
            payload
          } = extras;
          let {
            call_id
          } = extras;
          const dataObj = payload ? JSON.parse(payload) : {};
          dataObj.zim_call_id = call_id;
          ZegoPluginInvitationService.getInstance().getIOSOfflineDataHandler()(dataObj, uuid);
        });
        _ZegoUIKitCorePlugin.default.getCallKitPlugin().default.getInstance().on("providerDidReset", () => {
          (0, _logger.zloginfo)('#########providerDidReset');
        });
        _ZegoUIKitCorePlugin.default.getCallKitPlugin().default.getInstance().on("providerDidBegin", () => {
          (0, _logger.zloginfo)('#########providerDidBegin');
        });
        _ZegoUIKitCorePlugin.default.getCallKitPlugin().default.getInstance().on("didActivateAudioSession", () => {
          (0, _logger.zloginfo)('#########didActivateAudioSession');
        });
        _ZegoUIKitCorePlugin.default.getCallKitPlugin().default.getInstance().on("didDeactivateAudioSession", () => {
          (0, _logger.zloginfo)('#########didDeactivateAudioSession');
        });
        _ZegoUIKitCorePlugin.default.getCallKitPlugin().default.getInstance().on("timedOutPerformingAction", action => {
          (0, _logger.zloginfo)('#########timedOutPerformingAction', action);
          action.fulfill();
        });
        _ZegoUIKitCorePlugin.default.getCallKitPlugin().default.getInstance().on("performStartCallAction", action => {
          (0, _logger.zloginfo)('#########performStartCallAction', action);
          action.fulfill();
        });
        _ZegoUIKitCorePlugin.default.getCallKitPlugin().default.getInstance().on("performAnswerCallAction", action => {
          (0, _logger.zloginfo)('#########performAnswerCallAction', action);
          ZegoPluginInvitationService.getInstance().getAnswerCallHandle()(action);
        });
        _ZegoUIKitCorePlugin.default.getCallKitPlugin().default.getInstance().on("performEndCallAction", action => {
          (0, _logger.zloginfo)('#########performEndCallAction', action);
          ZegoPluginInvitationService.getInstance().getEndCallHandle()(action);
        });
        _ZegoUIKitCorePlugin.default.getCallKitPlugin().default.getInstance().on("performSetHeldCallAction", action => {
          (0, _logger.zloginfo)('#########performSetHeldCallAction', action);
          action.fulfill();
        });
        _ZegoUIKitCorePlugin.default.getCallKitPlugin().default.getInstance().on("performSetMutedCallAction", action => {
          (0, _logger.zloginfo)('#########performSetMutedCallAction', action);
          action.fulfill();
        });
        _ZegoUIKitCorePlugin.default.getCallKitPlugin().default.getInstance().on("performSetGroupCallAction", action => {
          (0, _logger.zloginfo)('#########performSetGroupCallAction', action);
          action.fulfill();
        });
        _ZegoUIKitCorePlugin.default.getCallKitPlugin().default.getInstance().on("performPlayDTMFCallAction", action => {
          (0, _logger.zloginfo)('#########performPlayDTMFCallAction', action);
          action.fulfill();
        });
      }
    }
  }
  setAdvancedConfig(key, value) {
    if (key === 'zim_voip_call_id') {
      this._advancedConfig[key] = value;
    } else {
      _core.default.getInstance().setAdvancedConfig(key, value);
    }
  }
  sendInvitation(inviterName, invitees, timeout, type, data, notificationConfig) {
    if (!invitees.length) {
      (0, _logger.zlogerror)('[Service]Send invitees is empty.');
      return Promise.reject(new _defines.default());
    }
    const config = {
      timeout
    };
    config.extendedData = JSON.stringify({
      inviter_name: inviterName,
      type,
      data
    });
    if (_ZegoUIKitCorePlugin.default.getZPNsPlugin()) {
      config.pushConfig = {
        title: notificationConfig.title ?? "",
        content: notificationConfig.message ?? "",
        resourcesID: notificationConfig.resourceID ?? "",
        payload: data,
        voIPConfig: {
          iOSVoIPHandleType: 1,
          iOSVoIPHandleValue: '',
          iOSVoIPHasVideo: type === 1
        }
      };
    }
    (0, _logger.zloginfo)(`[Service]Send invitation invitees: ${invitees}`);
    (0, _logger.zloginfo)(`[Service]Send invitation config, timeout: ${timeout}, extendedData: ${config.extendedData}`);
    (0, _logger.zloginfo)(`[Service]Send invitation config, pushConfig: ${config.pushConfig ? JSON.stringify(config.pushConfig) : ''}`);
    return _core.default.getInstance().invite(invitees, config);
  }
  cancelInvitation(invitees, data, notificationConfig) {
    invitees = invitees.map(invitee => invitee);
    if (!invitees.length) {
      (0, _logger.zlogerror)('[Service]Cancel invitees is empty.');
      return Promise.reject(new _defines.default());
    }
    const config = {
      extendedData: data
    };
    if (_ZegoUIKitCorePlugin.default.getZPNsPlugin()) {
      config.pushConfig = {
        title: notificationConfig && notificationConfig.title,
        content: notificationConfig && notificationConfig.message,
        resourcesID: notificationConfig && notificationConfig.resourceID,
        payload: data
      };
    }
    const callID = _core.default.getInstance().getCallIDByUserID(_core.default.getInstance().getLocalUser().userID);
    (0, _logger.zloginfo)(`[Service]Cancel invitation: callID: ${callID}, invitees: ${invitees}, data: ${data}.`);
    return _core.default.getInstance().cancel(invitees, callID, config);
  }
  refuseInvitation(inviterID, data) {
    let callID;
    // Parse data and adapt automatic rejection
    if (data) {
      const dataObj = JSON.parse(data);
      callID = dataObj.callID;
    } else {
      callID = _core.default.getInstance().getCallIDByUserID(inviterID);
    }
    if (!callID) {
      (0, _logger.zlogerror)('[Service]Call id corresponding to the inviterID is empty.');
      return Promise.reject(new _defines.default());
    }
    const config = {
      extendedData: data
    };
    (0, _logger.zloginfo)(`[Service]Refuse invitation: callID: ${callID}, inviter id: ${inviterID}, data: ${data}.`);
    return _core.default.getInstance().reject(callID, config);
  }
  acceptInvitation(inviterID, data) {
    let callID = _core.default.getInstance().getCallIDByUserID(inviterID);
    if (!callID) {
      if (data) {
        const dataObj = JSON.parse(data);
        callID = dataObj.callID;
      }
      if (!callID) {
        (0, _logger.zlogerror)('[Service]Call id corresponding to the inviterID is empty.', inviterID, data);
        return Promise.reject(new _defines.default('', 'Call id corresponding to the inviterID is empty.'));
      }
    }
    const config = {
      extendedData: data
    };
    (0, _logger.zloginfo)(`[Service]Accept invitation: callID: ${callID}, inviter id: ${inviterID}, data: ${data}.`);
    return _core.default.getInstance().accept(callID, config);
  }
  queryCallList(count, nextFlag) {
    const config = {
      count: count,
      nextFlag: nextFlag
    };
    return _core.default.getInstance().queryCallList(config);
  }
  renewToken(token) {
    return _core.default.getInstance().renewToken(token);
  }
  onConnectionStateChanged(callbackID, callback) {
    _core.default.getInstance().onConnectionStateChanged(callbackID, callback);
  }
  onCallInvitationReceived(callbackID, callback) {
    _core.default.getInstance().onCallInvitationReceived(callbackID, callback);
  }
  onCallInvitationTimeout(callbackID, callback) {
    _core.default.getInstance().onCallInvitationTimeout(callbackID, callback);
  }
  onCallInviteesAnsweredTimeout(callbackID, callback) {
    _core.default.getInstance().onCallInviteesAnsweredTimeout(callbackID, callback);
  }
  onCallInvitationAccepted(callbackID, callback) {
    _core.default.getInstance().onCallInvitationAccepted(callbackID, callback);
  }
  onCallInvitationRejected(callbackID, callback) {
    _core.default.getInstance().onCallInvitationRejected(callbackID, callback);
  }
  onCallInvitationCancelled(callbackID, callback) {
    _core.default.getInstance().onCallInvitationCancelled(callbackID, callback);
  }
  onRequireNewToken(callbackID, callback) {
    _core.default.getInstance().onRequireNewToken(callbackID, callback);
  }
  onLoginSuccess(callbackID, callback) {
    _core.default.getInstance().onLoginSuccess(callbackID, callback);
  }
}
exports.default = ZegoPluginInvitationService;
_ZegoPluginInvitationService = ZegoPluginInvitationService;
_defineProperty(ZegoPluginInvitationService, "shared", void 0);
