// @zegocloud/zego-uikit-rn/lib/commonjs/signal_plugin/services/room_properties_service.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _room_properties_core = _interopRequireDefault(require("../core/room_properties_core"));
var _ZegoPluginRoomPropertiesService;
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
class ZegoPluginRoomPropertiesService {
  constructor() {
    if (!ZegoPluginRoomPropertiesService.shared) {
      ZegoPluginRoomPropertiesService.shared = this;
    }
    return ZegoPluginRoomPropertiesService.shared;
  }
  static getInstance() {
    if (!ZegoPluginRoomPropertiesService.shared) {
      ZegoPluginRoomPropertiesService.shared = new ZegoPluginRoomPropertiesService();
    }
    return ZegoPluginRoomPropertiesService.shared;
  }
  updateRoomProperty(key, value, isDeleteAfterOwnerLeft, isForce = false, isUpdateOwner = false) {
    const attributes = {
      [key]: value
    };
    const config = {
      isForce,
      isDeleteAfterOwnerLeft,
      isUpdateOwner
    };
    return _room_properties_core.default.getInstance().updateRoomProperty(attributes, config);
  }
  deleteRoomProperties(keys = [], isForce) {
    const config = {
      isForce
    };
    return _room_properties_core.default.getInstance().deleteRoomProperties(keys, config);
  }
  beginRoomPropertiesBatchOperation(isDeleteAfterOwnerLeft = false, isForce = false, isUpdateOwner = false) {
    const config = {
      isDeleteAfterOwnerLeft,
      isForce,
      isUpdateOwner
    };
    return _room_properties_core.default.getInstance().beginRoomPropertiesBatchOperation(config);
  }
  endRoomPropertiesBatchOperation() {
    return _room_properties_core.default.getInstance().endRoomPropertiesBatchOperation();
  }
  queryRoomProperties() {
    return _room_properties_core.default.getInstance().queryRoomProperties();
  }
  onRoomPropertyUpdated(callbackID, callback) {
    _room_properties_core.default.getInstance().onRoomPropertyUpdated(callbackID, callback);
  }
}
exports.default = ZegoPluginRoomPropertiesService;
_ZegoPluginRoomPropertiesService = ZegoPluginRoomPropertiesService;
_defineProperty(ZegoPluginRoomPropertiesService, "shared", void 0);
