// @zegocloud/zego-uikit-prebuilt-call-rn/lib/commonjs/services/defines.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ZegoCallEndReason = exports.ONE_ON_ONE_VOICE_CALL_CONFIG = exports.ONE_ON_ONE_VIDEO_CALL_CONFIG = exports.GROUP_VOICE_CALL_CONFIG = exports.GROUP_VIDEO_CALL_CONFIG = void 0;
Object.defineProperty(exports, "ZegoMenuBarButtonName", {
  enumerable: true,
  get: function () {
    return _ZegoMenuBarButtonName.default;
  }
});
exports.ZegoMultiCertificate = void 0;
var _zegoUikitRn = require("@zegocloud/zego-uikit-rn");
var _ZegoMenuBarButtonName = _interopRequireDefault(require("../call/ZegoMenuBarButtonName"));
var _ZegoMenuBarStyle = _interopRequireDefault(require("../call/ZegoMenuBarStyle"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const ONE_ON_ONE_VIDEO_CALL_CONFIG = exports.ONE_ON_ONE_VIDEO_CALL_CONFIG = {
  turnOnCameraWhenJoining: true,
  turnOnMicrophoneWhenJoining: true,
  useSpeakerWhenJoining: true,
  layout: {
    mode: _zegoUikitRn.ZegoLayoutMode.pictureInPicture
  },
  bottomMenuBarConfig: {
    buttons: [_ZegoMenuBarButtonName.default.toggleCameraButton, _ZegoMenuBarButtonName.default.switchCameraButton, _ZegoMenuBarButtonName.default.hangUpButton, _ZegoMenuBarButtonName.default.toggleMicrophoneButton, _ZegoMenuBarButtonName.default.switchAudioOutputButton],
    style: _ZegoMenuBarStyle.default.light
  },
  topMenuBarConfig: {
    buttons: [],
    isVisible: false
  },
  audioVideoViewConfig: {
    useVideoViewAspectFill: true
  }
};
const ONE_ON_ONE_VOICE_CALL_CONFIG = exports.ONE_ON_ONE_VOICE_CALL_CONFIG = {
  turnOnCameraWhenJoining: false,
  turnOnMicrophoneWhenJoining: true,
  useSpeakerWhenJoining: false,
  layout: {
    mode: _zegoUikitRn.ZegoLayoutMode.pictureInPicture
  },
  bottomMenuBarConfig: {
    buttons: [_ZegoMenuBarButtonName.default.toggleMicrophoneButton, _ZegoMenuBarButtonName.default.hangUpButton, _ZegoMenuBarButtonName.default.switchAudioOutputButton],
    style: _ZegoMenuBarStyle.default.light
  },
  topMenuBarConfig: {
    buttons: [],
    isVisible: false
  },
  audioVideoViewConfig: {
    useVideoViewAspectFill: true
  }
};
const GROUP_VIDEO_CALL_CONFIG = exports.GROUP_VIDEO_CALL_CONFIG = {
  turnOnCameraWhenJoining: true,
  turnOnMicrophoneWhenJoining: true,
  useSpeakerWhenJoining: true,
  layout: {
    mode: _zegoUikitRn.ZegoLayoutMode.gallery
  },
  bottomMenuBarConfig: {
    buttons: [_ZegoMenuBarButtonName.default.toggleCameraButton, _ZegoMenuBarButtonName.default.switchCameraButton, _ZegoMenuBarButtonName.default.hangUpButton, _ZegoMenuBarButtonName.default.toggleMicrophoneButton, _ZegoMenuBarButtonName.default.switchAudioOutputButton]
  },
  topMenuBarConfig: {
    buttons: [_ZegoMenuBarButtonName.default.showMemberListButton]
  }
};
const GROUP_VOICE_CALL_CONFIG = exports.GROUP_VOICE_CALL_CONFIG = {
  turnOnCameraWhenJoining: false,
  turnOnMicrophoneWhenJoining: true,
  useSpeakerWhenJoining: true,
  layout: {
    mode: _zegoUikitRn.ZegoLayoutMode.gallery
  },
  bottomMenuBarConfig: {
    buttons: [_ZegoMenuBarButtonName.default.toggleMicrophoneButton, _ZegoMenuBarButtonName.default.hangUpButton, _ZegoMenuBarButtonName.default.switchAudioOutputButton]
  },
  topMenuBarConfig: {
    buttons: [_ZegoMenuBarButtonName.default.showMemberListButton]
  },
  memberListConfig: {
    showCameraState: false
  }
};
const ZegoMultiCertificate = exports.ZegoMultiCertificate = {
  first: 1,
  second: 2
};
const ZegoCallEndReason = exports.ZegoCallEndReason = {
  /// the call ended due to a local hang-up
  localHangUp: 0,
  /// the call ended when the remote user hung up, leaving only one local user in the call
  remoteHangUp: 1,
  /// the call ended due to being kicked out
  kickOut: 2
};
