// @zegocloud/zego-uikit-prebuilt-call-rn/lib/commonjs/services/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _reactNative = require("react-native");
var _zegoUikitRn = _interopRequireWildcard(require("@zegocloud/zego-uikit-rn"));
var _minimizing_helper = _interopRequireDefault(require("../call/services/minimizing_helper"));
var _prebuilt_helper = _interopRequireDefault(require("../call/services/prebuilt_helper"));
var _bell = _interopRequireDefault(require("../call_invitation/services/bell"));
var _callevent_notify_app = _interopRequireDefault(require("../call_invitation/services/callevent_notify_app"));
var _hangup_helper = _interopRequireDefault(require("../call_invitation/services/hangup_helper"));
var _inner_text_helper = _interopRequireDefault(require("../call_invitation/services/inner_text_helper"));
var _invite_state_manager = _interopRequireDefault(require("../call_invitation/services/invite_state_manager"));
var _notification_helper = _interopRequireDefault(require("../call_invitation/services/notification_helper"));
var _offline_call_event_listener = _interopRequireDefault(require("../call_invitation/services/offline_call_event_listener"));
var _plugins = _interopRequireDefault(require("../call_invitation/services/plugins"));
var _logger = require("../utils/logger");
var _report = _interopRequireDefault(require("../utils/report"));
var _package_version = require("../utils/package_version");
var _defines = require("./defines");
var _invitation = _interopRequireDefault(require("./invitation"));
var _timing_helper = _interopRequireDefault(require("./timing_helper"));
var _EventEmitter = require("../utils/EventEmitter");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
const {
  ZegoUIKitPrebuiltCallRNModule
} = _reactNative.NativeModules;
class ZegoUIKitPrebuiltCallService {
  constructor() {
    _defineProperty(this, "_instance", void 0);
    _defineProperty(this, "appInfo", {});
    _defineProperty(this, "localUser", {});
    _defineProperty(this, "config", {
      ringtoneConfig: {
        incomingCallFileName: 'zego_incoming.mp3',
        outgoingCallFileName: 'zego_outgoing.mp3'
      },
      innerText: {},
      androidNotificationConfig: {
        channelID: "CallInvitation",
        channelName: "CallInvitation"
      },
      notifyWhenAppRunningInBackgroundOrQuit: true,
      // false not supported
      isIOSSandboxEnvironment: undefined,
      waitingPageConfig: {
        // avatarBuilder
        // nameBuilder
        // stateBuilder
        // hangupBuilder
        // switchCameraBuilder,
        // backgroundColor,
        // foregroundBuilder,
      }
      // onIncomingCallDeclineButtonPressed,
      // onIncomingCallAcceptButtonPressed,
      // onIncomingCallReceived,
      // onIncomingCallCanceled,
      // onIncomingCallTimeout,
      // onIncomingCallRejectedCauseBusy,
      // onOutgoingCallCancelButtonPressed,
      // onOutgoingCallAccepted,
      // onOutgoingCallRejectedCauseBusy,
      // onOutgoingCallDeclined,
      // onOutgoingCallTimeout,
    });
    _defineProperty(this, "isInit", false);
    _defineProperty(this, "subscription", null);
    _defineProperty(this, "onInitCallbackMap", {});
    _defineProperty(this, "onRouteChangeCallbackMap", {});
  }
  static getInstance() {
    return this._instance || (this._instance = new ZegoUIKitPrebuiltCallService());
  }
  getModuleName() {
    return 'PrebuiltCall';
  }
  useSystemCallingUI(plugins) {
    (0, _logger.zloginfo)(`[ZegoUIKitPrebuiltCallService][useSystemCallingUI] plugins: ${plugins}`);
    _zegoUikitRn.ZegoUIKitLogger.logFlush();
    _offline_call_event_listener.default.getInstance().useSystemCallingUI(plugins);
  }
  init(appID, appSign, userID, userName, plugins, config = {}) {
    _zegoUikitRn.ZegoUIKitLogger.logSetUserID(userID);
    if (this.isInit) {
      this.notifyInit();
      Object.assign(this.config, config);
      return Promise.resolve();
    }
    (0, _logger.zloginfo)(`[ZegoUIKitPrebuiltCallService][init] appID: ${appID}, userID: ${userID}, userName: ${userName}, plugins: ${plugins}`);
    this.appInfo = {
      appID,
      appSign
    };
    this.localUser = {
      userID,
      userName
    };
    Object.assign(this.config, config);

    // Call invitation
    const {
      ringtoneConfig,
      innerText,
      notifyWhenAppRunningInBackgroundOrQuit,
      certificateIndex,
      isIOSSandboxEnvironment
    } = this.config;
    // Init inner text helper
    _inner_text_helper.default.instance().init(innerText);
    // Monitor application status
    this.subscription = _reactNative.AppState.addEventListener('change', nextAppState => {
      if (nextAppState === 'active') {
        _plugins.default.reconnectIfDisconnected();
      }
    });
    if (_zegoUikitRn.default.getPluginName('ZPNs')) {
      // Enable offline notification
      _zegoUikitRn.default.getSignalingPlugin().enableNotifyWhenAppRunningInBackgroundOrQuit(certificateIndex, isIOSSandboxEnvironment, this.config.appName);
    } else {
      (0, _logger.zloginfo)('[ZegoUIKitPrebuiltCallService] ZPNs not installed');
    }
    if (Platform.OS === 'ios' && !_zegoUikitRn.default.getPluginName('CallKit')) {
      (0, _logger.zloginfo)('[ZegoUIKitPrebuiltCallService] iOS CallKit not installed');
    }
    _notification_helper.default.getInstance().init();
    _report.default.create(appID, appSign, {
      'call_version': (0, _package_version.getPackageVersion)(),
      'user_id': userID
    });
    const eventBegin = Date.now();
    return new Promise((resolve, reject) => {
      setTimeout(() => {
        _plugins.default.init(appID, appSign, userID, userName, plugins).then(() => {
          _report.default.reportEvent('call/init', {
            'error': 0,
            'msg': '',
            'start_time': eventBegin
          });

          // You can customize the textual descriptions for permission requests within this callback.
          resolve();
          this.requestNotificationPermission({
            message: 'To display incoming call notifications in background, we need the "Display notifications" permission. Please allow it so you never miss a call.',
            allow: 'Allow',
            deny: 'Deny'
          });
          this.requestFullIntentPermission({
            message: 'To display incoming call notifications on the lock screen, we need the "Display full-screen notifications" permission. Please allow it so you never miss a call.',
            allow: 'Allow',
            deny: 'Deny'
          });
          _offline_call_event_listener.default.getInstance().init(this.config);
          _callevent_notify_app.default.getInstance().init(this.config);
          _invite_state_manager.default.init();
          _bell.default.initRingtoneConfig(ringtoneConfig);
          _bell.default.initIncomingSound();
          _bell.default.initOutgoingSound();
          this.isInit = true;
          this.notifyInit();
        }).catch(() => {
          _report.default.reportEvent('call/init', {
            'error': -1,
            'msg': 'unknown',
            'start_time': eventBegin
          });
          reject();
        });
      }, 500);
    });
  }
  uninit() {
    _report.default.reportEvent('call/unInit', null);
    if (this.isInit) {
      _plugins.default.uninit();
      _bell.default.releaseIncomingSound();
      _bell.default.releaseOutgoingSound();
      _invite_state_manager.default.uninit();
      this.subscription.remove();
      _offline_call_event_listener.default.getInstance().uninit();
      this.isInit = false;
    }
  }
  getInitUser() {
    return this.localUser;
  }
  getInitAppInfo() {
    return this.appInfo;
  }
  getInitConfig() {
    return this.config;
  }
  notifyInit() {
    Object.keys(this.onInitCallbackMap).forEach(callbackID => {
      if (this.onInitCallbackMap[callbackID]) {
        this.onInitCallbackMap[callbackID]();
      }
    });
  }
  onInit(callbackID, callback) {
    if (typeof callback !== 'function') {
      delete this.onInitCallbackMap[callbackID];
    } else {
      this.onInitCallbackMap[callbackID] = callback;
    }
  }
  hangUp(showConfirmation = false) {
    (0, _logger.zloginfo)(`[ZegoUIKitPrebuiltCallService][hangUp] showConfirmation: ${showConfirmation}`);
    const debounce = _hangup_helper.default.getInstance().getDebounce();
    const config = this.config.requireConfig ? this.config.requireConfig({}) : {};
    const {
      onCallEnd,
      onHangUpConfirmation,
      hangUpConfirmInfo
    } = config;
    if (debounce) {
      (0, _logger.zloginfo)(`[ZegoUIKitPrebuiltCallService][hangUp] debounce: ${debounce}, return`);
      return;
    }
    let routeParams = _prebuilt_helper.default.getInstance().getRouteParams();
    if (!routeParams.roomID) {
      (0, _logger.zloginfo)(`[ZegoUIKitPrebuiltCallService][hangUp] can't get roomID from routeParams, return`);
      return;
    }
    if (!showConfirmation) {
      _hangup_helper.default.getInstance().setDebounce(true);
      this._hangupProcess(onCallEnd, routeParams.roomID, _defines.ZegoCallEndReason.localHangUp);
      _hangup_helper.default.getInstance().setDebounce(false);
    } else {
      _hangup_helper.default.getInstance().setDebounce(true);
      const temp = onHangUpConfirmation || _hangup_helper.default.getInstance().showLeaveAlert.bind(this, hangUpConfirmInfo);
      temp().then(() => {
        this._hangupProcess(onCallEnd, routeParams.roomID, _defines.ZegoCallEndReason.localHangUp);
        _hangup_helper.default.getInstance().setDebounce(false);
      });
    }
  }
  minimizeWindow() {
    _minimizing_helper.default.getInstance().minimizeWindow();
  }
  showToast(toastData = {}) {
    (0, _logger.zloginfo)('[ZegoUIKitPrebuiltCallService][showToast]', toastData);
    _EventEmitter.eventEmitter.emit(_EventEmitter.EventName.SHOW_TOAST, toastData);
  }
  sendCallInvitation(invitees, isVideoCall, navigation, options) {
    return _invitation.default.getInstance().sendCallInvitation(invitees, isVideoCall, navigation, options);
  }
  requestNotificationPermission(alert) {
    if (Platform.OS === 'ios') {
      return;
    }
    ZegoUIKitPrebuiltCallRNModule.requestNotificationPermission(alert.message, alert.allow, alert.deny);
  }
  requestSystemAlertWindow(alert) {
    if (Platform.OS === 'ios') {
      return;
    }
    ZegoUIKitPrebuiltCallRNModule.requestSystemAlertWindow(alert.message, alert.allow, alert.deny);
  }
  requestFullIntentPermission(alert) {
    if (Platform.OS === 'ios') {
      return;
    }
    ZegoUIKitPrebuiltCallRNModule.requestFullIntentPermission(alert.message, alert.allow, alert.deny);
  }
  _hangupProcess(onCallEnd, roomID, endReason) {
    const isMinimize = _minimizing_helper.default.getInstance().getIsMinimize();
    (0, _logger.zloginfo)('[ZegoUIKitPrebuiltCallService][_hangupProcess] roomID: ', roomID, 'endReason: ', endReason, 'isMinimize: ', isMinimize);
    const duration = _timing_helper.default.getInstance().getDuration();
    if (typeof onCallEnd !== 'function') {
      _hangup_helper.default.getInstance().notifyAutoJump();
    } else {
      (0, _logger.zloginfo)('[ZegoUIKitPrebuiltCallService][hangUp] notify requireConfig.onCallEnd will');
      onCallEnd(roomID, endReason, duration);
      (0, _logger.zloginfo)('[ZegoUIKitPrebuiltCallService][hangUp] notify requireConfig.onCallEnd succeed');
      if (isMinimize) {
        _prebuilt_helper.default.getInstance().notifyDestroyPrebuilt();
      }
    }
  }
}
exports.default = ZegoUIKitPrebuiltCallService;
