// @zegocloud/zego-uikit-prebuilt-call-rn/lib/commonjs/services/invitation.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _defines = require("../call_invitation/services/defines");
var _plugins = _interopRequireDefault(require("../call_invitation/services/plugins"));
var _zegoUikitRn = _interopRequireDefault(require("@zegocloud/zego-uikit-rn"));
var _invite_state_manager = _interopRequireDefault(require("../call_invitation/services/invite_state_manager"));
var _inner_text_helper = _interopRequireDefault(require("../call_invitation/services/inner_text_helper"));
var _logger = require("../utils/logger");
var _offline_call_event_listener = _interopRequireDefault(require("../call_invitation/services/offline_call_event_listener"));
var _report = _interopRequireDefault(require("../utils/report"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
class ZegoUIKitPrebuiltCallInvitation {
  constructor() {}
  static getInstance() {
    return this._instance || (this._instance = new ZegoUIKitPrebuiltCallInvitation());
  }
  sendCallInvitation(invitees, isVideoCall, navigation, options) {
    const {
      resourceID = '',
      timeout = 60,
      notificationTitle,
      notificationMessage,
      callID,
      customData = '',
      callName,
      showWaitingPageWhenGroupCall = false
    } = options;
    const localUser = _plugins.default.getLocalUser();
    const roomID = callID && callID.length > 0 ? callID : `call_${localUser.userID}_${Date.now()}`;
    const nonNullNotificationTitle = notificationTitle ?? _inner_text_helper.default.instance().getIncomingCallDialogTitle(localUser.userName, isVideoCall ? _defines.ZegoInvitationType.videoCall : _defines.ZegoInvitationType.voiceCall, invitees.length);
    const nonNullNotificationMessage = notificationMessage ?? _inner_text_helper.default.instance().getIncomingCallDialogMessage(isVideoCall ? _defines.ZegoInvitationType.videoCall : _defines.ZegoInvitationType.voiceCall, invitees.length);
    const data = JSON.stringify({
      call_id: roomID,
      call_name: callName,
      invitees: invitees.map(invitee => {
        return {
          user_id: invitee.userID,
          user_name: invitee.userName
        };
      }),
      type: isVideoCall ? _defines.ZegoInvitationType.videoCall : _defines.ZegoInvitationType.voiceCall,
      inviter: {
        id: localUser.userID,
        name: localUser.userName
      },
      timeout: timeout,
      custom_data: customData,
      notificationTitle: nonNullNotificationTitle,
      notificationMessage: nonNullNotificationMessage
    });
    const inviteeIDs = invitees.map(invitee => {
      return invitee.userID;
    });
    const type = isVideoCall ? _defines.ZegoInvitationType.videoCall : _defines.ZegoInvitationType.voiceCall;

    // set notification config.
    const notificationConfig = {
      resourceID,
      title: nonNullNotificationTitle,
      message: nonNullNotificationMessage
    };
    return new Promise((resolve, reject) => {
      _zegoUikitRn.default.getSignalingPlugin().sendInvitation(inviteeIDs, timeout, type, data, notificationConfig).then(({
        code,
        message,
        callID,
        errorInvitees
      }) => {
        _report.default.reportEvent('call/invite', {
          'call_id': callID,
          'room_id': callID,
          'source': 'sendCallInvitation'
        });
        (0, _logger.zloginfo)(`[CallInvitation]Send invitation success, errCode: ${code}, message: ${message}, errorInvitees: ${errorInvitees}`);
        if (inviteeIDs.length > errorInvitees.length) {
          const successfulInvitees = JSON.parse(JSON.stringify(inviteeIDs));
          errorInvitees.forEach(errorInviteeID => {
            const index = successfulInvitees.findIndex(inviteeID => errorInviteeID === inviteeID);
            index !== -1 && successfulInvitees.splice(index, 1);
          });
          this.onInvitationSent(navigation, callID, invitees, successfulInvitees, roomID, isVideoCall, callName, showWaitingPageWhenGroupCall, timeout);
          resolve();
        } else {
          reject(-1, 'All invitees failed.');
        }
      }).catch(({
        code,
        message
      }) => {
        reject(code, message);
      });
    });
  }
  onInvitationSent(navigation, callID, allInvitees, successfulInvitees, roomID, isVideoCall, callName, showWaitingPageWhenGroupCall, timeout) {
    (0, _logger.zloginfo)(`[onInvitationSent] callID: ${callID}, roomID: ${roomID}`);
    const localUser = _plugins.default.getLocalUser();
    _invite_state_manager.default.addInviteData(callID, localUser.userID, successfulInvitees);
    if (allInvitees.length === 1 || showWaitingPageWhenGroupCall) {
      // Jump to call waiting page
      (0, _logger.zloginfo)('Jump to call waiting page.');
      navigation.navigate('ZegoUIKitPrebuiltCallWaitingScreen', {
        roomID,
        isVideoCall,
        invitees: allInvitees,
        inviterID: localUser.userID,
        invitationID: callID,
        callName,
        timeout
      });
    } else {
      // Jump to call room page
      (0, _logger.zloginfo)('Jump to call room page.');
      const inviteeIDs = allInvitees.map(invitee => {
        return invitee.userID;
      });
      (0, _logger.zloginfo)('Navigate to ZegoUIKitPrebuiltCallInCallScreen');
      navigation.navigate('ZegoUIKitPrebuiltCallInCallScreen', {
        roomID,
        isVideoCall,
        invitees: inviteeIDs,
        inviter: localUser.userID,
        invitationID: callID
      });
    }
  }
}
exports.default = ZegoUIKitPrebuiltCallInvitation;
_defineProperty(ZegoUIKitPrebuiltCallInvitation, "_instance", void 0);
