// @zegocloud/zego-uikit-prebuilt-call-rn/lib/commonjs/services/timing_helper.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _reactNative = require("react-native");
var _logger = require("../utils/logger");
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
class TimingHelper {
  constructor() {
    this._duration = 0;
    this._durationStart = Date.now();
    this._onDurationUpdateCallbackMap = {};
    this._timer = null;
  }
  static getInstance() {
    if (TimingHelper._instance) {
      return TimingHelper._instance;
    }
    TimingHelper._instance = new TimingHelper();
    return TimingHelper._instance;
  }
  startTimer() {
    (0, _logger.zloginfo)('[TimingHelper][startTimer]');
    if (!this._timer) {
      this._durationStart = Date.now();
      this._duration = 0;
      this._timer = setInterval(() => {
        this._increaseDuration();
      }, 1000);
    }
  }
  stopTimer() {
    (0, _logger.zloginfo)('[TimingHelper][stopTimer]');
    if (this._timer) {
      clearInterval(this._timer);
      this._timer = null;
      this._durationStart = Date.now();
      this._duration = 0;
    }
  }
  resetDuration() {
    (0, _logger.zloginfo)('[TimingHelper][resetDuration]');
    this.stopTimer();
    this._duration = 0;
    this._durationStart = Date.now();
  }
  _increaseDuration() {
    if (_reactNative.Platform.OS === 'android') {
      // The timing depends on the hardware, and on machines with poor performance, the timing may be slower than the actual
      const realDuration = Math.floor((Date.now() - this._durationStart) / 1000);
      const duration = this._duration + 1;
      if (realDuration >= duration + 1) {
        this._duration = realDuration;
      } else {
        this._duration = duration;
      }
    } else {
      this._duration = Math.floor((Date.now() - this._durationStart) / 1000);
    }
    this.notifyDurationUpdate(this._duration);
  }
  getDuration() {
    const realDuration = Math.floor((Date.now() - this._durationStart) / 1000);
    if (realDuration - this._duration > 2) {
      return realDuration;
    }
    return this._duration;
  }
  onDurationUpdate(callbackID, callback) {
    if (typeof callback !== 'function') {
      delete this._onDurationUpdateCallbackMap[callbackID];
    } else {
      this._onDurationUpdateCallbackMap[callbackID] = callback;
    }
  }
  async notifyDurationUpdate(duration) {
    Object.keys(this._onDurationUpdateCallbackMap).forEach(callbackID => {
      if (this._onDurationUpdateCallbackMap[callbackID]) {
        this._onDurationUpdateCallbackMap[callbackID](duration);
      }
    });
  }
}
exports.default = TimingHelper;
_defineProperty(TimingHelper, "_instance", void 0);
