// @zegocloud/zego-uikit-prebuilt-call-rn/lib/commonjs/call/AudioVideoForegroundView.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = AudioVideoForegroundView;
var _react = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
var _zegoUikitRn = require("@zegocloud/zego-uikit-rn");
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
function AudioVideoForegroundView(props) {
  const {
    userInfo,
    showUserNameOnView,
    showCameraStateOnView,
    showMicrophoneStateOnView
  } = props;
  const {
    userID = '',
    userName = ''
  } = userInfo;
  const [bottomContainerMaxWidth, setBottomContainerMaxWidth] = (0, _react.useState)(0);
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.foregroundViewContainer,
    onLayout: event => {
      const {
        width
      } = event.nativeEvent.layout;
      setBottomContainerMaxWidth(width - 10);
    }
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.bottomContainer, {
      maxWidth: bottomContainerMaxWidth
    }]
  }, showUserNameOnView ? /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.nameLabelContainer
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: styles.nameLabel
  }, userName)) : /*#__PURE__*/_react.default.createElement(_reactNative.View, null), showCameraStateOnView ? /*#__PURE__*/_react.default.createElement(_zegoUikitRn.ZegoCameraStateIcon, {
    userID: userID,
    style: styles.deviceIcon
  }) : /*#__PURE__*/_react.default.createElement(_reactNative.View, null), showMicrophoneStateOnView ? /*#__PURE__*/_react.default.createElement(_zegoUikitRn.ZegoMicrophoneStateIcon, {
    userID: userID,
    style: styles.deviceIcon
  }) : /*#__PURE__*/_react.default.createElement(_reactNative.View, null)));
}
const styles = _reactNative.StyleSheet.create({
  foregroundViewContainer: {
    flex: 1,
    position: 'absolute',
    width: '100%',
    height: '100%',
    zIndex: 10
  },
  bottomContainer: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'flex-end',
    alignItems: 'center',
    backgroundColor: '#2A2A2A',
    opacity: 0.5,
    position: 'absolute',
    alignSelf: 'center',
    paddingLeft: 5,
    paddingRight: 5,
    paddingBottom: 3,
    paddingTop: 3,
    borderRadius: 6,
    bottom: 5,
    right: 5
  },
  nameLabelContainer: {
    alignSelf: 'center',
    flexShrink: 1
  },
  nameLabel: {
    color: '#FFFFFF',
    fontSize: 12
  },
  deviceIcon: {
    flex: 1,
    position: 'absolute'
  }
});
