// @zegocloud/zego-uikit-prebuilt-call-rn/lib/commonjs/call/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
var _native = require("@react-navigation/native");
var _reactDelegateComponent = _interopRequireDefault(require("react-delegate-component"));
var _reactNativeKeepAwake = require("@sayem314/react-native-keep-awake");
var _reactNativeSafeAreaContext = require("react-native-safe-area-context");
var _zegoUikitRn = _interopRequireWildcard(require("@zegocloud/zego-uikit-rn"));
var _defines = require("../services/defines");
var _timing_helper = _interopRequireDefault(require("../services/timing_helper"));
var _keyboard = require("../utils/keyboard");
var _logger = require("../utils/logger");
var _minimizing_helper = _interopRequireDefault(require("./services/minimizing_helper"));
var _prebuilt_helper = _interopRequireDefault(require("./services/prebuilt_helper"));
var _AudioVideoForegroundView = _interopRequireDefault(require("./AudioVideoForegroundView"));
var _ZegoBottomBar = _interopRequireDefault(require("./ZegoBottomBar"));
var _ZegoCallMemberList = _interopRequireDefault(require("./ZegoCallMemberList"));
var _ZegoMenuBarButtonName = _interopRequireDefault(require("./ZegoMenuBarButtonName"));
var _ZegoMenuBarStyle = _interopRequireDefault(require("./ZegoMenuBarStyle"));
var _ZegoPrebuiltForegroundView = _interopRequireDefault(require("./ZegoPrebuiltForegroundView"));
var _ZegoTopMenuBar = _interopRequireDefault(require("./ZegoTopMenuBar"));
var _EventEmitter = require("../utils/EventEmitter");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
function ZegoUIKitPrebuiltCall(props, ref) {
  const navigation = (0, _native.useNavigation)();
  const insets = (0, _reactNativeSafeAreaContext.useSafeAreaInsets)();
  const isMinimizeSwitch = _minimizing_helper.default.getInstance().getIsMinimizeSwitch();
  !isMinimizeSwitch && _minimizing_helper.default.getInstance().notifyEntryNormal();
  const {
    appID,
    appSign,
    userID,
    userName,
    callID,
    config,
    token = '',
    onRequireNewToken
  } = props;
  _minimizing_helper.default.getInstance().setInitParams(appID, appSign, userID, userName, callID, {
    ...config,
    token,
    onRequireNewToken
  });
  const {
    audioVideoViewConfig = {},
    // turnOnCameraWhenJoining = true,
    // turnOnMicrophoneWhenJoining = true,
    // useSpeakerWhenJoining = true,

    bottomMenuBarConfig = {},
    topMenuBarConfig = {},
    memberListConfig = {},
    layout = {},
    // {mode, config}

    hangUpConfirmInfo,
    // {title: '', cancelButtonName: '', confirmButtonName: ''}

    onHangUpConfirmation,
    onJoinRoom,
    onCallEnd,
    onUserJoin,
    durationConfig = {},
    // Deprecate
    timingConfig = {},
    avatarBuilder,
    foregroundBuilder: prebuiltForegroundBuilder,
    messageItemBuilder,
    useFrontFacingCamera = true
  } = config;
  const {
    showMicrophoneStateOnView = true,
    showCameraStateOnView = false,
    showUserNameOnView = true,
    showSoundWavesInAudioMode = true,
    useVideoViewAspectFill = true,
    foregroundBuilder
  } = audioVideoViewConfig;
  const {
    buttons = [_ZegoMenuBarButtonName.default.toggleCameraButton, _ZegoMenuBarButtonName.default.switchCameraButton, _ZegoMenuBarButtonName.default.hangUpButton, _ZegoMenuBarButtonName.default.toggleMicrophoneButton, _ZegoMenuBarButtonName.default.switchAudioOutputButton],
    maxCount = 5,
    extendButtons = [],
    hideAutomatically = true,
    hideByClick = true,
    style = _ZegoMenuBarStyle.default.light
  } = bottomMenuBarConfig;
  const {
    isVisible = true,
    title: topTitle = userName || '',
    buttons: topButtons = [],
    maxCount: topMaxCount = 3,
    extendButtons: topExtendButtons = [],
    hideAutomatically: topHideAutomatically = true,
    hideByClick: topHideByClick = true,
    style: topStyle = _ZegoMenuBarStyle.default.light
  } = topMenuBarConfig;
  const {
    showMicrophoneState = true,
    showCameraState = true,
    itemBuilder
  } = memberListConfig;
  const {
    enableTiming = durationConfig.isVisible || true,
    isDurationVisible = true,
    onDurationUpdate = durationConfig.onDurationUpdate
  } = timingConfig;
  const stateData = (0, _react.useRef)(_prebuilt_helper.default.getInstance().getStateData());
  const [isFrontCamera, setIsFrontCamera] = (0, _react.useState)(stateData.current.isFrontCamera || useFrontFacingCamera);
  const [turnOnCameraWhenJoining, setTurnOnCameraWhenJoining] = (0, _react.useState)(stateData.current.turnOnCameraWhenJoining !== undefined ? stateData.current.turnOnCameraWhenJoining : config.turnOnCameraWhenJoining);
  const [turnOnMicrophoneWhenJoining, setTurnOnMicrophoneWhenJoining] = (0, _react.useState)(stateData.current.turnOnMicrophoneWhenJoining !== undefined ? stateData.current.turnOnMicrophoneWhenJoining : config.turnOnMicrophoneWhenJoining);
  const [useSpeakerWhenJoining, setUseSpeakerWhenJoining] = (0, _react.useState)(stateData.current.useSpeakerWhenJoining !== undefined ? stateData.current.useSpeakerWhenJoining : config.useSpeakerWhenJoining);
  const [isBottomMenubarVisable, setBottomMenubarVisable] = (0, _react.useState)(true);
  const [isTopMenubarVisable, setTopMenubarVisable] = (0, _react.useState)(true);
  const [isCallMemberListVisable, setIsCallMemberListVisable] = (0, _react.useState)(false);
  const [textInputVisable, setTextInputVisable] = (0, _react.useState)(false);
  const [isToastVisable, setIsToastVisable] = (0, _react.useState)(false);
  const [toastExtendedData, setToastExtendedData] = (0, _react.useState)({
    type: _zegoUikitRn.ZegoToastType.default,
    text: ''
  });
  const toastInvisableTimeoutRef = (0, _react.useRef)(null);
  const keyboardHeight = (0, _keyboard.useKeyboard)();
  var hideCountdown = 5;
  var hideCountdownOnTopMenu = 5;
  const debounce = (0, _react.useRef)(false);
  if (stateData.current.callbackID) {
    stateData.current.callbackID = 'ZegoUIKitPrebuiltCall' + String(Math.floor(Math.random() * 10000));
  }
  const callbackID = stateData.current.callbackID;
  const isPageInBackground = () => {
    const isMinimize = _minimizing_helper.default.getInstance().getIsMinimize();
    return isMinimize;
  };
  const onFullPageTouch = () => {
    (0, _logger.zloginfo)('####onFullPageTouch####');
    hideCountdown = 5;
    hideCountdownOnTopMenu = 5;
    if (isBottomMenubarVisable) {
      if (hideByClick) {
        setBottomMenubarVisable(false);
        setIsCallMemberListVisable(false);
      }
    } else {
      setBottomMenubarVisable(true);
    }
    if (isTopMenubarVisable) {
      if (topHideByClick) {
        setTopMenubarVisable(false);
        setIsCallMemberListVisable(false);
      }
    } else {
      setTopMenubarVisable(true);
    }
    setTextInputVisable(false);
  };
  const grantPermissions = async callback => {
    // Android: Dynamically obtaining device permissions
    if (_reactNative.Platform.OS === 'android') {
      // Check if permission granted
      let grantedAudio = _reactNative.PermissionsAndroid.check(_reactNative.PermissionsAndroid.PERMISSIONS.RECORD_AUDIO);
      let grantedCamera = _reactNative.PermissionsAndroid.check(_reactNative.PermissionsAndroid.PERMISSIONS.CAMERA);
      const ungrantedPermissions = [];
      try {
        const isAudioGranted = await grantedAudio;
        const isVideoGranted = await grantedCamera;
        if (!isAudioGranted) {
          ungrantedPermissions.push(_reactNative.PermissionsAndroid.PERMISSIONS.RECORD_AUDIO);
        }
        const isVideoCall = buttons.includes(_ZegoMenuBarButtonName.default.toggleCameraButton) || turnOnCameraWhenJoining;
        if (!isVideoGranted && isVideoCall) {
          ungrantedPermissions.push(_reactNative.PermissionsAndroid.PERMISSIONS.CAMERA);
        }
      } catch (error) {
        ungrantedPermissions.push(_reactNative.PermissionsAndroid.PERMISSIONS.RECORD_AUDIO, _reactNative.PermissionsAndroid.PERMISSIONS.CAMERA);
      }
      // If not, request it
      (0, _logger.zloginfo)(`[ZegoUIKitPrebuiltCall] requestMultiple, ungrantedPermissions: ${ungrantedPermissions}`);
      return _reactNative.PermissionsAndroid.requestMultiple(ungrantedPermissions).then(result => {
        (0, _logger.zloginfo)(`[ZegoUIKitPrebuiltCall] requestMultiple, result: ${result}`);
        if (callback) {
          callback();
        }
      });
    } else if (callback) {
      callback();
    }
  };
  // Default operation for click the leave button
  const showLeaveAlert = () => {
    return new Promise((resolve, reject) => {
      if (hangUpConfirmInfo) {
        const {
          title = "Leave the call",
          message = "Are you sure to leave the call?",
          cancelButtonName = "Cancel",
          confirmButtonName = "Confirm"
        } = hangUpConfirmInfo;
        _reactNative.Alert.alert(title, message, [{
          text: cancelButtonName,
          onPress: () => {
            reject();
          },
          style: "cancel"
        }, {
          text: confirmButtonName,
          onPress: () => {
            resolve();
          }
        }], {
          cancelable: false
        });
      } else {
        resolve();
      }
    });
  };
  const sortAudioVideo = globalAudioVideoUserList => {
    if (layout.mode === _zegoUikitRn.ZegoLayoutMode.pictureInPicture) {
      if (globalAudioVideoUserList.length > 1) {
        const index = globalAudioVideoUserList.findIndex(user => user.userID === userID);
        if (index !== -1) {
          const localUser = globalAudioVideoUserList.splice(index, 1)[0];
          globalAudioVideoUserList.splice(1, 0, localUser);
        }
      }
      return globalAudioVideoUserList;
    } else {
      // Do not deal with
      return globalAudioVideoUserList;
    }
  };
  const startCallTimingTimer = (0, _react.useCallback)(() => {
    if (!enableTiming) return;
    _timing_helper.default.getInstance().stopTimer();
    _timing_helper.default.getInstance().startTimer();
  }, []);
  const destroyCallTimingTimer = (0, _react.useCallback)(isFromMinimize => {
    (0, _logger.zloginfo)('[ZegoUIKitPrebuiltCall] destroyCallTimingTimer, isFromMinimize:', isFromMinimize);
    _timing_helper.default.getInstance().stopTimer();

    // we need to reset duration when call is ended from minimize
    if (isFromMinimize) {
      _timing_helper.default.getInstance().resetDuration();
    }
    // we need to reset duration when the call is normal ended.
    else if (!_minimizing_helper.default.getInstance().getIsMinimizeSwitch()) {
      _timing_helper.default.getInstance().resetDuration();
    }
  }, []);
  const onOpenCallMemberList = () => {
    setIsCallMemberListVisable(true);
  };
  const onCloseCallMemberList = () => {
    setIsCallMemberListVisable(false);
  };
  const onSwitchCamera = () => {
    (0, _logger.zloginfo)('onSwitchCamera', isFrontCamera);

    // Default rear camera
    const result = !isFrontCamera;
    stateData.current.isFrontCamera = result;
    setIsFrontCamera(result);
  };
  const showMessageListView = () => {
    return buttons.includes(_ZegoMenuBarButtonName.default.messageButton);
  };
  const resetToastInvisableTimeout = (0, _react.useCallback)((duration = 3000) => {
    clearTimeout(toastInvisableTimeoutRef.current);
    toastInvisableTimeoutRef.current = setTimeout(() => {
      setIsToastVisable(false);
    }, duration);
  }, []);
  const showToast = (0, _react.useCallback)(({
    type = _zegoUikitRn.ZegoToastType.default,
    text = '',
    duration = 3000
  } = {}) => {
    if (!text) return;
    setToastExtendedData({
      type,
      text
    });
    setIsToastVisable(true);
    resetToastInvisableTimeout(duration);
  }, [resetToastInvisableTimeout]);
  (0, _react.useImperativeHandle)(ref, () => ({
    hangUp: (showConfirmation = false) => {
      (0, _logger.zloginfo)(`[ZegoUIKitPrebuiltCall][hangUp] showConfirmation: ${showConfirmation}`);
      if (debounce.current) return;
      if (!showConfirmation) {
        debounce.current = true;
        const duration = _timing_helper.default.getInstance().getDuration();
        typeof onCallEnd == 'function' && onCallEnd(callID, _defines.ZegoCallEndReason.localHangUp, duration);
        debounce.current = false;
      } else {
        debounce.current = true;
        const temp = onHangUpConfirmation || showLeaveAlert;
        temp().then(() => {
          const duration = _timing_helper.default.getInstance().getDuration();
          typeof onCallEnd == 'function' && onCallEnd(callID, _defines.ZegoCallEndReason.localHangUp, duration);
          debounce.current = false;
        });
      }
    },
    minimizeWindow: () => {
      _minimizing_helper.default.getInstance().minimizeWindow();
    },
    showToast
  }));
  const handleBackButton = () => {
    return true;
  };
  const _getCacheAVUserListWhenMinimizeSwitched = () => {
    if (isMinimizeSwitch) {
      if (layout && layout.config && layout.config.removeViewWhenAudioVideoUnavailable == false) {
        let allUsers = _zegoUikitRn.default.getAllUsers().filter(user => user.userID);
        return sortAudioVideo(allUsers);
      } else {
        let hasStreamUsers = _zegoUikitRn.default.getAllUsers().filter(user => user.userID && (user.isCameraOn || user.isMicrophoneOn));
        return sortAudioVideo(hasStreamUsers);
      }
    } else {
      return null;
    }
  };
  (0, _react.useEffect)(() => {
    (0, _logger.zloginfo)(`[ZegoUIKitPrebuiltCall] useEffect`);
    _zegoUikitRn.default.onJoinRoom(callbackID, onJoinRoom);
    _zegoUikitRn.default.onOnlySelfInRoom(callbackID, () => {
      (0, _logger.zloginfo)('[ZegoUIKitPrebuiltCall][onOnlySelfInRoom]');
      const duration = _timing_helper.default.getInstance().getDuration();
      if (typeof onCallEnd == 'function') {
        onCallEnd(callID, _defines.ZegoCallEndReason.remoteHangUp, duration);
      }
    });
    _zegoUikitRn.default.onUserJoin(callbackID, onUserJoin);
    _zegoUikitRn.default.onMeRemovedFromRoom(callbackID, () => {
      (0, _logger.zloginfo)('[ZegoUIKitPrebuiltCall][onMeRemovedFromRoom]');
      const duration = _timing_helper.default.getInstance().getDuration();
      if (typeof onCallEnd == 'function') {
        onCallEnd(callID, _defines.ZegoCallEndReason.kickOut, duration);
      }
    });
    _zegoUikitRn.default.onRequireNewToken(callbackID, onRequireNewToken);
    _EventEmitter.eventEmitter.on(_EventEmitter.EventName.SHOW_TOAST, showToast);
    _zegoUikitRn.default.onMicrophoneOn(callbackID, (targetUserID, isOn) => {
      if (targetUserID === userID) {
        stateData.current.turnOnMicrophoneWhenJoining = !!isOn;
      }
    });
    _zegoUikitRn.default.onCameraOn(callbackID, (targetUserID, isOn) => {
      if (targetUserID === userID) {
        stateData.current.turnOnCameraWhenJoining = !!isOn;
      }
    });
    _zegoUikitRn.default.onAudioOutputDeviceChanged(callbackID, type => {
      stateData.current.useSpeakerWhenJoining = type === 0;
    });
    _prebuilt_helper.default.getInstance().onPrebuiltDestroy(callbackID, () => {
      // Calls are usually in the minimized mode.
      (0, _logger.zloginfo)('[ZegoUIKitPrebuiltCall] onPrebuiltDestroy', callbackID);
      (0, _logger.zloginfo)('[ZegoUIKitPrebuiltCall] leave room, from: onPrebuiltDestroy');
      _zegoUikitRn.default.leaveRoom();
      _zegoUikitRn.default.onJoinRoom(callbackID);
      _zegoUikitRn.default.onOnlySelfInRoom(callbackID);
      _zegoUikitRn.default.onUserJoin(callbackID);
      _zegoUikitRn.default.onMeRemovedFromRoom(callbackID);
      _zegoUikitRn.default.onRequireNewToken(callbackID);
      destroyCallTimingTimer(true);
      _prebuilt_helper.default.getInstance().clearState();
      _prebuilt_helper.default.getInstance().clearRouteParams();
      _prebuilt_helper.default.getInstance().clearNotify();
      _minimizing_helper.default.getInstance().setIsMinimizeSwitch(false, 'onPrebuiltDestroy');
      _minimizing_helper.default.getInstance().notifyEntryNormal();
      (0, _logger.zloginfo)(`[ZegoUIKitPrebuiltCall] KeepAwake deactivate`);
      (0, _reactNativeKeepAwake.deactivateKeepAwake)();
    });
    let curMinimizeSwitch = _minimizing_helper.default.getInstance().getIsMinimizeSwitch();
    _zegoUikitRn.default.init(appID, appSign, {
      userID: userID,
      userName: userName
    }).then(() => {
      _minimizing_helper.default.getInstance().notifyPrebuiltInit();
      _zegoUikitRn.default.setAudioOutputToSpeaker(useSpeakerWhenJoining);
      (0, _logger.zloginfo)(`[ZegoUIKitPrebuiltCall] joinRoom: ${callID}`);
      if (appSign) {
        _zegoUikitRn.default.joinRoom(callID);
      } else {
        _zegoUikitRn.default.joinRoom(callID, token || (typeof onRequireNewToken === 'function' ? onRequireNewToken() || '' : ''));
      }

      // Skip timer reset if just switched from minimized mode.
      if (!curMinimizeSwitch) {
        startCallTimingTimer();
      }
      grantPermissions(() => {
        // RTC need it
        _zegoUikitRn.default.turnCameraOn('', false);
        _zegoUikitRn.default.turnCameraOn('', turnOnCameraWhenJoining);
        _zegoUikitRn.default.useFrontFacingCamera(isFrontCamera);
        _zegoUikitRn.default.turnMicrophoneOn('', turnOnMicrophoneWhenJoining);
      });
    });
    const backHandler = _reactNative.BackHandler.addEventListener('hardwareBackPress', handleBackButton);
    navigation.setOptions({
      gestureEnabled: false
    });
    (0, _logger.zloginfo)(`[ZegoUIKitPrebuiltCall] KeepAwake activate`);
    (0, _reactNativeKeepAwake.activateKeepAwake)();
    _minimizing_helper.default.getInstance().setIsMinimizeSwitch(false, 'PrebuiltCall useEffect');
    _timing_helper.default.getInstance().onDurationUpdate(callbackID, duration => {
      if (typeof onDurationUpdate === 'function') {
        (0, _logger.zloginfo)(`[ZegoUIKitPrebuiltCall] onDurationUpdate execute will, don't log succeed`);
        onDurationUpdate(duration);
      }
    });
    return () => {
      (0, _logger.zloginfo)('[ZegoUIKitPrebuiltCall] useEffect return');
      const isMinimizeSwitch = _minimizing_helper.default.getInstance().getIsMinimizeSwitch();
      if (!isMinimizeSwitch) {
        (0, _logger.zloginfo)('[ZegoUIKitPrebuiltCall] leave room, from: not isMinimizeSwitch');
        _zegoUikitRn.default.leaveRoom();
        _zegoUikitRn.default.onJoinRoom(callbackID);
        _zegoUikitRn.default.onOnlySelfInRoom(callbackID);
        _zegoUikitRn.default.onUserJoin(callbackID);
        _zegoUikitRn.default.onMeRemovedFromRoom(callbackID);
        _zegoUikitRn.default.onRequireNewToken(callbackID);
        _prebuilt_helper.default.getInstance().clearState();
        _prebuilt_helper.default.getInstance().clearRouteParams();
        _prebuilt_helper.default.getInstance().clearNotify();
        destroyCallTimingTimer(false);
        (0, _logger.zloginfo)('[ZegoUIKitPrebuiltCall] KeepAwake deactivate');
        (0, _reactNativeKeepAwake.deactivateKeepAwake)();
      }
      _timing_helper.default.getInstance().onDurationUpdate(callbackID);
      _EventEmitter.eventEmitter.off(_EventEmitter.EventName.SHOW_TOAST, showToast);
      clearTimeout(toastInvisableTimeoutRef.current);
      backHandler.remove();
      navigation.setOptions({
        gestureEnabled: true
      });
    };
  }, []);
  function useInterval(callback, delay) {
    const savedCallback = (0, _react.useRef)();
    (0, _react.useEffect)(() => {
      savedCallback.current = callback;
    });
    (0, _react.useEffect)(() => {
      function tick() {
        savedCallback.current();
      }
      if (delay !== null) {
        let id = setInterval(tick, delay);
        return () => clearInterval(id);
      }
    }, [delay]);
  }
  useInterval(() => {
    hideCountdown--;
    if (hideCountdown <= 0) {
      hideCountdown = 5;
      if (hideAutomatically) {
        setBottomMenubarVisable(false);
      }
    }
  }, 1000);
  useInterval(() => {
    hideCountdownOnTopMenu--;
    if (hideCountdownOnTopMenu <= 0) {
      hideCountdownOnTopMenu = 5;
      if (topHideAutomatically) {
        setTopMenubarVisable(false);
      }
    }
  }, 1000);
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.container, styles.fillParent]
  }, /*#__PURE__*/_react.default.createElement(_zegoUikitRn.ZegoToast, {
    visable: isToastVisable,
    type: toastExtendedData.type,
    text: toastExtendedData.text
  }), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.foregroundContainer,
    pointerEvents: "none"
  }, /*#__PURE__*/_react.default.createElement(_ZegoPrebuiltForegroundView.default, {
    style: styles.foregroundView,
    isDurationVisible: isDurationVisible
  })), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.foregroundBuilderContainer,
    pointerEvents: "box-none"
  }, /*#__PURE__*/_react.default.createElement(_reactDelegateComponent.default
  // style={styles.foregroundBuilderDelegate}
  , {
    to: prebuiltForegroundBuilder
  })), isVisible && isTopMenubarVisable ? /*#__PURE__*/_react.default.createElement(_ZegoTopMenuBar.default, {
    menuTitle: topTitle,
    menuBarButtonsMaxCount: topMaxCount,
    menuBarButtons: topButtons,
    menuBarExtendedButtons: topExtendButtons,
    onHangUp: () => {
      const duration = _timing_helper.default.getInstance().getDuration();
      typeof onCallEnd == 'function' && onCallEnd(callID, _defines.ZegoCallEndReason.localHangUp, duration);
    },
    onHangUpConfirmation: onHangUpConfirmation ? onHangUpConfirmation : showLeaveAlert,
    turnOnCameraWhenJoining: turnOnCameraWhenJoining,
    turnOnMicrophoneWhenJoining: turnOnMicrophoneWhenJoining,
    useSpeakerWhenJoining: useSpeakerWhenJoining,
    onOpenCallMemberList: onOpenCallMemberList,
    onSwitchCamera: onSwitchCamera,
    onMessagePress: () => {
      setTextInputVisable(true);
      setBottomMenubarVisable(false);
      setTopMenubarVisable(false);
    }
  }) : /*#__PURE__*/_react.default.createElement(_reactNative.View, null), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.fillParent, {
      bottom: insets.bottom,
      right: insets.right
    }],
    pointerEvents: "auto",
    onTouchStart: onFullPageTouch
  }, /*#__PURE__*/_react.default.createElement(_zegoUikitRn.ZegoAudioVideoContainer, {
    style: [styles.audioVideoView, styles.fillParent],
    audioVideoConfig: {
      showSoundWavesInAudioMode: showSoundWavesInAudioMode,
      useVideoViewAspectFill: useVideoViewAspectFill,
      cacheAudioVideoUserList: _getCacheAVUserListWhenMinimizeSwitched()
    },
    layout: layout,
    avatarBuilder: avatarBuilder,
    foregroundBuilder: foregroundBuilder ? foregroundBuilder : ({
      userInfo
    }) => /*#__PURE__*/_react.default.createElement(_AudioVideoForegroundView.default, {
      userInfo: userInfo,
      showMicrophoneStateOnView: showMicrophoneStateOnView,
      showCameraStateOnView: showCameraStateOnView,
      showUserNameOnView: showUserNameOnView
    }),
    sortAudioVideo: sortAudioVideo
  }), showMessageListView() ? /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.messageListView
  }, /*#__PURE__*/_react.default.createElement(_zegoUikitRn.ZegoInRoomMessageView, {
    itemBuilder: messageItemBuilder,
    style: styles.fillParent
  })) : null), isBottomMenubarVisable ? /*#__PURE__*/_react.default.createElement(_ZegoBottomBar.default, {
    menuBarButtonsMaxCount: maxCount,
    menuBarButtons: buttons,
    menuBarExtendedButtons: extendButtons,
    onHangUp: () => {
      const duration = _timing_helper.default.getInstance().getDuration();
      typeof onCallEnd == 'function' && onCallEnd(callID, _defines.ZegoCallEndReason.localHangUp, duration);
    },
    onHangUpConfirmation: onHangUpConfirmation ? onHangUpConfirmation : showLeaveAlert,
    turnOnCameraWhenJoining: turnOnCameraWhenJoining,
    turnOnMicrophoneWhenJoining: turnOnMicrophoneWhenJoining,
    useSpeakerWhenJoining: useSpeakerWhenJoining,
    onMorePress: () => {
      hideCountdown = 5;
    },
    onSwitchCamera: onSwitchCamera,
    onMessagePress: () => {
      setTextInputVisable(true);
      setBottomMenubarVisable(false);
      setTopMenubarVisable(false);
    }
  }) : /*#__PURE__*/_react.default.createElement(_reactNative.View, null), isCallMemberListVisable ? /*#__PURE__*/_react.default.createElement(_ZegoCallMemberList.default, {
    showMicrophoneState: showMicrophoneState,
    showCameraState: showCameraState,
    itemBuilder: itemBuilder,
    avatarBuilder: avatarBuilder,
    onCloseCallMemberList: onCloseCallMemberList
  }) : /*#__PURE__*/_react.default.createElement(_reactNative.View, null), textInputVisable ? /*#__PURE__*/_react.default.createElement(_reactNative.KeyboardAvoidingView, {
    style: styles.fillParent,
    behavior: _reactNative.Platform.OS == "ios" ? "padding" : "height",
    keyboardVerticalOffset: _reactNative.Platform.OS === 'android' && _reactNative.Platform.Version < 35 ? 32 : 0
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.messageInputPannel, {
      bottom: _reactNative.Platform.OS == 'ios' ? keyboardHeight : 0,
      height: 45
    }]
  }, /*#__PURE__*/_react.default.createElement(_zegoUikitRn.ZegoInRoomMessageInput, {
    ref: input => {
      // setTextInput(input);
    }
    // @ts-ignore
    ,
    onContentSizeChange: height => {
      // setTextInputHeight(height);
    },
    placeholder: 'Say something...',
    onSumit: () => {
      setTextInputVisable(false);
    }
  }))) : null);
}
var _default = exports.default = /*#__PURE__*/(0, _react.forwardRef)(ZegoUIKitPrebuiltCall);
const styles = _reactNative.StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    zIndex: 0
  },
  fillParent: {
    width: '100%',
    height: '100%',
    position: 'absolute'
  },
  audioVideoView: {
    flex: 1,
    zIndex: 2,
    right: 0,
    top: 0
  },
  foregroundContainer: {
    flex: 1,
    position: 'absolute',
    width: '100%',
    height: '100%',
    top: 6,
    zIndex: 11,
    alignItems: 'center'
  },
  foregroundView: {
    flex: 1,
    position: 'absolute',
    width: '100%',
    height: '100%',
    alignItems: 'center'
  },
  foregroundBuilderDelegate: {
    flex: 1,
    position: 'absolute',
    width: '100%',
    height: '100%',
    alignItems: 'center'
  },
  foregroundBuilderContainer: {
    flex: 1,
    position: 'absolute',
    width: '100%',
    height: '100%',
    zIndex: 21,
    alignItems: 'center'
  },
  messageInputPannel: {
    position: 'absolute',
    borderTopLeftRadius: 12,
    borderTopRightRadius: 12,
    backgroundColor: 'rgba(0, 0, 0, 0.7500)',
    width: '100%'
  },
  messageListView: {
    position: 'absolute',
    left: 16,
    bottom: 110,
    width: 230,
    maxHeight: 200,
    zIndex: 10
  }
});
