// @zegocloud/zego-uikit-prebuilt-call-rn/lib/commonjs/call/ZegoPrebuiltForegroundView.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ZegoPrebuiltForegroundView;
var _react = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
var _utils = require("../utils");
var _timing_helper = _interopRequireDefault(require("../services/timing_helper"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
function ZegoPrebuiltForegroundView(props) {
  const {
    isDurationVisible
  } = props;
  const [duration, setDuration] = (0, _react.useState)(_timing_helper.default.getInstance().getDuration());
  (0, _react.useEffect)(() => {
    _timing_helper.default.getInstance().onDurationUpdate("ZegoPrebuiltForegroundViewDuration", duration => setDuration(duration));
    return () => {
      _timing_helper.default.getInstance().onDurationUpdate("ZegoPrebuiltForegroundViewDuration");
    };
  }, []);
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.container
  }, isDurationVisible ? /*#__PURE__*/_react.default.createElement(_reactNative.SafeAreaView, null, /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: styles.timing
  }, (0, _utils.durationFormat)(duration))) : null);
}
const paddingTop = _reactNative.Platform.OS === "android" ? _reactNative.StatusBar.currentHeight : 26;
const styles = _reactNative.StyleSheet.create({
  container: {
    flex: 1,
    position: 'absolute',
    width: '100%',
    height: '100%',
    top: paddingTop,
    zIndex: 5,
    alignItems: 'center'
  },
  timing: {
    color: 'white'
  }
});
