// @zegocloud/zego-uikit-prebuilt-call-rn/lib/commonjs/call/ZegoCountdownLabel.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ZegoCountdownLabel;
var _react = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
var _utils = require("../utils");
var _timing_helper = _interopRequireDefault(require("../services/timing_helper"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
function ZegoCountdownLabel(props) {
  const {
    maxDuration,
    onCountdownFinished
  } = props;
  const [duration, setDuration] = (0, _react.useState)(Math.max(maxDuration - _timing_helper.default.getInstance().getDuration(), 0));
  const [countdownFinished, setCountdownFinished] = (0, _react.useState)(false);
  (0, _react.useEffect)(() => {
    _timing_helper.default.getInstance().onDurationUpdate("ZegoCountdownLabelDuration", v => {
      if (v == maxDuration) {
        if (typeof onCountdownFinished === "function") {
          onCountdownFinished();
        }
      } else if (v > maxDuration) return;
      setDuration(Math.max(maxDuration - v, 0));
    });
    return () => {
      _timing_helper.default.getInstance().onDurationUpdate("ZegoCountdownLabelDuration");
    };
  }, []);
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.container,
    pointerEvents: 'box-none'
  }, /*#__PURE__*/_react.default.createElement(_reactNative.SafeAreaView, null, /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: styles.timing
  }, (0, _utils.durationFormat)(duration))));
}
const styles = _reactNative.StyleSheet.create({
  container: {
    flex: 1,
    position: 'absolute',
    width: '100%',
    height: '100%',
    top: 26,
    zIndex: 5,
    alignItems: 'center'
  },
  timing: {
    color: 'white'
  }
});
