// @zegocloud/zego-uikit-prebuilt-call-rn/lib/commonjs/call/ZegoTopMenuBar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ZegoTopBar;
var _react = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
var _ZegoMenuBarButtonName = _interopRequireDefault(require("./ZegoMenuBarButtonName"));
var _ZegoMemberButton = _interopRequireDefault(require("./ZegoMemberButton"));
var _ZegoMinimizingButton = _interopRequireDefault(require("./ZegoMinimizingButton"));
var _zegoUikitRn = require("@zegocloud/zego-uikit-rn");
var _ZegoMessageButton = _interopRequireDefault(require("./ZegoMessageButton"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
function ZegoTopBar(props) {
  const {
    menuTitle,
    menuBarButtonsMaxCount,
    menuBarButtons,
    menuBarExtendedButtons,
    onHangUp,
    onHangUpConfirmation,
    onSwitchCamera,
    onOpenCallMemberList,
    turnOnCameraWhenJoining,
    turnOnMicrophoneWhenJoining,
    useSpeakerWhenJoining,
    onMessagePress
  } = props;
  const getButtonByButtonIndex = buttonIndex => {
    switch (buttonIndex) {
      case _ZegoMenuBarButtonName.default.toggleCameraButton:
        return /*#__PURE__*/_react.default.createElement(_zegoUikitRn.ZegoToggleCameraButton, {
          key: 1,
          width: 30,
          height: 30,
          isOn: turnOnCameraWhenJoining
        });
      case _ZegoMenuBarButtonName.default.toggleMicrophoneButton:
        return /*#__PURE__*/_react.default.createElement(_zegoUikitRn.ZegoToggleMicrophoneButton, {
          key: 2,
          width: 30,
          height: 30,
          isOn: turnOnMicrophoneWhenJoining
        });
      case _ZegoMenuBarButtonName.default.hangUpButton:
        return /*#__PURE__*/_react.default.createElement(_zegoUikitRn.ZegoLeaveButton, {
          key: 0,
          width: 30,
          height: 30,
          onLeaveConfirmation: onHangUpConfirmation,
          onPressed: onHangUp
        });
      case _ZegoMenuBarButtonName.default.switchAudioOutputButton:
        return /*#__PURE__*/_react.default.createElement(_zegoUikitRn.ZegoSwitchAudioOutputButton, {
          key: 4,
          width: 30,
          height: 30,
          useSpeaker: useSpeakerWhenJoining
        });
      case _ZegoMenuBarButtonName.default.switchCameraButton:
        return /*#__PURE__*/_react.default.createElement(_zegoUikitRn.ZegoSwitchCameraButton, {
          key: 3,
          width: 30,
          height: 30,
          onPress: onSwitchCamera
        });
      case _ZegoMenuBarButtonName.default.showMemberListButton:
        return /*#__PURE__*/_react.default.createElement(_ZegoMemberButton.default, {
          key: 5,
          onPressed: onOpenCallMemberList
        });
      case _ZegoMenuBarButtonName.default.minimizingButton:
        return /*#__PURE__*/_react.default.createElement(_ZegoMinimizingButton.default, {
          key: 6
        });
      case _ZegoMenuBarButtonName.default.messageButton:
        return /*#__PURE__*/_react.default.createElement(_ZegoMessageButton.default, {
          key: 7,
          width: 30,
          height: 30,
          onPress: onMessagePress
        });
    }
  };
  const getDisplayButtons = () => {
    let allButtons = [];
    menuBarButtons.slice(0, menuBarButtonsMaxCount).forEach(buttonIndex => {
      allButtons.push(getButtonByButtonIndex(buttonIndex));
    });
    allButtons = allButtons.concat(menuBarExtendedButtons);
    return allButtons;
  };
  const getButtonStyle = () => {
    return styles.customIconContainer;
  };
  const allButtons = getDisplayButtons();
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.topBarContainer
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.left
  }, /*#__PURE__*/_react.default.createElement(_reactNative.TouchableWithoutFeedback, null, /*#__PURE__*/_react.default.createElement(_reactNative.Image, {
    style: styles.downArrowIcon,
    source: require('./resources/white_button_arrow.png')
  })), /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: styles.title
  }, menuTitle)), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.right
  }, allButtons.map((button, index) => /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    key: index,
    style: getButtonStyle()
  }, button))));
}
const styles = _reactNative.StyleSheet.create({
  topBarContainer: {
    flex: 1,
    position: 'absolute',
    flexDirection: 'row',
    alignItems: 'center',
    width: '100%',
    top: 35,
    height: 44,
    zIndex: 3,
    justifyContent: 'space-between',
    paddingLeft: 3.5,
    paddingRight: 13.5
  },
  left: {
    opacity: 0,
    flexDirection: 'row',
    alignItems: "center"
  },
  right: {
    flexDirection: 'row',
    alignItems: "center"
  },
  title: {
    color: '#FFFFFF',
    fontSize: 18,
    marginLeft: 4
  },
  customIconContainer: {
    marginLeft: 10
  }
});
