// @zegocloud/zego-uikit-prebuilt-call-rn/lib/commonjs/call_invitation/components/ZegoSendCallInvitationButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ZegoSendCallInvitationButton;
var _react = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
var _zegoUikitRn = require("@zegocloud/zego-uikit-rn");
var _defines = require("../services/defines");
var _plugins = _interopRequireDefault(require("../services/plugins"));
var _native = require("@react-navigation/native");
var _inner_text_helper = _interopRequireDefault(require("../services/inner_text_helper"));
var _invitation = _interopRequireDefault(require("../../services/invitation"));
var _logger = require("../../utils/logger");
var _EventEmitter = require("../../utils/EventEmitter");
var _report = _interopRequireDefault(require("../../utils/report"));
var _invite_state_manager = _interopRequireDefault(require("../services/invite_state_manager"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
function ZegoSendCallInvitationButton(props) {
  const navigation = (0, _native.useNavigation)();
  const {
    icon,
    text,
    invitees = [],
    isVideoCall = false,
    timeout = 60,
    onWillPressed,
    onPressed,
    width = 42,
    height = 42,
    verticalLayout = false,
    textColor,
    fontSize,
    backgroundColor,
    borderRadius,
    borderWidth,
    borderColor,
    borderStyle,
    callID,
    callName,
    showWaitingPageWhenGroupCall = false,
    resourceID: _resourceID = '',
    customData
  } = props;
  const getInviteeIDList = () => {
    return invitees.map(invitee => {
      return invitee.userID;
    });
  };
  const [localUser, setLocalUser] = (0, _react.useState)({
    ..._plugins.default.getLocalUser()
  });
  let _roomID = '';
  const isCanSendInvitation = () => {
    let canSendInvitation = false;
    let isOnCall = _invite_state_manager.default.isOncall();
    if (isOnCall) {
      canSendInvitation = false;
    } else if (callID && typeof callID !== 'string') {
      (0, _logger.zlogwarning)('[ZegoSendCallInvitationButton][isCanSendInvitation] callID should be string');
      canSendInvitation = false;
    } else if (!onWillPressed) {
      canSendInvitation = true;
    } else if (typeof onWillPressed === 'function') {
      canSendInvitation = onWillPressed();
    } else {
      // don't support promise
      canSendInvitation = false;
    }
    (0, _logger.zloginfo)('isCanSendInvitation', canSendInvitation);
    return canSendInvitation;
  };
  const onPress = ({
    errorCode,
    errorMessage,
    errorInvitees,
    invitationID,
    invitees: successfulInvitees
  }) => {
    _report.default.reportEvent('call/invite', {
      'call_id': invitationID,
      'room_id': callID,
      'source': 'button'
    });
    _invitation.default.getInstance().onInvitationSent(navigation, invitationID, invitees, successfulInvitees, _roomID, isVideoCall, callName, showWaitingPageWhenGroupCall);
    if (typeof onPressed === 'function') {
      onPressed(errorCode, errorMessage, errorInvitees);
    }
  };
  const onFailure = ({
    code: errorCode,
    message: errorMessage
  }) => {
    if (typeof onPressed === 'function') {
      onPressed(errorCode, errorMessage, undefined);
    }
  };
  const onRequestStatusChanged = isInviteCalling => {
    _invite_state_manager.default.setOutgoingCallStatus(isInviteCalling);
  };
  const loginHandler = ({
    userID,
    userName
  }) => {
    if (localUser.userID === userID) {
      (0, _logger.zloginfo)(`[ZegoSendCallInvitationButton][loginHandler] userID not changed, userID: ${userID}`);
      return;
    }
    (0, _logger.zloginfo)(`[ZegoSendCallInvitationButton][loginHandler] userID: ${userID}, userName: ${userName}`);
    setLocalUser({
      userID,
      userName
    });
  };
  (0, _react.useEffect)(() => {
    (0, _logger.zloginfo)(`[ZegoSendCallInvitationButton] useEffect, callName: ${callName}`);
    const unsubscribe1 = navigation.addListener('blur', () => {
      (0, _logger.zloginfo)(`[Navigation] ZegoSendCallInvitationButton blur, isVideoCall: ${isVideoCall}`);
    });
    const unsubscribe2 = navigation.addListener('focus', () => {
      (0, _logger.zloginfo)(`[Navigation] ZegoSendCallInvitationButton focus, isVideoCall: ${isVideoCall}`);
    });
    const unsubscribe3 = navigation.addListener('beforeRemove', () => {
      (0, _logger.zloginfo)(`[Navigation] ZegoSendCallInvitationButton beforeRemove, isVideoCall: ${isVideoCall}`);
    });
    _EventEmitter.eventEmitter.on(_EventEmitter.EventName.LOGINED, loginHandler);
    return () => {
      (0, _logger.zloginfo)(`[ZegoSendCallInvitationButton] useEffect return`);
      unsubscribe1();
      unsubscribe2();
      unsubscribe3();
      _EventEmitter.eventEmitter.off(_EventEmitter.EventName.LOGINED, loginHandler);
    };
  }, []);
  const updateData = () => {
    _roomID = callID ?? `call_${localUser.userID}_${Date.now()}`;
    (0, _logger.zloginfo)(`ZegoSendCallInvitationButton _roomID: ${_roomID}`);
    const _data = JSON.stringify({
      call_id: _roomID,
      call_name: callName,
      invitees: invitees.map(invitee => {
        return {
          user_id: invitee.userID,
          user_name: invitee.userName
        };
      }),
      type: isVideoCall ? _defines.ZegoInvitationType.videoCall : _defines.ZegoInvitationType.voiceCall,
      inviter: {
        id: localUser.userID,
        name: localUser.userName
      },
      timeout: timeout,
      custom_data: customData
    });
    return _data;
  };
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.container
  }, /*#__PURE__*/_react.default.createElement(_zegoUikitRn.ZegoSendInvitationButton, {
    icon: icon,
    text: text,
    invitees: getInviteeIDList(),
    type: isVideoCall ? _defines.ZegoInvitationType.videoCall : _defines.ZegoInvitationType.voiceCall,
    onRequestData: () => {
      return updateData();
    },
    timeout: timeout,
    onWillPressed: () => {
      return isCanSendInvitation();
    },
    onPressed: onPress,
    onFailure: onFailure,
    onRequestStatusChanged: onRequestStatusChanged,
    resourceID: _resourceID,
    width: width,
    height: height,
    color: textColor,
    fontSize: fontSize,
    backgroundColor: backgroundColor,
    verticalLayout: verticalLayout,
    borderRadius: borderRadius,
    borderWidth: borderWidth,
    borderColor: borderColor,
    borderStyle: borderStyle,
    notificationTitle: callName ? callName : _inner_text_helper.default.instance().getIncomingCallDialogTitle(localUser.userName, isVideoCall ? _defines.ZegoInvitationType.videoCall : _defines.ZegoInvitationType.voiceCall, invitees.length),
    notificationMessage: _inner_text_helper.default.instance().getIncomingCallDialogMessage(isVideoCall ? _defines.ZegoInvitationType.videoCall : _defines.ZegoInvitationType.voiceCall, invitees.length)
  }));
}
const styles = _reactNative.StyleSheet.create({
  container: {}
});
