// @zegocloud/zego-uikit-prebuilt-call-rn/lib/commonjs/call_invitation/components/ZegoCallInvitationDialog.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ZegoCallInvitationDialog;
var _react = _interopRequireWildcard(require("react"));
var _reactNative = require("react-native");
var _native = require("@react-navigation/native");
var _reactDelegateComponent = _interopRequireDefault(require("react-delegate-component"));
var _zegoUikitRn = _interopRequireWildcard(require("@zegocloud/zego-uikit-rn"));
var _services = _interopRequireDefault(require("../../services"));
var _logger = require("../../utils/logger");
var _report = _interopRequireDefault(require("../../utils/report"));
var _bell = _interopRequireDefault(require("../services/bell"));
var _callevent_notify_app = _interopRequireDefault(require("../services/callevent_notify_app"));
var _call_invite_helper = _interopRequireDefault(require("../services/call_invite_helper"));
var _defines = require("../services/defines");
var _inner_text_helper = _interopRequireDefault(require("../services/inner_text_helper"));
var _invite_state_manager = _interopRequireDefault(require("../services/invite_state_manager"));
var _notification_helper = _interopRequireDefault(require("../services/notification_helper"));
var _plugins = _interopRequireDefault(require("../services/plugins"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
function ZegoCallInvitationDialog(props) {
  const TAG = 'ZegoCallInvitationDialog';
  const initConfig = _services.default.getInstance().getInitConfig();
  const {
    showDeclineButton = true,
    avatarBuilder
  } = initConfig;
  const navigation = (0, _native.useNavigation)();
  const [isInit, setIsInit] = (0, _react.useState)(false);
  const [isDialogVisable, setIsDialogVisable] = (0, _react.useState)(false);
  const isDialogVisableRef = (0, _react.useRef)(false);
  const [isFullScreen, setIsFullScreen] = (0, _react.useState)(false);
  const [inviteType, setInviteType] = (0, _react.useState)(_defines.ZegoInvitationType.voiceCall);
  const [inviter, setInviter] = (0, _react.useState)({});
  const [extendData, setExtendData] = (0, _react.useState)({});
  const [callID, setCallID] = (0, _react.useState)('');
  const zimCallIDRef = (0, _react.useRef)('');
  const getDialogTitle = () => {
    if (extendData.notificationTitle && extendData.notificationTitle.length > 0) {
      return extendData.notificationTitle;
    } else if (extendData.call_name && extendData.call_name.length > 0) {
      return extendData.call_name;
    } else {
      const count = extendData.invitees ? extendData.invitees.length : 0;
      return _inner_text_helper.default.instance().getIncomingCallDialogTitle(inviter.name, inviteType, count);
    }
  };
  const getDialogMessage = () => {
    if (extendData.notificationMessage && extendData.notificationMessage.length > 0) {
      return extendData.notificationMessage;
    } else {
      const count = extendData.invitees ? extendData.invitees.length : 0;
      return _inner_text_helper.default.instance().getIncomingCallDialogMessage(inviteType, count);
    }
  };
  const getShortName = name => {
    if (!name) {
      return '';
    }
    const nl = name.split(' ');
    var shotName = '';
    nl.forEach(part => {
      if (part !== '') {
        shotName += part.substring(0, 1);
      }
    });
    return shotName;
  };
  const getImageSourceByPath = () => {
    if (inviteType === _defines.ZegoInvitationType.videoCall) {
      return require('../resources/button_call_video_accept.png');
    }
  };
  const onRefuseCallback = () => {
    (0, _logger.zloginfo)("onRefuseCallback");
    _callevent_notify_app.default.getInstance().notifyEvent('onIncomingCallDeclineButtonPressed', navigation);
    dismissForegroundNotification();
  };
  const refuseHandle = () => {
    _call_invite_helper.default.getInstance().refuseCall(callID);
  };
  const refuseSuccHandle = () => {
    (0, _logger.zloginfo)(`Refuse Call Invitation succ`);
    refuseHandle();
    _report.default.reportEvent('call/respondInvitation', {
      'call_id': callID,
      'app_state': _reactNative.AppState.currentState,
      'action': 'refuseSucc'
    });
  };
  const refuseFailHandle = error => {
    (0, _logger.zloginfo)(`Refuse Call Invitation failed, code: ${error.code}, message: ${error.message}`);
    refuseHandle();
    _report.default.reportEvent('call/respondInvitation', {
      'call_id': callID,
      'app_state': _reactNative.AppState.currentState,
      'action': 'refuseFail'
    });
  };
  const onAcceptCallback = data => {
    (0, _logger.zloginfo)("onAcceptCallback", data.call_id, data.inviter.id);
    _callevent_notify_app.default.getInstance().notifyEvent('onIncomingCallAcceptButtonPressed', navigation);
    dismissForegroundNotification();
    (0, _logger.zloginfo)('Navigate to ZegoUIKitPrebuiltCallInCallScreen');
    navigation.navigate('ZegoUIKitPrebuiltCallInCallScreen', {
      roomID: data.call_id,
      isVideoCall: data.type === _defines.ZegoInvitationType.videoCall,
      invitees: data.invitees,
      inviter: data.inviter.id
    });
  };
  const acceptHandle = () => {
    _call_invite_helper.default.getInstance().acceptCall(callID, {
      ...extendData,
      inviteType,
      inviter
    });
    _report.default.reportEvent('call/respondInvitation', {
      'call_id': callID,
      'app_state': _reactNative.AppState.currentState,
      'action': 'acceptSucc'
    });
  };
  const acceptFailHandle = error => {
    (0, _logger.zloginfo)(`Accept Call Invitation failed, code: ${error.code}, message: ${error.message}`);
    refuseHandle();
    _report.default.reportEvent('call/respondInvitation', {
      'call_id': callID,
      'app_state': _reactNative.AppState.currentState,
      'action': 'acceptFail'
    });
  };
  const pressHandle = () => {
    setIsFullScreen(true);
  };
  const dismissForegroundNotification = () => {
    (0, _logger.zloginfo)('[ZegoCallInvitationDialog] dismissForegroundNotification');
    _bell.default.stopIncomingSound();
    _bell.default.cancleVirate();
    setIsDialogVisable(false);
    setIsFullScreen(false);
  };
  (0, _react.useEffect)(() => {
    const callbackID = 'ZegoCallInvitationDialog' + String(Math.floor(Math.random() * 10000));
    _services.default.getInstance().onInit(callbackID, () => {
      setIsInit(true);
    });
    _call_invite_helper.default.getInstance().onCallAccepted(callbackID, onAcceptCallback);
    _call_invite_helper.default.getInstance().onCallRefused(callbackID, onRefuseCallback);
    return () => {
      _services.default.getInstance().onInit(callbackID);
      _call_invite_helper.default.getInstance().onCallAccepted(callbackID);
      _call_invite_helper.default.getInstance().onCallRefused(callbackID);
    };
  }, []);
  (0, _react.useEffect)(() => {
    if (isInit) {
      (0, _logger.zloginfo)('[ZegoCallInvitationDialog] Register callbacks after init');
      const callbackID = 'ZegoCallInvitationDialog' + String(Math.floor(Math.random() * 10000));
      _zegoUikitRn.default.getSignalingPlugin().onInvitationReceived(callbackID, async ({
        callID: invitationID,
        type,
        inviter,
        data
      }) => {
        (0, _logger.zloginfo)(`onInvitationReceived implement by ${TAG}, callID: ${invitationID}, isVideoCall: ${type}, inviter: ${JSON.stringify(inviter)}, data: ${data}`);
        let onCall = _invite_state_manager.default.isOncall(invitationID);
        const onRoom = _zegoUikitRn.default.isRoomConnected();
        const offlineData = _call_invite_helper.default.getInstance().getOfflineData();
        (0, _logger.zloginfo)(`[ZegoCallInvitationDialog][onInvitationReceived] onCall: ${onCall}, onRoom: ${onRoom}, offlineData: ${offlineData}, currentState: ${_reactNative.AppState.currentState}`);
        if (onCall || !offlineData && onRoom) {
          (0, _logger.zloginfo)(`Automatically declining invitations, onCall: ${onCall}, onRoom: ${onRoom}`);

          // Automatically declining invitations
          const dataParsed = data ? JSON.parse(data) : {};
          _zegoUikitRn.default.getSignalingPlugin().refuseInvitation(inviter.id, JSON.stringify({
            callID: invitationID,
            reason: 'busy',
            call_id: dataParsed.call_id,
            'invitee': {
              userID: _plugins.default.getLocalUser().userID,
              userName: _plugins.default.getLocalUser().userName
            }
          }));
          _invite_state_manager.default.updateInviteDataAfterRejected(invitationID);
          _callevent_notify_app.default.getInstance().notifyEvent('onIncomingCallRejectedCauseBusy', dataParsed.call_id, {
            userID: dataParsed.inviter.id,
            userName: dataParsed.inviter.name
          });
        } else {
          const currentData = JSON.parse(data);
          _callevent_notify_app.default.getInstance().notifyEvent('onIncomingCallReceived', currentData.call_id, {
            userID: currentData.inviter.id,
            userName: currentData.inviter.name
          }, currentData.type, currentData.invitees.map(invitee => {
            return {
              userID: invitee.user_id,
              userName: invitee.user_name
            };
          }), currentData.custom_data);
          if (offlineData && offlineData.call_id === currentData.call_id && offlineData.inviter.id === currentData.inviter.id) {
            (0, _logger.zloginfo)('[ZegoCallInvitationDialog][onInvitationReceived] received offline call');
            _call_invite_helper.default.getInstance().acceptCall(invitationID, offlineData);
            _zegoUikitRn.default.getSignalingPlugin().acceptInvitation(inviter.id, JSON.stringify({
              call_id: currentData.call_id,
              'invitee': {
                userID: _plugins.default.getLocalUser().userID,
                userName: _plugins.default.getLocalUser().userName
              }
            }));
            _call_invite_helper.default.getInstance().setOfflineData(undefined);
          } else {
            (0, _logger.zloginfo)(`[ZegoCallInvitationDialog][onInvitationReceived] received online call, currentState: ${_reactNative.AppState.currentState}`);
            if (_reactNative.AppState.currentState === 'background' || _reactNative.AppState.currentState === 'inactive') {
              (0, _logger.zloginfo)('[ZegoCallInvitationDialog][onInvitationReceived] show background online notification');
              _notification_helper.default.getInstance().showOnlineNotification(invitationID, {
                roomID: currentData.call_id,
                type,
                inviter,
                invitees: currentData.invitees
              });
            }
            (0, _logger.zloginfo)('[ZegoCallInvitationDialog][onInvitationReceived] show foreground online notification');
            setCallID(invitationID);
            setInviteType(type);
            setInviter(inviter);
            setExtendData(JSON.parse(data));
            setIsDialogVisable(true);
            _report.default.reportEvent('call/displayNotification', {
              'call_id': invitationID,
              'app_state': _reactNative.AppState.currentState
            });
            _bell.default.vibrate();
            if (_reactNative.AppState.currentState !== 'background') {
              _bell.default.playIncomingSound();
            }

            // Unable to receive timeout due to network issues; added local timeout to turn off incoming call reminder.
            let timeout = (currentData.timeout ?? 60) + 5;
            (0, _logger.zloginfo)(`[ZegoCallInvitationDialog][onInvitationReceived] will config local timeout process after ${timeout} seconds, callID: ${invitationID}`);
            setTimeout(() => {
              (0, _logger.zloginfo)(`[ZegoCallInvitationDialog][onInvitationReceived] local timeout process, current callID: ${zimCallIDRef.current}, isDialogVisable: ${isDialogVisableRef.current}`);
              if (isDialogVisableRef.current && zimCallIDRef.current === invitationID) {
                dismissForegroundNotification();
                _notification_helper.default.getInstance().dismissNotification(invitationID, 'Timeout', TAG);
              }
            }, timeout * 1000);
          }
        }
        (0, _logger.zloginfo)(`onInvitationReceived implement by ${TAG}, process done`);
      }, TAG);
      _zegoUikitRn.default.getSignalingPlugin().onInvitationTimeout(callbackID, ({
        callID: _zimCallID,
        data
      }) => {
        (0, _logger.zloginfo)(`onInvitationTimeout implement by ${TAG}, callID: ${_zimCallID}, data: ${JSON.stringify(data)}`);
        const dataParsed = data ? JSON.parse(data) : {};
        const dataInDataParsed = dataParsed.data ? JSON.parse(dataParsed.data) : {};
        _callevent_notify_app.default.getInstance().notifyEvent('onIncomingCallTimeout', dataInDataParsed.call_id, dataParsed.inviter);
        if (_zimCallID === zimCallIDRef.current) {
          dismissForegroundNotification();
        }
        _notification_helper.default.getInstance().dismissNotification(_zimCallID, 'Timeout', TAG);
        (0, _logger.zloginfo)(`onInvitationTimeout implement by ${TAG}, process done`);
      });
      _zegoUikitRn.default.getSignalingPlugin().onInvitationCanceled(callbackID, ({
        callID,
        inviter,
        data
      }) => {
        (0, _logger.zloginfo)(`onInvitationCanceled implement by ${TAG}, callID: ${callID}, inviter: ${JSON.stringify(inviter)} data: ${data}`);
        let dataParsed = JSON.parse(data);
        if (_plugins.default.getLocalUser().userID != inviter.id) {
          _callevent_notify_app.default.getInstance().notifyEvent('onIncomingCallCanceled', dataParsed.call_id, dataParsed.inviter);
        }
        if (callID === zimCallIDRef.current) {
          dismissForegroundNotification();
        }
        _notification_helper.default.getInstance().dismissNotification(callID, 'BeCancelled', TAG);
        (0, _logger.zloginfo)(`onInvitationCanceled implement by ${TAG}, process done`);
      });
      return () => {
        _zegoUikitRn.default.getSignalingPlugin().onInvitationReceived(callbackID);
        _zegoUikitRn.default.getSignalingPlugin().onInvitationTimeout(callbackID);
        _zegoUikitRn.default.getSignalingPlugin().onInvitationCanceled(callbackID);
        _bell.default.stopIncomingSound();
        _bell.default.cancleVirate();
      };
    }
  }, [isInit]);
  (0, _react.useEffect)(() => {
    zimCallIDRef.current = callID;
  }, [callID]);
  (0, _react.useEffect)(() => {
    isDialogVisableRef.current = isDialogVisable;
  }, [isDialogVisable]);
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: [styles.container, isDialogVisable ? styles.show : null]
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Modal, {
    animationType: "fade",
    transparent: true,
    visible: isDialogVisable,
    style: styles.modal
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.mask
  }), !isFullScreen ? /*#__PURE__*/_react.default.createElement(_reactNative.TouchableOpacity, {
    onPress: pressHandle,
    activeOpacity: 0.9
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.dialog
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.left
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.avatar
  }, !avatarBuilder ? /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: styles.nameLabel
  }, getShortName(inviter.name), " ") : /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: {
      width: '100%',
      height: '100%',
      alignItems: 'center',
      justifyContent: 'center'
    }
  }, /*#__PURE__*/_react.default.createElement(_reactDelegateComponent.default, {
    to: avatarBuilder,
    props: {
      userInfo: {
        userID: inviter.id,
        userName: inviter.name
      }
    }
  }))), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.callInfo
  }, /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: styles.callName,
    numberOfLines: 1,
    ellipsizeMode: "tail"
  }, getDialogTitle()), /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: styles.callTitle
  }, getDialogMessage()))), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.right
  }, showDeclineButton ? /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.refuse
  }, /*#__PURE__*/_react.default.createElement(_zegoUikitRn.ZegoRefuseInvitationButton, {
    inviterID: inviter.id,
    onPressed: refuseSuccHandle,
    onFailure: refuseFailHandle,
    data: JSON.stringify({
      inviterID: inviter.id,
      reason: 'decline',
      callID,
      call_id: extendData.call_id,
      'invitee': {
        userID: _plugins.default.getLocalUser().userID,
        userName: _plugins.default.getLocalUser().userName
      }
    })
  })) : null, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.accept
  }, /*#__PURE__*/_react.default.createElement(_zegoUikitRn.ZegoAcceptInvitationButton, {
    icon: getImageSourceByPath(),
    inviterID: inviter.id,
    data: JSON.stringify({
      call_id: extendData.call_id,
      'invitee': {
        userID: _plugins.default.getLocalUser().userID,
        userName: _plugins.default.getLocalUser().userName
      }
    }),
    onPressed: acceptHandle,
    onFailure: acceptFailHandle
  }))))) : /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.fullDialog
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.content
  }, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.fullAvatar
  }, !avatarBuilder ? /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: styles.fullNameLabel
  }, getShortName(inviter.name)) : /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: {
      width: '100%',
      height: '100%',
      alignItems: 'center',
      justifyContent: 'center'
    }
  }, /*#__PURE__*/_react.default.createElement(_reactDelegateComponent.default, {
    to: avatarBuilder,
    props: {
      userInfo: {
        userID: inviter.id,
        userName: inviter.name
      }
    }
  }))), /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: styles.fullCallName
  }, getDialogTitle()), /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: styles.calling
  }, getDialogMessage())), /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.bottomBarContainer
  }, showDeclineButton ? /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.fullRefuse
  }, /*#__PURE__*/_react.default.createElement(_zegoUikitRn.ZegoRefuseInvitationButton, {
    inviterID: inviter.id,
    onPressed: refuseSuccHandle,
    onFailure: refuseFailHandle,
    data: JSON.stringify({
      inviterID: inviter.id,
      reason: 'decline',
      callID,
      call_id: extendData.call_id,
      'invitee': {
        userID: _plugins.default.getLocalUser().userID,
        userName: _plugins.default.getLocalUser().userName
      }
    })
  }), /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: styles.fullRefuseTitle
  }, _inner_text_helper.default.instance().getInnerText().incomingCallPageDeclineButton)) : null, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.fullAccept
  }, /*#__PURE__*/_react.default.createElement(_zegoUikitRn.ZegoAcceptInvitationButton, {
    icon: getImageSourceByPath(),
    inviterID: inviter.id,
    data: JSON.stringify({
      call_id: extendData.call_id,
      'invitee': {
        userID: _plugins.default.getLocalUser().userID,
        userName: _plugins.default.getLocalUser().userName
      }
    }),
    onPressed: acceptHandle,
    onFailure: acceptFailHandle
  }), /*#__PURE__*/_react.default.createElement(_reactNative.Text, {
    style: styles.fullAcceptTitle
  }, _inner_text_helper.default.instance().getInnerText().incomingCallPageAcceptButton))))));
}
const styles = _reactNative.StyleSheet.create({
  container: {
    flex: 1,
    position: 'absolute',
    zIndex: 10000,
    alignItems: 'center'
  },
  mask: {
    position: 'absolute',
    top: 0,
    bottom: 0,
    right: 0,
    left: 0,
    backgroundColor: '#333333',
    opacity: 0.4
  },
  dialog: {
    height: 80,
    backgroundColor: '#333333',
    borderRadius: 8,
    marginTop: 40,
    flexDirection: 'row',
    justifyContent: 'space-between',
    paddingLeft: 12,
    paddingRight: 12,
    marginLeft: 8,
    marginRight: 8
  },
  left: {
    flexDirection: 'row',
    alignItems: 'center',
    flex: 1
  },
  avatar: {
    backgroundColor: '#ffffff',
    borderRadius: 1000,
    color: '#2A2A2A',
    width: 42,
    height: 42,
    justifyContent: 'center',
    alignItems: 'center',
    marginRight: 13,
    overflow: 'hidden'
  },
  callInfo: {
    flexDirection: 'column',
    marginRight: 5,
    flex: 1
  },
  nameLabel: {
    fontSize: 26
  },
  callName: {
    fontSize: 24,
    color: '#ffffff',
    lineHeight: 25,
    marginBottom: 3.5
  },
  callTitle: {
    fontSize: 12,
    color: '#ffffff',
    opacity: 0.8,
    lineHeight: 16.5
  },
  right: {
    flexDirection: 'row',
    alignItems: 'center'
  },
  refuse: {
    marginRight: 20
  },
  fullDialog: {
    backgroundColor: '#333333',
    width: '100%',
    height: '100%'
  },
  content: {
    width: '100%',
    alignItems: 'center',
    marginTop: 114
  },
  fullAvatar: {
    width: 100,
    height: 100,
    borderRadius: 1000,
    backgroundColor: '#ffffff',
    alignItems: 'center',
    justifyContent: 'center',
    marginBottom: 5,
    overflow: 'hidden'
  },
  fullNameLabel: {
    color: '#222222',
    fontSize: 22
  },
  fullCallName: {
    fontSize: 21,
    lineHeight: 29.5,
    marginBottom: 23.5,
    color: '#ffffff'
  },
  calling: {
    fontSize: 16,
    lineHeight: 22.5,
    color: '#ffffff',
    opacity: 0.7
  },
  bottomBarContainer: {
    width: '100%',
    position: 'absolute',
    bottom: 100,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-around',
    paddingLeft: 30,
    paddingRight: 30
  },
  fullRefuse: {},
  fullRefuseTitle: {
    fontSize: 13,
    opacity: 0.7,
    color: '#FFFFFF',
    lineHeight: 18.5,
    marginTop: 7.5
  },
  fullAccept: {},
  fullAcceptTitle: {
    fontSize: 13,
    opacity: 0.7,
    color: '#FFFFFF',
    lineHeight: 18.5,
    marginTop: 7.5
  }
});
