// @zegocloud/zego-uikit-prebuilt-call-rn/lib/commonjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "GROUP_VIDEO_CALL_CONFIG", {
  enumerable: true,
  get: function () {
    return _defines2.GROUP_VIDEO_CALL_CONFIG;
  }
});
Object.defineProperty(exports, "GROUP_VOICE_CALL_CONFIG", {
  enumerable: true,
  get: function () {
    return _defines2.GROUP_VOICE_CALL_CONFIG;
  }
});
Object.defineProperty(exports, "ONE_ON_ONE_VIDEO_CALL_CONFIG", {
  enumerable: true,
  get: function () {
    return _defines2.ONE_ON_ONE_VIDEO_CALL_CONFIG;
  }
});
Object.defineProperty(exports, "ONE_ON_ONE_VOICE_CALL_CONFIG", {
  enumerable: true,
  get: function () {
    return _defines2.ONE_ON_ONE_VOICE_CALL_CONFIG;
  }
});
Object.defineProperty(exports, "ZegoCallEndReason", {
  enumerable: true,
  get: function () {
    return _defines2.ZegoCallEndReason;
  }
});
Object.defineProperty(exports, "ZegoCallInvitationDialog", {
  enumerable: true,
  get: function () {
    return _ZegoCallInvitationDialog.default;
  }
});
Object.defineProperty(exports, "ZegoCountdownLabel", {
  enumerable: true,
  get: function () {
    return _ZegoCountdownLabel.default;
  }
});
Object.defineProperty(exports, "ZegoInvitationType", {
  enumerable: true,
  get: function () {
    return _defines.ZegoInvitationType;
  }
});
Object.defineProperty(exports, "ZegoMenuBarButtonName", {
  enumerable: true,
  get: function () {
    return _ZegoMenuBarButtonName.default;
  }
});
Object.defineProperty(exports, "ZegoMenuBarStyle", {
  enumerable: true,
  get: function () {
    return _ZegoMenuBarStyle.default;
  }
});
Object.defineProperty(exports, "ZegoMultiCertificate", {
  enumerable: true,
  get: function () {
    return _defines2.ZegoMultiCertificate;
  }
});
Object.defineProperty(exports, "ZegoSendCallInvitationButton", {
  enumerable: true,
  get: function () {
    return _ZegoSendCallInvitationButton.default;
  }
});
Object.defineProperty(exports, "ZegoUIKitPrebuiltCall", {
  enumerable: true,
  get: function () {
    return _call.default;
  }
});
Object.defineProperty(exports, "ZegoUIKitPrebuiltCallFloatingMinimizedView", {
  enumerable: true,
  get: function () {
    return _ZegoUIKitPrebuiltCallFloatingMinimizedView.default;
  }
});
Object.defineProperty(exports, "ZegoUIKitPrebuiltCallInCallScreen", {
  enumerable: true,
  get: function () {
    return _ZegoCallInvitationRoom.default;
  }
});
Object.defineProperty(exports, "ZegoUIKitPrebuiltCallWaitingScreen", {
  enumerable: true,
  get: function () {
    return _ZegoCallInvitationWaiting.default;
  }
});
exports.default = void 0;
var _call = _interopRequireDefault(require("./call"));
var _ZegoUIKitPrebuiltCallFloatingMinimizedView = _interopRequireDefault(require("./call/ZegoUIKitPrebuiltCallFloatingMinimizedView"));
var _ZegoCallInvitationDialog = _interopRequireDefault(require("./call_invitation/components/ZegoCallInvitationDialog"));
var _ZegoCallInvitationWaiting = _interopRequireDefault(require("./call_invitation/pages/ZegoCallInvitationWaiting"));
var _ZegoCallInvitationRoom = _interopRequireDefault(require("./call_invitation/pages/ZegoCallInvitationRoom"));
var _defines = require("./call_invitation/services/defines");
var _ZegoMenuBarButtonName = _interopRequireDefault(require("./call/ZegoMenuBarButtonName"));
var _ZegoMenuBarStyle = _interopRequireDefault(require("./call/ZegoMenuBarStyle"));
var _ZegoSendCallInvitationButton = _interopRequireDefault(require("./call_invitation/components/ZegoSendCallInvitationButton"));
var _services = _interopRequireDefault(require("./services"));
var _ZegoCountdownLabel = _interopRequireDefault(require("./call/ZegoCountdownLabel"));
var _defines2 = require("./services/defines");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
var _default = exports.default = _services.default.getInstance();
